--- a/net/minecraft/core/particles/ParticleTypes.java
+++ b/net/minecraft/core/particles/ParticleTypes.java
@@ -4,6 +_,7 @@
 import java.util.function.Function;
 import net.minecraft.core.Registry;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class ParticleTypes {
    public static final SimpleParticleType AMBIENT_ENTITY_EFFECT = register("ambient_entity_effect", false);
    public static final SimpleParticleType ANGRY_VILLAGER = register("angry_villager", false);
