--- a/net/minecraft/data/tags/ItemTagsProvider.java
+++ b/net/minecraft/data/tags/ItemTagsProvider.java
@@ -14,8 +_,14 @@
 public class ItemTagsProvider extends TagsProvider<Item> {
    private final Function<TagKey<Block>, Tag.Builder> blockTags;
 
+   /** @deprecated Forge: Use the {@link #ItemTagsProvider(DataGenerator, BlockTagsProvider, String, net.minecraftforge.common.data.ExistingFileHelper) mod id variant} */
+   @Deprecated
    public ItemTagsProvider(DataGenerator p_126530_, BlockTagsProvider p_126531_) {
       super(p_126530_, Registry.ITEM);
+      this.blockTags = p_126531_::getOrCreateRawBuilder;
+   }
+   public ItemTagsProvider(DataGenerator p_126530_, BlockTagsProvider p_126531_, String modId, @javax.annotation.Nullable net.minecraftforge.common.data.ExistingFileHelper existingFileHelper) {
+      super(p_126530_, Registry.ITEM, modId, existingFileHelper);
       this.blockTags = p_126531_::getOrCreateRawBuilder;
    }
 
