--- a/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -33,7 +_,44 @@
          this.connection.disconnect(DISCONNECT_REASON);
       } else {
          this.hasRequestedStatus = true;
-         this.connection.send(new ClientboundStatusResponsePacket(this.server.getStatus()));
+//         this.connection.send(new ClientboundStatusResponsePacket(this.server.getStatus()));
+         // CraftBukkit start
+         org.magmafoundation.magma.craftbukkit.ServerListPingEventImpl event = new org.magmafoundation.magma.craftbukkit.ServerListPingEventImpl(connection, server);
+         this.server.server.getPluginManager().callEvent(event);
+
+         java.util.List<com.mojang.authlib.GameProfile> profiles = new java.util.ArrayList<com.mojang.authlib.GameProfile>(event.getPlayers().length);
+         for (Object player : event.getPlayers()) {
+            if (player != null) {
+               net.minecraft.server.level.ServerPlayer entityPlayer = ((net.minecraft.server.level.ServerPlayer) player);
+               if (entityPlayer.allowsListing()) {
+                  profiles.add(entityPlayer.getGameProfile());
+               } else {
+                  profiles.add(MinecraftServer.ANONYMOUS_PLAYER_PROFILE);
+               }
+            }
+         }
+
+         net.minecraft.network.protocol.status.ServerStatus.Players playerSample = new net.minecraft.network.protocol.status.ServerStatus.Players(event.getMaxPlayers(), profiles.size());
+         if (!this.server.hidesOnlinePlayers()) {
+            // Spigot Start
+            if ( !profiles.isEmpty() )
+            {
+               java.util.Collections.shuffle( profiles ); // This sucks, its inefficient but we have no simple way of doing it differently
+               profiles = profiles.subList( 0, Math.min( profiles.size(), org.spigotmc.SpigotConfig.playerSample ) ); // Cap the sample to n (or less) displayed players, ie: Vanilla behaviour
+            }
+            // Spigot End
+            playerSample.setSample(profiles.toArray(new com.mojang.authlib.GameProfile[profiles.size()]));
+         }
+
+         net.minecraft.network.protocol.status.ServerStatus ping = new net.minecraft.network.protocol.status.ServerStatus();
+         ping.setFavicon(event.icon.value);
+         ping.setDescription(org.bukkit.craftbukkit.v1_18_R2.util.CraftChatMessage.fromString(event.getMotd(), true)[0]);
+         ping.setPlayers(playerSample);
+         int version = net.minecraft.SharedConstants.getCurrentVersion().getProtocolVersion();
+         ping.setVersion(new net.minecraft.network.protocol.status.ServerStatus.Version(server.getServerModName() + " " + server.getServerVersion(), version));
+         ping.setForgeData(this.server.getStatus().getForgeData());
+         this.connection.send(new ClientboundStatusResponsePacket(ping));
+         // CraftBukkit end
       }
    }
 
