--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -118,11 +_,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.AbstractArrow;
 import net.minecraft.world.entity.vehicle.Boat;
-import net.minecraft.world.inventory.AbstractContainerMenu;
-import net.minecraft.world.inventory.AnvilMenu;
-import net.minecraft.world.inventory.BeaconMenu;
-import net.minecraft.world.inventory.MerchantMenu;
-import net.minecraft.world.inventory.RecipeBookMenu;
+import net.minecraft.world.inventory.*;
 import net.minecraft.world.item.BlockItem;
 import net.minecraft.world.item.BucketItem;
 import net.minecraft.world.item.Item;
@@ -141,7 +_,6 @@
 import net.minecraft.world.level.block.entity.JigsawBlockEntity;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
 import net.minecraft.world.level.block.entity.StructureBlockEntity;
-import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
@@ -149,7 +_,18 @@
 import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
-import org.apache.commons.lang3.StringUtils;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_18_R2.CraftServer;
+import org.bukkit.craftbukkit.v1_18_R2.entity.CraftPlayer;
+import org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_18_R2.util.CraftMagicNumbers;
+import org.bukkit.craftbukkit.v1_18_R2.util.LazyPlayerSet;
+import org.bukkit.event.inventory.*;
+import org.bukkit.event.inventory.ClickType;
+import org.bukkit.event.player.PlayerCommandPreprocessEvent;
+import org.bukkit.inventory.CraftingInventory;
+import org.bukkit.inventory.InventoryView;
+import org.bukkit.inventory.SmithingInventory;
 import org.slf4j.Logger;
 
 public class ServerGamePacketListenerImpl implements ServerPlayerConnection, ServerGamePacketListener {
@@ -189,6 +_,20 @@
    private int f_9740_;
    private int f_9741_;
 
+   // CraftBukkit start
+   private final CraftServer cserver;
+   public boolean processedDisconnect;
+   private int lastTick = MinecraftServer.currentTick;
+   private int allowedPlayerTicks = 1;
+   private int lastDropTick = MinecraftServer.currentTick;
+   private int lastBookTick  = MinecraftServer.currentTick;
+   private int dropCount = 0;
+
+   public CraftPlayer getCraftPlayer() {
+      return (this.f_9743_ == null) ? null : (CraftPlayer) this.f_9743_.getBukkitEntity();
+   }
+   // CraftBukkit end
+
    public ServerGamePacketListenerImpl(MinecraftServer p_9770_, Connection p_9771_, ServerPlayer p_9772_) {
       this.f_9745_ = p_9770_;
       this.f_9742_ = p_9771_;
@@ -197,6 +_,8 @@
       p_9772_.f_8906_ = this;
       this.f_9747_ = Util.m_137550_();
       p_9772_.m_8967_().m_7674_();
+
+      this.cserver = p_9770_.server; // CraftBukkit
    }
 
    public void m_9933_() {
@@ -386,9 +_,11 @@
             }
 
             entity.m_19890_(d3, d4, d5, f, f1);
+            this.f_9743_.m_19890_(d3, d4, d5, this.f_9743_.m_146908_(), this.f_9743_.m_146909_()); // Forge - Resync player position on vehicle moving
             boolean flag3 = serverlevel.m_45756_(entity, entity.m_142469_().m_82406_(0.0625D));
             if (flag && (flag2 || !flag3)) {
                entity.m_19890_(d0, d1, d2, f, f1);
+               this.f_9743_.m_19890_(d3, d4, d5, this.f_9743_.m_146908_(), this.f_9743_.m_146909_()); // Forge - Resync player position on vehicle moving
                this.f_9742_.m_129512_(new ClientboundMoveVehiclePacket(entity));
                return;
             }
@@ -405,7 +_,7 @@
    }
 
    private boolean m_9793_(Entity p_9794_) {
-      return p_9794_.f_19853_.m_45556_(p_9794_.m_142469_().m_82400_(0.0625D).m_82363_(0.0D, -0.55D, 0.0D)).allMatch(BlockBehaviour.BlockStateBase::m_60795_);
+      return BlockPos.m_121921_(p_9794_.m_142469_().m_82400_(0.0625D).m_82363_(0.0D, -0.55D, 0.0D)).allMatch(b -> p_9794_.f_19853_.m_8055_(b).m_60795_());
    }
 
    public void m_7376_(ServerboundAcceptTeleportationPacket p_9835_) {
@@ -979,7 +_,9 @@
             this.f_9743_.m_9243_();
             int i = this.f_9743_.f_19853_.m_151558_();
             if (blockpos.m_123342_() < i) {
-               if (this.f_9766_ == null && this.f_9743_.m_20275_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.5D, (double)blockpos.m_123343_() + 0.5D) < 64.0D && serverlevel.m_7966_(this.f_9743_, blockpos)) {
+               double dist = f_9743_.m_21051_(net.minecraftforge.common.ForgeMod.REACH_DISTANCE.get()).m_22135_() + 3;
+               dist *= dist;
+               if (this.f_9766_ == null && this.f_9743_.m_20275_((double)blockpos.m_123341_() + 0.5D, (double)blockpos.m_123342_() + 0.5D, (double)blockpos.m_123343_() + 0.5D) < dist && serverlevel.m_7966_(this.f_9743_, blockpos)) {
                   InteractionResult interactionresult = this.f_9743_.f_8941_.m_7179_(this.f_9743_, serverlevel, itemstack, interactionhand, blockhitresult);
                   if (direction == Direction.UP && !interactionresult.m_19077_() && blockpos.m_123342_() >= i - 1 && m_9790_(this.f_9743_, itemstack)) {
                      Component component = (new TranslatableComponent("build.tooHigh", i - 1)).m_130940_(ChatFormatting.RED);
@@ -1099,7 +_,7 @@
    }
 
    public void m_7388_(ServerboundChatPacket p_9841_) {
-      String s = StringUtils.normalizeSpace(p_9841_.m_133837_());
+      String s = org.apache.commons.lang3.StringUtils.normalizeSpace(p_9841_.m_133837_());
 
       for(int i = 0; i < s.length(); ++i) {
          if (!SharedConstants.m_136188_(s.charAt(i))) {
@@ -1129,8 +_,11 @@
             String s1 = p_143629_.m_143722_();
             Component component = s1.isEmpty() ? null : new TranslatableComponent("chat.type.text", this.f_9743_.m_5446_(), s1);
             Component component1 = new TranslatableComponent("chat.type.text", this.f_9743_.m_5446_(), s);
-            this.f_9745_.m_6846_().m_143991_(component1, (p_184197_) -> {
-               return this.f_9743_.m_143421_(p_184197_) ? component : component1;
+            component1 = net.minecraftforge.common.ForgeHooks.onServerChatEvent(this, s, component1);
+            Component finalComponent = component1;
+            if (finalComponent != null)
+               this.f_9745_.m_6846_().m_143991_(component1, (p_184197_) -> {
+               return this.f_9743_.m_143421_(p_184197_) ? component : finalComponent;
             }, ChatType.CHAT, this.f_9743_.m_142081_());
          }
 
@@ -1143,7 +_,29 @@
    }
 
    private void m_9957_(String p_9958_) {
-      this.f_9745_.m_129892_().m_82117_(this.f_9743_.m_20203_(), p_9958_);
+      // CraftBukkit start - whole method
+      this.f_9744_.info(this.f_9743_.m_6302_() + " issued server command: " + p_9958_);
+
+      CraftPlayer player = this.getCraftPlayer();
+
+      PlayerCommandPreprocessEvent event = new PlayerCommandPreprocessEvent(player, p_9958_, new LazyPlayerSet(f_9745_));
+      this.cserver.getPluginManager().callEvent(event);
+
+      if (event.isCancelled()) {
+         return;
+      }
+
+      try {
+         if (this.cserver.dispatchCommand(event.getPlayer(), event.getMessage().substring(1))) {
+            return;
+         }
+      } catch (org.bukkit.command.CommandException ex) {
+         player.sendMessage(org.bukkit.ChatColor.RED + "An internal error occurred while attempting to perform this command");
+         java.util.logging.Logger.getLogger(ServerGamePacketListenerImpl.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
+         return;
+      }
+      // this.server.getCommands().performCommand(this.player.createCommandSourceStack(), p_9958_);
+      // CraftBukkit end
    }
 
    public void m_7953_(ServerboundSwingPacket p_9926_) {
@@ -1222,6 +_,7 @@
                private void m_143678_(InteractionHand p_143679_, ServerGamePacketListenerImpl.EntityInteraction p_143680_) {
                   ItemStack itemstack = ServerGamePacketListenerImpl.this.f_9743_.m_21120_(p_143679_).m_41777_();
                   InteractionResult interactionresult = p_143680_.m_143694_(ServerGamePacketListenerImpl.this.f_9743_, entity, p_143679_);
+                  if (net.minecraftforge.common.ForgeHooks.onInteractEntityAt(f_9743_, entity, entity.m_20182_(), p_143679_) != null) return;
                   if (interactionresult.m_19077_()) {
                      CriteriaTriggers.f_10565_.m_61494_(ServerGamePacketListenerImpl.this.f_9743_, itemstack, entity);
                      if (interactionresult.m_19080_()) {
@@ -1290,9 +_,11 @@
 
    public void m_5914_(ServerboundContainerClickPacket p_9856_) {
       PacketUtils.m_131359_(p_9856_, this, this.f_9743_.m_183503_());
+      if (this.f_9743_.m_6107_()) return; // CraftBukkit
       this.f_9743_.m_9243_();
-      if (this.f_9743_.f_36096_.f_38840_ == p_9856_.m_133959_()) {
-         if (this.f_9743_.m_5833_()) {
+      if (this.f_9743_.f_36096_.f_38840_ == p_9856_.m_133959_() && this.f_9743_.f_36096_.m_6875_(this.f_9743_)) { // CraftBukkit
+         boolean cancelled = this.f_9743_.m_5833_(); // CraftBukkit - see below if
+         if (false/*this.player.isSpectator()*/) { // CraftBukkit
             this.f_9743_.f_36096_.m_150429_();
          } else {
             int i = p_9856_.m_133962_();
@@ -1301,7 +_,269 @@
             } else {
                boolean flag = p_9856_.m_182741_() != this.f_9743_.f_36096_.m_182424_();
                this.f_9743_.f_36096_.m_150443_();
-               this.f_9743_.f_36096_.m_150399_(i, p_9856_.m_133963_(), p_9856_.m_133966_(), this.f_9743_);
+//               this.player.containerMenu.clicked(i, p_9856_.getButtonNum(), p_9856_.getClickType(), this.player);
+
+               // CraftBukkit start - Call InventoryClickEvent
+               if (p_9856_.m_133962_() < -1 && p_9856_.m_133962_() != -999) {
+                  return;
+               }
+
+               InventoryView inventory = this.f_9743_.f_36096_.getBukkitView();
+               InventoryType.SlotType type = inventory.getSlotType(p_9856_.m_133962_());
+
+               InventoryClickEvent event;
+               ClickType click = ClickType.UNKNOWN;
+               InventoryAction action = InventoryAction.UNKNOWN;
+
+               ItemStack itemstack = ItemStack.f_41583_;
+
+               switch (p_9856_.m_133966_()) {
+                  case PICKUP:
+                     if (p_9856_.m_133963_() == 0) {
+                        click = ClickType.LEFT;
+                     } else if (p_9856_.m_133963_() == 1) {
+                        click = ClickType.RIGHT;
+                     }
+                     if (p_9856_.m_133963_() == 0 || p_9856_.m_133963_() == 1) {
+                        action = InventoryAction.NOTHING; // Don't want to repeat ourselves
+                        if (p_9856_.m_133962_() == -999) {
+                           if (!f_9743_.f_36096_.m_142621_().m_41619_()) {
+                              action = p_9856_.m_133963_() == 0 ? InventoryAction.DROP_ALL_CURSOR : InventoryAction.DROP_ONE_CURSOR;
+                           }
+                        } else if (p_9856_.m_133962_() < 0)  {
+                           action = InventoryAction.NOTHING;
+                        } else {
+                           Slot slot = this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_());
+                           if (slot != null) {
+                              ItemStack clickedItem = slot.m_7993_();
+                              ItemStack cursor = f_9743_.f_36096_.m_142621_();
+                              if (clickedItem.m_41619_()) {
+                                 if (!cursor.m_41619_()) {
+                                    action = p_9856_.m_133963_() == 0 ? InventoryAction.PLACE_ALL : InventoryAction.PLACE_ONE;
+                                 }
+                              } else if (slot.m_8010_(f_9743_)) {
+                                 if (cursor.m_41619_()) {
+                                    action = p_9856_.m_133963_() == 0 ? InventoryAction.PICKUP_ALL : InventoryAction.PICKUP_HALF;
+                                 } else if (slot.m_5857_(cursor)) {
+                                    if (clickedItem.m_41656_(cursor) && ItemStack.m_41658_(clickedItem, cursor)) {
+                                       int toPlace = p_9856_.m_133963_() == 0 ? cursor.m_41613_() : 1;
+                                       toPlace = Math.min(toPlace, clickedItem.m_41741_() - clickedItem.m_41613_());
+                                       toPlace = Math.min(toPlace, slot.f_40218_.m_6893_() - clickedItem.m_41613_());
+                                       if (toPlace == 1) {
+                                          action = InventoryAction.PLACE_ONE;
+                                       } else if (toPlace == cursor.m_41613_()) {
+                                          action = InventoryAction.PLACE_ALL;
+                                       } else if (toPlace < 0) {
+                                          action = toPlace != -1 ? InventoryAction.PICKUP_SOME : InventoryAction.PICKUP_ONE; // this happens with oversized stacks
+                                       } else if (toPlace != 0) {
+                                          action = InventoryAction.PLACE_SOME;
+                                       }
+                                    } else if (cursor.m_41613_() <= slot.m_6641_()) {
+                                       action = InventoryAction.SWAP_WITH_CURSOR;
+                                    }
+                                 } else if (cursor.m_41720_() == clickedItem.m_41720_() && ItemStack.m_41658_(cursor, clickedItem)) {
+                                    if (clickedItem.m_41613_() >= 0) {
+                                       if (clickedItem.m_41613_() + cursor.m_41613_() <= cursor.m_41741_()) {
+                                          // As of 1.5, this is result slots only
+                                          action = InventoryAction.PICKUP_ALL;
+                                       }
+                                    }
+                                 }
+                              }
+                           }
+                        }
+                     }
+                     break;
+                  // TODO check on updates
+                  case QUICK_MOVE:
+                     if (p_9856_.m_133963_() == 0) {
+                        click = ClickType.SHIFT_LEFT;
+                     } else if (p_9856_.m_133963_() == 1) {
+                        click = ClickType.SHIFT_RIGHT;
+                     }
+                     if (p_9856_.m_133963_() == 0 || p_9856_.m_133963_() == 1) {
+                        if (p_9856_.m_133962_() < 0) {
+                           action = InventoryAction.NOTHING;
+                        } else {
+                           Slot slot = this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_());
+                           if (slot != null && slot.m_8010_(this.f_9743_) && slot.m_6657_()) {
+                              action = InventoryAction.MOVE_TO_OTHER_INVENTORY;
+                           } else {
+                              action = InventoryAction.NOTHING;
+                           }
+                        }
+                     }
+                     break;
+                  case SWAP:
+                     if ((p_9856_.m_133963_() >= 0 && p_9856_.m_133963_() < 9) || p_9856_.m_133963_() == 40) {
+                        click = (p_9856_.m_133963_() == 40) ? ClickType.SWAP_OFFHAND : ClickType.NUMBER_KEY;
+                        Slot clickedSlot = this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_());
+                        if (clickedSlot.m_8010_(f_9743_)) {
+                           ItemStack hotbar = this.f_9743_.m_150109_().m_8020_(p_9856_.m_133963_());
+                           boolean canCleanSwap = hotbar.m_41619_() || (clickedSlot.f_40218_ == f_9743_.m_150109_() && clickedSlot.m_5857_(hotbar)); // the slot will accept the hotbar item
+                           if (clickedSlot.m_6657_()) {
+                              if (canCleanSwap) {
+                                 action = InventoryAction.HOTBAR_SWAP;
+                              } else {
+                                 action = InventoryAction.HOTBAR_MOVE_AND_READD;
+                              }
+                           } else if (!clickedSlot.m_6657_() && !hotbar.m_41619_() && clickedSlot.m_5857_(hotbar)) {
+                              action = InventoryAction.HOTBAR_SWAP;
+                           } else {
+                              action = InventoryAction.NOTHING;
+                           }
+                        } else {
+                           action = InventoryAction.NOTHING;
+                        }
+                     }
+                     break;
+                  case CLONE:
+                     if (p_9856_.m_133963_() == 2) {
+                        click = ClickType.MIDDLE;
+                        if (p_9856_.m_133962_() < 0) {
+                           action = InventoryAction.NOTHING;
+                        } else {
+                           Slot slot = this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_());
+                           if (slot != null && slot.m_6657_() && f_9743_.m_150110_().f_35937_ && f_9743_.f_36096_.m_142621_().m_41619_()) {
+                              action = InventoryAction.CLONE_STACK;
+                           } else {
+                              action = InventoryAction.NOTHING;
+                           }
+                        }
+                     } else {
+                        click = ClickType.UNKNOWN;
+                        action = InventoryAction.UNKNOWN;
+                     }
+                     break;
+                  case THROW:
+                     if (p_9856_.m_133962_() >= 0) {
+                        if (p_9856_.m_133963_() == 0) {
+                           click = ClickType.DROP;
+                           Slot slot = this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_());
+                           if (slot != null && slot.m_6657_() && slot.m_8010_(f_9743_) && !slot.m_7993_().m_41619_() && slot.m_7993_().m_41720_() != Item.m_41439_(Blocks.f_50016_)) {
+                              action = InventoryAction.DROP_ONE_SLOT;
+                           } else {
+                              action = InventoryAction.NOTHING;
+                           }
+                        } else if (p_9856_.m_133963_() == 1) {
+                           click = ClickType.CONTROL_DROP;
+                           Slot slot = this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_());
+                           if (slot != null && slot.m_6657_() && slot.m_8010_(f_9743_) && !slot.m_7993_().m_41619_() && slot.m_7993_().m_41720_() != Item.m_41439_(Blocks.f_50016_)) {
+                              action = InventoryAction.DROP_ALL_SLOT;
+                           } else {
+                              action = InventoryAction.NOTHING;
+                           }
+                        }
+                     } else {
+                        // Sane default (because this happens when they are holding nothing. Don't ask why.)
+                        click = ClickType.LEFT;
+                        if (p_9856_.m_133963_() == 1) {
+                           click = ClickType.RIGHT;
+                        }
+                        action = InventoryAction.NOTHING;
+                     }
+                     break;
+                  case QUICK_CRAFT:
+                     this.f_9743_.f_36096_.m_150399_(p_9856_.m_133962_(), p_9856_.m_133963_(), p_9856_.m_133966_(), this.f_9743_);
+                     break;
+                  case PICKUP_ALL:
+                     click = ClickType.DOUBLE_CLICK;
+                     action = InventoryAction.NOTHING;
+                     if (p_9856_.m_133962_() >= 0 && !this.f_9743_.f_36096_.m_142621_().m_41619_()) {
+                        ItemStack cursor = this.f_9743_.f_36096_.m_142621_();
+                        action = InventoryAction.NOTHING;
+                        // Quick check for if we have any of the item
+                        if (inventory.getTopInventory().contains(CraftMagicNumbers.getMaterial(cursor.m_41720_())) || inventory.getBottomInventory().contains(CraftMagicNumbers.getMaterial(cursor.m_41720_()))) {
+                           action = InventoryAction.COLLECT_TO_CURSOR;
+                        }
+                     }
+                     break;
+                  default:
+                     break;
+               }
+               if (p_9856_.m_133966_() != net.minecraft.world.inventory.ClickType.QUICK_CRAFT) {
+                  if (click == ClickType.NUMBER_KEY) {
+                     event = new InventoryClickEvent(inventory, type, p_9856_.m_133962_(), click, action, p_9856_.m_133963_());
+                  } else {
+                     event = new InventoryClickEvent(inventory, type, p_9856_.m_133962_(), click, action);
+                  }
+                  org.bukkit.inventory.Inventory top = inventory.getTopInventory();
+                  if (p_9856_.m_133962_() == 0 && top instanceof CraftingInventory) {
+                     org.bukkit.inventory.Recipe recipe = ((CraftingInventory) top).getRecipe();
+                     if (recipe != null) {
+                        if (click == ClickType.NUMBER_KEY) {
+                           event = new CraftItemEvent(recipe, inventory, type, p_9856_.m_133962_(), click, action, p_9856_.m_133963_());
+                        } else {
+                           event = new CraftItemEvent(recipe, inventory, type, p_9856_.m_133962_(), click, action);
+                        }
+                     }
+                  }
+                  if (p_9856_.m_133962_() == 2 && top instanceof SmithingInventory) {
+                     org.bukkit.inventory.ItemStack result = ((SmithingInventory) top).getResult();
+                     if (result != null) {
+                        if (click == ClickType.NUMBER_KEY) {
+                           event = new SmithItemEvent(inventory, type, p_9856_.m_133962_(), click, action, p_9856_.m_133963_());
+                        } else {
+                           event = new SmithItemEvent(inventory, type, p_9856_.m_133962_(), click, action);
+                        }
+                     }
+                  }
+                  event.setCancelled(cancelled);
+                  AbstractContainerMenu oldContainer = this.f_9743_.f_36096_; // SPIGOT-1224
+                  cserver.getPluginManager().callEvent(event);
+                  if (this.f_9743_.f_36096_ != oldContainer) {
+                     return;
+                  }
+                  switch (event.getResult()) {
+                     case ALLOW:
+                     case DEFAULT:
+                        this.f_9743_.f_36096_.m_150399_(i, p_9856_.m_133963_(), p_9856_.m_133966_(), this.f_9743_);
+                        break;
+                     case DENY:
+                        switch (action) {
+                           // Modified other slots
+                           case PICKUP_ALL:
+                           case MOVE_TO_OTHER_INVENTORY:
+                           case HOTBAR_MOVE_AND_READD:
+                           case HOTBAR_SWAP:
+                           case COLLECT_TO_CURSOR:
+                           case UNKNOWN:
+                              this.f_9743_.f_36096_.m_150429_();
+                              break;
+                           // Modified cursor and clicked
+                           case PICKUP_SOME:
+                           case PICKUP_HALF:
+                           case PICKUP_ONE:
+                           case PLACE_ALL:
+                           case PLACE_SOME:
+                           case PLACE_ONE:
+                           case SWAP_WITH_CURSOR:
+                              this.f_9743_.f_8906_.m_141995_(new ClientboundContainerSetSlotPacket(-1, -1, this.f_9743_.f_36095_.m_182425_(), this.f_9743_.f_36096_.m_142621_()));
+                              this.f_9743_.f_8906_.m_141995_(new ClientboundContainerSetSlotPacket(this.f_9743_.f_36096_.f_38840_, this.f_9743_.f_36095_.m_182425_(), p_9856_.m_133962_(), this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_()).m_7993_()));
+                              break;
+                           // Modified clicked only
+                           case DROP_ALL_SLOT:
+                           case DROP_ONE_SLOT:
+                              this.f_9743_.f_8906_.m_141995_(new ClientboundContainerSetSlotPacket(this.f_9743_.f_36096_.f_38840_, this.f_9743_.f_36095_.m_182425_(), p_9856_.m_133962_(), this.f_9743_.f_36096_.m_38853_(p_9856_.m_133962_()).m_7993_()));
+                              break;
+                           // Modified cursor only
+                           case DROP_ALL_CURSOR:
+                           case DROP_ONE_CURSOR:
+                           case CLONE_STACK:
+                              this.f_9743_.f_8906_.m_141995_(new ClientboundContainerSetSlotPacket(-1, -1, this.f_9743_.f_36095_.m_182425_(), this.f_9743_.f_36096_.m_142621_()));
+                              break;
+                           // Nothing
+                           case NOTHING:
+                              break;
+                        }
+                  }
+                  if (event instanceof CraftItemEvent || event instanceof SmithItemEvent) {
+                     // Need to update the inventory on crafting to
+                     // correctly support custom recipes
+                     f_9743_.f_36096_.m_150429_();
+                  }
+               }
+               // CraftBukkit end
 
                for(Entry<ItemStack> entry : Int2ObjectMaps.fastIterable(p_9856_.m_179582_())) {
                   this.f_9743_.f_36096_.m_182414_(entry.getIntKey(), entry.getValue());
@@ -1332,6 +_,7 @@
 
    public void m_6557_(ServerboundContainerButtonClickPacket p_9854_) {
       PacketUtils.m_131359_(p_9854_, this, this.f_9743_.m_183503_());
+      if (this.f_9743_.m_6107_()) return; // CraftBukkit
       this.f_9743_.m_9243_();
       if (this.f_9743_.f_36096_.f_38840_ == p_9854_.m_133935_() && !this.f_9743_.m_5833_()) {
          boolean flag = this.f_9743_.f_36096_.m_6366_(this.f_9743_, p_9854_.m_133938_());
@@ -1358,6 +_,40 @@
 
          boolean flag1 = p_9915_.m_134561_() >= 1 && p_9915_.m_134561_() <= 45;
          boolean flag2 = itemstack.m_41619_() || itemstack.m_41773_() >= 0 && itemstack.m_41613_() <= 64 && !itemstack.m_41619_();
+         if (flag || (flag1 && !ItemStack.m_41728_(this.f_9743_.f_36095_.m_38853_(p_9915_.m_134561_()).m_7993_(), p_9915_.m_134564_()))) { // Insist on valid slot
+            // CraftBukkit start - Call click event
+            InventoryView inventory = this.f_9743_.f_36095_.getBukkitView();
+            org.bukkit.inventory.ItemStack item = CraftItemStack.asBukkitCopy(p_9915_.m_134564_());
+            InventoryType.SlotType type = InventoryType.SlotType.QUICKBAR;
+            if (flag) {
+               type = InventoryType.SlotType.OUTSIDE;
+            } else if (p_9915_.m_134561_() < 36) {
+               if (p_9915_.m_134561_() >= 5 && p_9915_.m_134561_() < 9) {
+                  type = InventoryType.SlotType.ARMOR;
+               } else {
+                  type = InventoryType.SlotType.CONTAINER;
+               }
+            }
+            InventoryCreativeEvent event = new InventoryCreativeEvent(inventory, type, flag ? -999 : p_9915_.m_134561_(), item);
+            cserver.getPluginManager().callEvent(event);
+            itemstack = CraftItemStack.asNMSCopy(event.getCursor());
+            switch (event.getResult()) {
+               case ALLOW:
+                  // Plugin cleared the id / stacksize checks
+                  flag2 = true;
+                  break;
+               case DEFAULT:
+                  break;
+               case DENY:
+                  // Reset the slot
+                  if (p_9915_.m_134561_() >= 0) {
+                     this.f_9743_.f_8906_.m_141995_(new ClientboundContainerSetSlotPacket(this.f_9743_.f_36095_.f_38840_, this.f_9743_.f_36095_.m_182425_(), p_9915_.m_134561_(), this.f_9743_.f_36095_.m_38853_(p_9915_.m_134561_()).m_7993_()));
+                     this.f_9743_.f_8906_.m_141995_(new ClientboundContainerSetSlotPacket(-1, this.f_9743_.f_36095_.m_182425_(), -1, ItemStack.f_41583_));
+                  }
+                  return;
+            }
+         }
+         // CraftBukkit end
          if (flag1 && flag2) {
             this.f_9743_.f_36095_.m_38853_(p_9915_.m_134561_()).m_5852_(itemstack);
             this.f_9743_.f_36095_.m_38946_();
@@ -1430,6 +_,8 @@
    }
 
    public void m_7423_(ServerboundCustomPayloadPacket p_9860_) {
+      PacketUtils.m_131359_(p_9860_, this, this.f_9743_.m_183503_());
+      net.minecraftforge.network.NetworkHooks.onCustomPayload(p_9860_, this.f_9742_);
    }
 
    public void m_7477_(ServerboundChangeDifficultyPacket p_9839_) {
@@ -1448,6 +_,20 @@
 
    public ServerPlayer m_142253_() {
       return this.f_9743_;
+   }
+
+   // Magma Stub
+    public void disconnect(String s) {
+    }
+
+   public void chat(String msg, boolean b) {
+   }
+
+   public void teleport(Location to) {
+   }
+
+   public final boolean isDisconnected() {
+      return !this.f_9743_.joining && !this.f_9742_.m_129536_();
    }
 
    @FunctionalInterface
