--- a/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerStatusPacketListenerImpl.java
@@ -1,14 +_,14 @@
 package net.minecraft.server.network;
 
 import net.minecraft.network.Connection;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.network.protocol.status.ClientboundPongResponsePacket;
 import net.minecraft.network.protocol.status.ClientboundStatusResponsePacket;
 import net.minecraft.network.protocol.status.ServerStatusPacketListener;
 import net.minecraft.network.protocol.status.ServerboundPingRequestPacket;
 import net.minecraft.network.protocol.status.ServerboundStatusRequestPacket;
 import net.minecraft.server.MinecraftServer;
 
 public class ServerStatusPacketListenerImpl implements ServerStatusPacketListener {
    private static final Component f_10081_ = new TranslatableComponent("multiplayer.status.request_handled");
@@ -33,7 +_,44 @@
          this.f_10083_.m_129507_(f_10081_);
       } else {
          this.f_10084_ = true;
-         this.f_10083_.m_129512_(new ClientboundStatusResponsePacket(this.f_10082_.m_129928_()));
+//         this.connection.send(new ClientboundStatusResponsePacket(this.server.getStatus()));
+         // CraftBukkit start
+         org.magmafoundation.magma.craftbukkit.ServerListPingEventImpl event = new org.magmafoundation.magma.craftbukkit.ServerListPingEventImpl(f_10083_, f_10082_);
+         this.f_10082_.server.getPluginManager().callEvent(event);
+
+         java.util.List<com.mojang.authlib.GameProfile> profiles = new java.util.ArrayList<com.mojang.authlib.GameProfile>(event.getPlayers().length);
+         for (Object player : event.getPlayers()) {
+            if (player != null) {
+               net.minecraft.server.level.ServerPlayer entityPlayer = ((net.minecraft.server.level.ServerPlayer) player);
+               if (entityPlayer.m_184128_()) {
+                  profiles.add(entityPlayer.m_36316_());
+               } else {
+                  profiles.add(MinecraftServer.f_195496_);
+               }
+            }
+         }
+
+         net.minecraft.network.protocol.status.ServerStatus.Players playerSample = new net.minecraft.network.protocol.status.ServerStatus.Players(event.getMaxPlayers(), profiles.size());
+         if (!this.f_10082_.m_183306_()) {
+            // Spigot Start
+            if ( !profiles.isEmpty() )
+            {
+               java.util.Collections.shuffle( profiles ); // This sucks, its inefficient but we have no simple way of doing it differently
+               profiles = profiles.subList( 0, Math.min( profiles.size(), org.spigotmc.SpigotConfig.playerSample ) ); // Cap the sample to n (or less) displayed players, ie: Vanilla behaviour
+            }
+            // Spigot End
+            playerSample.m_134924_(profiles.toArray(new com.mojang.authlib.GameProfile[profiles.size()]));
+         }
+
+         net.minecraft.network.protocol.status.ServerStatus ping = new net.minecraft.network.protocol.status.ServerStatus();
+         ping.m_134906_(event.icon.value);
+         ping.m_134908_(org.bukkit.craftbukkit.v1_18_R2.util.CraftChatMessage.fromString(event.getMotd(), true)[0]);
+         ping.m_134910_(playerSample);
+         int version = net.minecraft.SharedConstants.m_183709_().getProtocolVersion();
+         ping.m_134912_(new net.minecraft.network.protocol.status.ServerStatus.Version(f_10082_.m_130001_() + " " + f_10082_.m_7630_(), version));
+         ping.setForgeData(this.f_10082_.m_129928_().getForgeData());
+         this.f_10083_.m_129512_(new ClientboundStatusResponsePacket(ping));
+         // CraftBukkit end
       }
    }
 
