--- a/net/minecraft/network/CompressionEncoder.java
+++ b/net/minecraft/network/CompressionEncoder.java
@@ -3,12 +_,14 @@
 import io.netty.buffer.ByteBuf;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.MessageToByteEncoder;
 import java.util.zip.Deflater;
 
 public class CompressionEncoder extends MessageToByteEncoder<ByteBuf> {
    private final byte[] f_129444_ = new byte[8192];
    private final Deflater f_129445_;
    private int f_129446_;
+   private static final boolean DISABLE_PACKET_DEBUG = Boolean.parseBoolean(System.getProperty("forge.disablePacketCompressionDebug", "false"));
+   private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger();
 
    public CompressionEncoder(int p_129448_) {
       this.f_129446_ = p_129448_;
@@ -22,6 +_,13 @@
          friendlybytebuf.m_130130_(0);
          friendlybytebuf.writeBytes(p_129453_);
       } else {
+         int maxSize = org.magmafoundation.magma.configuration.MagmaConfig.instance.forgePacketCompressionThreshold.getValues();
+         if (!DISABLE_PACKET_DEBUG && i > 2097152 && i > maxSize) { //Magma - make packet size configurable
+             p_129453_.markReaderIndex();
+             LOGGER.error("Attempted to send packet over maximum protocol size: {} > {}", i, maxSize);
+             LOGGER.error("[PACKET DUMP DISABLED BY MAGMA]"); //Magma - you're welcome
+             p_129453_.resetReaderIndex();
+         }
          byte[] abyte = new byte[i];
          p_129453_.readBytes(abyte);
          friendlybytebuf.m_130130_(abyte.length);
