--- a/net/minecraft/network/CompressionEncoder.java
+++ b/net/minecraft/network/CompressionEncoder.java
@@ -9,6 +_,8 @@
    private final byte[] encodeBuf = new byte[8192];
    private final Deflater deflater;
    private int threshold;
+   private static final boolean DISABLE_PACKET_DEBUG = Boolean.parseBoolean(System.getProperty("forge.disablePacketCompressionDebug", "false"));
+   private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger();
 
    public CompressionEncoder(int p_129448_) {
       this.threshold = p_129448_;
@@ -22,6 +_,13 @@
          friendlybytebuf.writeVarInt(0);
          friendlybytebuf.writeBytes(p_129453_);
       } else {
+         int maxSize = org.magmafoundation.magma.configuration.MagmaConfig.instance.forgePacketCompressionThreshold.getValues();
+         if (!DISABLE_PACKET_DEBUG && i > 2097152 && i > maxSize) { //Magma - make packet size configurable
+             p_129453_.markReaderIndex();
+             LOGGER.error("Attempted to send packet over maximum protocol size: {} > {}", i, maxSize);
+             LOGGER.error("[PACKET DUMP DISABLED BY MAGMA]"); //Magma - you're welcome
+             p_129453_.resetReaderIndex();
+         }
          byte[] abyte = new byte[i];
          p_129453_.readBytes(abyte);
          friendlybytebuf.writeVarInt(abyte.length);
