--- a/net/minecraft/network/PacketEncoder.java
+++ b/net/minecraft/network/PacketEncoder.java
@@ -8,6 +_,7 @@
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketFlow;
 import net.minecraft.util.profiling.jfr.JvmProfiler;
+import org.magmafoundation.magma.configuration.MagmaConfig;
 import org.slf4j.Logger;
 
 public class PacketEncoder extends MessageToByteEncoder<Packet<?>> {
@@ -38,8 +_,8 @@
                int i = friendlybytebuf.writerIndex();
                p_130546_.m_5779_(friendlybytebuf);
                int j = friendlybytebuf.writerIndex() - i;
-               if (j > 8388608) {
-                  throw new IllegalArgumentException("Packet too big (is " + j + ", should be less than 8388608): " + p_130546_);
+               if (j > 8388608 && j > MagmaConfig.instance.forgeMaxPacketSize.getValues()) {
+                  throw new IllegalArgumentException("Packet too big (is " + j + ", should be less than "+MagmaConfig.instance.forgeMaxPacketSize.getValues()+"): " + p_130546_);
                } else {
                   int k = p_130545_.channel().attr(Connection.f_129461_).get().m_129582_();
                   JvmProfiler.f_185340_.m_183508_(k, integer, p_130545_.channel().remoteAddress(), j);
