--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -45,6 +_,19 @@
                ItemStack itemstack1 = p_155307_.getRecipeManager().getRecipeFor(RecipeType.CAMPFIRE_COOKING, container, p_155307_).map((p_155305_) -> {
                   return p_155305_.assemble(container);
                }).orElse(itemstack);
+
+               // CraftBukkit start - fire BlockCookEvent
+               org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack source = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asCraftMirror(itemstack);
+               org.bukkit.inventory.ItemStack result = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asBukkitCopy(itemstack1);
+               org.bukkit.event.block.BlockCookEvent blockCookEvent = new org.bukkit.event.block.BlockCookEvent(org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_155307_, p_155308_), source, result);
+               p_155307_.getCraftServer().getPluginManager().callEvent(blockCookEvent);
+               if (blockCookEvent.isCancelled()) {
+                  return;
+               }
+               result = blockCookEvent.getResult();
+               itemstack1 = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asNMSCopy(result);
+               // CraftBukkit end
+
                Containers.dropItemStack(p_155307_, (double)p_155308_.getX(), (double)p_155308_.getY(), (double)p_155308_.getZ(), itemstack1);
                p_155310_.items.set(i, ItemStack.EMPTY);
                p_155307_.sendBlockUpdated(p_155308_, p_155309_, p_155309_, 3);
