--- a/net/minecraft/world/level/block/entity/LecternBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/LecternBlockEntity.java
@@ -26,12 +_,14 @@
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
-public class LecternBlockEntity extends BlockEntity implements Clearable, MenuProvider {
+public class LecternBlockEntity extends BlockEntity implements Clearable, MenuProvider, CommandSource { // CraftBukkit - CommandSource
    public static final int DATA_PAGE = 0;
    public static final int NUM_DATA = 1;
    public static final int SLOT_BOOK = 0;
    public static final int NUM_SLOTS = 1;
-   public final Container bookAccess = new Container() {
+   public final Container bookAccess = new LecternInventory();
+
+   public class LecternInventory implements  Container {
       public int getContainerSize() {
          return 1;
       }
@@ -69,10 +_,18 @@
       }
 
       public void setItem(int p_59585_, ItemStack p_59586_) {
+         // CraftBukkit start
+         if (p_59585_ == 0) {
+            LecternBlockEntity.this.setBook(p_59586_);
+            if (LecternBlockEntity.this.getLevel() != null) {
+               LecternBlock.resetBookState(LecternBlockEntity.this.getLevel(), LecternBlockEntity.this.getBlockPos(), LecternBlockEntity.this.getBlockState(), LecternBlockEntity.this.hasBook());
+            }
+         }
+         // CraftBukkit end
       }
 
       public int getMaxStackSize() {
-         return 1;
+         return maxStack; // CraftBukkit
       }
 
       public void setChanged() {
@@ -91,9 +_,54 @@
          return false;
       }
 
+      // CraftBukkit start
+      public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
+      private int maxStack = 1;
+
+      @Override
+      public java.util.List<ItemStack> getContents() {
+         return java.util.Arrays.asList(book);
+      }
+
+      @Override
+      public void onOpen(org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity who) {
+         transaction.add(who);
+      }
+
+      @Override
+      public void onClose(org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity who) {
+         transaction.remove(who);
+      }
+
+      @Override
+      public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+         return transaction;
+      }
+
+      @Override
+      public org.bukkit.inventory.InventoryHolder getOwner() {
+         return LecternBlockEntity.this.getOwner();
+      }
+
+      @Override
+      public void setMaxStackSize(int size) {
+         maxStack = size;
+      }
+
+      @Override
+      public org.bukkit.Location getLocation() {
+         return new org.bukkit.Location(level.getWorld(), worldPosition.getX(), worldPosition.getY(), worldPosition.getZ());
+      }
+
+      public LecternBlockEntity getLectern() {
+         return LecternBlockEntity.this;
+      }
+      // CraftBukkit end
+
       public void clearContent() {
       }
-   };
+   }
+
    private final ContainerData dataAccess = new ContainerData() {
       public int get(int p_59600_) {
          return p_59600_ == 0 ? LecternBlockEntity.this.page : 0;
@@ -148,7 +_,7 @@
       if (i != this.page) {
          this.page = i;
          this.setChanged();
-         LecternBlock.signalPageChange(this.getLevel(), this.getBlockPos(), this.getBlockState());
+         if (this.level != null) LecternBlock.signalPageChange(this.getLevel(), this.getBlockPos(), this.getBlockState()); // CraftBukkit
       }
 
    }
@@ -170,6 +_,32 @@
       return p_59555_;
    }
 
+   // CraftBukkit start
+   @Override
+   public void sendMessage(Component ichatbasecomponent, java.util.UUID uuid) {
+   }
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return wrapper.getEntity() != null ? wrapper.getEntity().getBukkitSender(wrapper) : new org.bukkit.craftbukkit.v1_18_R2.command.CraftBlockCommandSender(wrapper, this);
+   }
+
+   @Override
+   public boolean acceptsSuccess() {
+      return false;
+   }
+
+   @Override
+   public boolean acceptsFailure() {
+      return false;
+   }
+
+   @Override
+   public boolean shouldInformAdmins() {
+      return false;
+   }
+   // CraftBukkit end
+
    private CommandSourceStack createCommandSourceStack(@Nullable Player p_59535_) {
       String s;
       Component component;
@@ -215,7 +_,7 @@
    }
 
    public AbstractContainerMenu createMenu(int p_59562_, Inventory p_59563_, Player p_59564_) {
-      return new LecternMenu(p_59562_, this.bookAccess, this.dataAccess);
+      return new LecternMenu(p_59562_, this.bookAccess, this.dataAccess, p_59563_); // CraftBukkit
    }
 
    public Component getDisplayName() {
