--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -23,7 +_,7 @@
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 
-public class SignBlockEntity extends BlockEntity {
+public class SignBlockEntity extends BlockEntity implements CommandSource { // CraftBukkit - implements
    public static final int LINES = 4;
    private static final String[] RAW_TEXT_FIELD_NAMES = new String[]{"Text1", "Text2", "Text3", "Text4"};
    private static final String[] FILTERED_TEXT_FIELD_NAMES = new String[]{"FilteredText1", "FilteredText2", "FilteredText3", "FilteredText4"};
@@ -55,6 +_,12 @@
          }
       }
 
+      // CraftBukkit start
+      if (Boolean.getBoolean("convertLegacySigns")) {
+         p_187515_.putBoolean("Bukkit.isConverted", true);
+      }
+      // CraftBukkit end
+
       p_187515_.putString("Color", this.color.getName());
       p_187515_.putBoolean("GlowingText", this.hasGlowingText);
    }
@@ -64,8 +_,24 @@
       super.load(p_155716_);
       this.color = DyeColor.byName(p_155716_.getString("Color"), DyeColor.BLACK);
 
+      // CraftBukkit start - Add an option to convert signs correctly
+      // This is done with a flag instead of all the time because
+      // we have no way to tell whether a sign is from 1.7.10 or 1.8
+      boolean oldSign = Boolean.getBoolean("convertLegacySigns") && !p_155716_.getBoolean("Bukkit.isConverted");
+      // CraftBukkit end
+
       for(int i = 0; i < 4; ++i) {
          String s = p_155716_.getString(RAW_TEXT_FIELD_NAMES[i]);
+         // CraftBukkit start
+         if (s != null && s.length() > 2048) {
+            s = "\"\"";
+         }
+
+         if (oldSign) {
+            messages[i] = org.bukkit.craftbukkit.v1_18_R2.util.CraftChatMessage.fromString(s)[0];
+            continue;
+         }
+         // CraftBukkit end
          Component component = this.loadLine(s);
          this.messages[i] = component;
          String s1 = FILTERED_TEXT_FIELD_NAMES[i];
@@ -98,6 +_,10 @@
          if (component != null) {
             return component;
          }
+         // CraftBukkit start
+      } catch (com.google.gson.JsonParseException jsonparseexception) {
+         return new TextComponent(p_155721_);
+         // CraftBukkit end
       } catch (Exception exception) {
       }
 
@@ -180,10 +_,36 @@
       return true;
    }
 
+   // CraftBukkit start
+   @Override
+   public void sendMessage(Component component, java.util.UUID uuid) {}
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return wrapper.getEntity() != null ? wrapper.getEntity().getBukkitSender(wrapper) : new org.bukkit.craftbukkit.v1_18_R2.command.CraftBlockCommandSender(wrapper, this);
+   }
+
+   @Override
+   public boolean acceptsSuccess() {
+      return false;
+   }
+
+   @Override
+   public boolean acceptsFailure() {
+      return false;
+   }
+
+   @Override
+   public boolean shouldInformAdmins() {
+      return false;
+   }
+   // CraftBukkit end
+
    public CommandSourceStack createCommandSourceStack(@Nullable ServerPlayer p_59736_) {
       String s = p_59736_ == null ? "Sign" : p_59736_.getName().getString();
       Component component = (Component)(p_59736_ == null ? new TextComponent("Sign") : p_59736_.getDisplayName());
-      return new CommandSourceStack(CommandSource.NULL, Vec3.atCenterOf(this.worldPosition), Vec2.ZERO, (ServerLevel)this.level, 2, s, component, this.level.getServer(), p_59736_);
+      // CraftBukkit - this
+      return new CommandSourceStack(this, Vec3.atCenterOf(this.worldPosition), Vec2.ZERO, (ServerLevel) this.level, 2, s, (Component) component, this.level.getServer(), p_59736_);
    }
 
    public DyeColor getColor() {
@@ -216,6 +_,6 @@
 
    private void markUpdated() {
       this.setChanged();
-      this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3);
+      if (this.level != null) this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3); // CraftBukkit - skip notify if world is null (SPIGOT-5122)
    }
 }
