--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -28,6 +_,6 @@
 import net.minecraft.world.level.levelgen.feature.configurations.EndGatewayConfiguration;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
 
 public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
@@ -180,6 +_,24 @@
             } else {
                entity = p_155832_.m_20201_();
             }
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (entity instanceof ServerPlayer) {
+               org.bukkit.craftbukkit.v1_18_R2.entity.CraftPlayer player = (org.bukkit.craftbukkit.v1_18_R2.entity.CraftPlayer) entity.getBukkitEntity();
+               org.bukkit.Location location = new org.bukkit.Location(p_155829_.getWorld(), (double) blockpos1.m_123341_() + 0.5D, (double) blockpos1.m_123342_() + 0.5D, (double) blockpos1.m_123343_() + 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+               org.bukkit.event.player.PlayerTeleportEvent teleEvent = new org.bukkit.event.player.PlayerTeleportEvent(player, player.getLocation(), location, org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               org.bukkit.Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+               entity.m_20091_();
+               ((ServerPlayer) entity).f_8906_.teleport(teleEvent.getTo());
+               m_155849_(p_155829_, p_155830_, p_155831_, p_155833_); // CraftBukkit - call at end of method
+               return;
+            }
+            // CraftBukkit end
 
             entity.m_20091_();
             entity.m_20324_((double)blockpos1.m_123341_() + 0.5D, (double)blockpos1.m_123342_(), (double)blockpos1.m_123343_() + 0.5D);
