--- a/net/minecraft/server/rcon/RconConsoleSource.java
+++ b/net/minecraft/server/rcon/RconConsoleSource.java
@@ -34,8 +_,19 @@
       return new CommandSourceStack(this, Vec3.atLowerCornerOf(serverlevel.getSharedSpawnPos()), Vec2.ZERO, serverlevel, 4, "Rcon", RCON_COMPONENT, this.server, (Entity)null);
    }
 
+   // CraftBukkit start - Send a String
+   public void sendMessage(String message) {
+      this.buffer.append(message);
+   }
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+      return server.remoteConsole;
+   }
+   // CraftBukkit end
+
    public void sendMessage(Component p_11509_, UUID p_11510_) {
-      this.buffer.append(p_11509_.getString());
+      this.buffer.append(p_11509_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean acceptsSuccess() {
