--- a/net/minecraft/world/entity/npc/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/AbstractVillager.java
@@ -43,7 +_,16 @@
    private Player tradingPlayer;
    @Nullable
    protected MerchantOffers offers;
-   private final SimpleContainer inventory = new SimpleContainer(8);
+   private final SimpleContainer inventory = new SimpleContainer(8, (org.bukkit.craftbukkit.v1_18_R2.entity.CraftAbstractVillager) this.getBukkitEntity()); // // CraftBukkit add argument
+
+   // CraftBukkit start
+   private org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchant craftMerchant;
+
+   @Override
+   public org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchant getCraftMerchant() {
+      return (craftMerchant == null) ? craftMerchant = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchant(this) : craftMerchant;
+   }
+   // CraftBukkit end
 
    public AbstractVillager(EntityType<? extends AbstractVillager> p_35267_, Level p_35268_) {
       super(p_35267_, p_35268_);
@@ -164,9 +_,9 @@
    }
 
    @Nullable
-   public Entity changeDimension(ServerLevel p_35295_) {
+   public Entity changeDimension(ServerLevel p_35295_, net.minecraftforge.common.util.ITeleporter teleporter) {
       this.stopTrading();
-      return super.changeDimension(p_35295_);
+      return super.changeDimension(p_35295_, teleporter);
    }
 
    protected void stopTrading() {
@@ -219,7 +_,16 @@
          VillagerTrades.ItemListing villagertrades$itemlisting = p_35279_[integer];
          MerchantOffer merchantoffer = villagertrades$itemlisting.getOffer(this, this.random);
          if (merchantoffer != null) {
-            p_35278_.add(merchantoffer);
+            // CraftBukkit start
+            org.bukkit.event.entity.VillagerAcquireTradeEvent event = new org.bukkit.event.entity.VillagerAcquireTradeEvent((org.bukkit.entity.AbstractVillager) getBukkitEntity(), merchantoffer.asBukkit());
+            // Suppress during worldgen
+            if (this.valid) {
+               org.bukkit.Bukkit.getPluginManager().callEvent(event);
+            }
+            if (!event.isCancelled()) {
+               p_35278_.add(org.bukkit.craftbukkit.v1_18_R2.inventory.CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+            }
+            // CraftBukkit end
          }
       }
 
