--- a/net/minecraft/network/chat/Style.java
+++ b/net/minecraft/network/chat/Style.java
@@ -133,6 +_,30 @@
       return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, p_178525_, this.clickEvent, this.hoverEvent, this.insertion, this.font);
    }
 
+   /**
+    * @deprecated Use {@link #withUnderlined(Boolean)} instead.
+    */
+   @Deprecated(forRemoval = true, since = "1.18.2")
+   public Style setUnderlined(@Nullable Boolean underlined) {
+      return new Style(this.color, this.bold, this.italic, underlined, this.strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
+   /**
+    * @deprecated Use {@link #withStrikethrough(Boolean)} instead.
+    */
+   @Deprecated(forRemoval = true, since = "1.18.2")
+   public Style setStrikethrough(@Nullable Boolean strikethrough) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, strikethrough, this.obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
+   /**
+    * @deprecated Use {@link #withObfuscated(Boolean)} instead.
+    */
+   @Deprecated(forRemoval = true, since = "1.18.2")
+   public Style setObfuscated(@Nullable Boolean obfuscated) {
+      return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, obfuscated, this.clickEvent, this.hoverEvent, this.insertion, this.font);
+   }
+
    public Style withClickEvent(@Nullable ClickEvent p_131143_) {
       return new Style(this.color, this.bold, this.italic, this.underlined, this.strikethrough, this.obfuscated, p_131143_, this.hoverEvent, this.insertion, this.font);
    }
