--- a/net/minecraft/network/chat/TextColor.java
+++ b/net/minecraft/network/chat/TextColor.java
@@ -11,7 +_,7 @@
 public final class TextColor {
    private static final String CUSTOM_COLOR_PREFIX = "#";
    private static final Map<ChatFormatting, TextColor> LEGACY_FORMAT_TO_COLOR = Stream.of(ChatFormatting.values()).filter(ChatFormatting::isColor).collect(ImmutableMap.toImmutableMap(Function.identity(), (p_131276_) -> {
-      return new TextColor(p_131276_.getColor(), p_131276_.getName());
+      return new TextColor(p_131276_.getColor(), p_131276_.getName(), p_131276_); // CraftBukkit
    }));
    private static final Map<String, TextColor> NAMED_COLORS = LEGACY_FORMAT_TO_COLOR.values().stream().collect(ImmutableMap.toImmutableMap((p_131273_) -> {
       return p_131273_.name;
@@ -20,15 +_,28 @@
    @Nullable
    private final String name;
 
+   // CraftBukkit start
+   @Nullable
+   public final ChatFormatting format;
+
+   private TextColor(int p_131263_, String p_131264_, ChatFormatting format) {
+      this.value = p_131263_;
+      this.name = p_131264_;
+      this.format = format;
+   }
+
    private TextColor(int p_131263_, String p_131264_) {
       this.value = p_131263_;
       this.name = p_131264_;
+      this.format = null;
    }
 
    private TextColor(int p_131261_) {
       this.value = p_131261_;
       this.name = null;
+      this.format = null;
    }
+   // CraftBukkit end
 
    public int getValue() {
       return this.value;
