--- a/com/mojang/realmsclient/gui/screens/RealmsGenericErrorScreen.java
+++ b/com/mojang/realmsclient/gui/screens/RealmsGenericErrorScreen.java
@@ -67,6 +_,15 @@
       return (new TextComponent("")).append(this.lines.getFirst()).append(": ").append(this.lines.getSecond());
    }
 
+    @Override
+    public boolean keyPressed(int key, int scanCode, int modifiers) {
+       if (key == org.lwjgl.glfw.GLFW.GLFW_KEY_ESCAPE) {
+          minecraft.setScreen(this.nextScreen);
+          return true;
+       }
+       return super.keyPressed(key, scanCode, modifiers);
+    }
+
    public void render(PoseStack p_88679_, int p_88680_, int p_88681_, float p_88682_) {
       this.renderBackground(p_88679_);
       drawCenteredString(p_88679_, this.font, this.lines.getFirst(), this.width / 2, 80, 16777215);
