--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -100,6 +_,7 @@
 
          this.dragonKilled = p_64080_.getBoolean("DragonKilled");
          this.previouslyKilled = p_64080_.getBoolean("PreviouslyKilled");
+         this.needsStateScanning = !p_64080_.getBoolean("LegacyScanPerformed"); // Forge: fix MC-105080
          if (p_64080_.getBoolean("IsRespawning")) {
             this.respawnStage = DragonRespawnAnimation.START;
          }
@@ -131,13 +_,14 @@
       compoundtag.putBoolean("NeedsStateScanning", this.needsStateScanning);
       if (this.dragonUUID != null) {
          compoundtag.putUUID("Dragon", this.dragonUUID);
-      }
 
       compoundtag.putBoolean("DragonKilled", this.dragonKilled);
       compoundtag.putBoolean("PreviouslyKilled", this.previouslyKilled);
+      compoundtag.putBoolean("LegacyScanPerformed", !this.needsStateScanning); // Forge: fix MC-105080
       if (this.portalLocation != null) {
          compoundtag.put("ExitPortalLocation", NbtUtils.writeBlockPos(this.portalLocation));
       }
+      }
 
       ListTag listtag = new ListTag();
 
@@ -507,6 +_,13 @@
             endcrystal.setBeamTarget((BlockPos)null);
          }
       }
-
+   }
+
+   public void addPlayer(ServerPlayer player) {
+      this.dragonEvent.addPlayer(player);
+   }
+
+   public void removePlayer(ServerPlayer player) {
+      this.dragonEvent.removePlayer(player);
    }
 }
