--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
@@ -18,7 +_,8 @@
    }
 
    public boolean test(LootContext p_81930_) {
-      return p_81930_.getRandom().nextFloat() < this.probability;
+      int i = p_81930_.getLootingModifier();
+      return p_81930_.getRandom().nextFloat() < this.probability + (float) i;
    }
 
    public static LootItemCondition.Builder randomChance(float p_81928_) {
