--- a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -51,7 +_,11 @@
       if (this.level instanceof ServerLevel) {
          BlockPos blockpos = this.blockPosition();
          if (((ServerLevel)this.level).dragonFight() != null && this.level.getBlockState(blockpos).isAir()) {
-            this.level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level, blockpos));
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockIgniteEvent(this.level, blockpos, this).isCancelled()) {
+               this.level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(this.level, blockpos));
+            }
+            // CraftBukkit end
          }
       }
 
@@ -87,9 +_,23 @@
          return false;
       } else {
          if (!this.isRemoved() && !this.level.isClientSide) {
+            // CraftBukkit start - All non-living entities need this
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31050_, p_31051_, false)) {
+               return false;
+            }
+            // CraftBukkit end
             this.remove(Entity.RemovalReason.KILLED);
             if (!p_31050_.isExplosion()) {
-               this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), 6.0F, Explosion.BlockInteraction.DESTROY);
+               // CraftBukkit start
+               org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, false);
+               this.level.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  this.unsetRemoved();
+                  return false;
+               }
+               this.level.prepareExplosion(event.getFire()); //Magma - fix mixin inject error
+               this.level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), Explosion.BlockInteraction.DESTROY);
+               // CraftBukkit end
             }
 
             this.onDestroyedBy(p_31050_);
