--- a/net/minecraft/client/player/LocalPlayer.java
+++ b/net/minecraft/client/player/LocalPlayer.java
@@ -145,6 +_,7 @@
    }
 
    public boolean hurt(DamageSource p_108662_, float p_108663_) {
+      net.minecraftforge.common.ForgeHooks.onPlayerAttack(this, p_108662_, p_108663_);
       return false;
    }
 
@@ -461,6 +_,11 @@
    }
 
    public void playSound(SoundEvent p_108651_, float p_108652_, float p_108653_) {
+      net.minecraftforge.event.entity.PlaySoundAtEntityEvent event = net.minecraftforge.event.ForgeEventFactory.onPlaySoundAtEntity(this, p_108651_, this.getSoundSource(), p_108652_, p_108653_);
+      if (event.isCanceled() || event.getSound() == null) return;
+      p_108651_ = event.getSound();
+      p_108652_ = event.getVolume();
+      p_108653_ = event.getPitch();
       this.level.playLocalSound(this.getX(), this.getY(), this.getZ(), p_108651_, this.getSoundSource(), p_108652_, p_108653_, false);
    }
 
@@ -616,6 +_,7 @@
       boolean flag2 = this.hasEnoughImpulseToStartSprinting();
       this.crouching = !this.getAbilities().flying && !this.isSwimming() && this.canEnterPose(Pose.CROUCHING) && (this.isShiftKeyDown() || !this.isSleeping() && !this.canEnterPose(Pose.STANDING));
       this.input.tick(this.isMovingSlowly());
+      net.minecraftforge.client.ForgeHooksClient.onMovementInputUpdate(this, this.input);
       this.minecraft.getTutorial().onInput(this.input);
       if (this.isUsingItem() && !this.isPassenger()) {
          this.input.leftImpulse *= 0.2F;
@@ -688,7 +_,7 @@
 
       if (this.input.jumping && !flag7 && !flag && !this.getAbilities().flying && !this.isPassenger() && !this.onClimbable()) {
          ItemStack itemstack = this.getItemBySlot(EquipmentSlot.CHEST);
-         if (itemstack.is(Items.ELYTRA) && ElytraItem.isFlyEnabled(itemstack) && this.tryToStartFallFlying()) {
+         if (itemstack.canElytraFly(this) && this.tryToStartFallFlying()) {
             this.connection.send(new ServerboundPlayerCommandPacket(this, ServerboundPlayerCommandPacket.Action.START_FALL_FLYING));
          }
       }
@@ -806,6 +_,7 @@
 
    public void rideTick() {
       super.rideTick();
+      if (this.wantsToStopRiding() && this.isPassenger()) this.input.shiftKeyDown = false;
       this.handsBusy = false;
       if (this.getVehicle() instanceof Boat) {
          Boat boat = (Boat)this.getVehicle();
@@ -1019,6 +_,18 @@
       } else {
          return super.getRopeHoldPosition(p_108758_);
       }
+   }
+
+   public void updateSyncFields(LocalPlayer old) {
+      this.xLast = old.xLast;
+      this.yLast1 = old.yLast1;
+      this.zLast = old.zLast;
+      this.yRotLast = old.yRotLast;
+      this.xRotLast = old.xRotLast;
+      this.lastOnGround = old.lastOnGround;
+      this.wasShiftKeyDown = old.wasShiftKeyDown;
+      this.wasSprinting = old.wasSprinting;
+      this.positionReminder = old.positionReminder;
    }
 
    public void updateTutorialInventoryAction(ItemStack p_172532_, ItemStack p_172533_, ClickAction p_172534_) {
