--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -91,6 +_,28 @@
       }
    }
 
+   // CraftBukkit start
+   public EntitySelector parse(StringReader p_91451_, boolean overridePermissions) throws CommandSyntaxException {
+      int i = 0;
+      EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_91451_);
+      EntitySelector entityselector = entityselectorparser.parse(overridePermissions);
+      // CraftBukkit end
+      if (entityselector.getMaxResults() > 1 && this.single) {
+         if (this.playersOnly) {
+            p_91451_.setCursor(0);
+            throw ERROR_NOT_SINGLE_PLAYER.createWithContext(p_91451_);
+         } else {
+            p_91451_.setCursor(0);
+            throw ERROR_NOT_SINGLE_ENTITY.createWithContext(p_91451_);
+         }
+      } else if (entityselector.includesEntities() && this.playersOnly && !entityselector.isSelfSelector()) {
+         p_91451_.setCursor(0);
+         throw ERROR_ONLY_PLAYERS_ALLOWED.createWithContext(p_91451_);
+      } else {
+         return entityselector;
+      }
+   }
+
    public EntitySelector parse(StringReader p_91451_) throws CommandSyntaxException {
       int i = 0;
       EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_91451_);
