--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -155,8 +_,17 @@
    });
 
    public ModelBakery(ResourceManager p_119247_, BlockColors p_119248_, ProfilerFiller p_119249_, int p_119250_) {
+      this(p_119247_, p_119248_, true);
+      processLoading(p_119249_, p_119250_);
+   }
+
+   protected ModelBakery(ResourceManager p_119247_, BlockColors p_119248_, boolean vanillaBakery) {
       this.resourceManager = p_119247_;
       this.blockColors = p_119248_;
+   }
+
+   protected void processLoading(ProfilerFiller p_119249_, int p_119250_) {
+      net.minecraftforge.client.model.ModelLoaderRegistry.onModelLoadingStart();
       p_119249_.push("missing_model");
 
       try {
@@ -190,12 +_,16 @@
       p_119249_.popPush("special");
       this.loadTopLevel(new ModelResourceLocation("minecraft:trident_in_hand#inventory"));
       this.loadTopLevel(new ModelResourceLocation("minecraft:spyglass_in_hand#inventory"));
+      for (ResourceLocation rl : getSpecialModels()) {
+         addModelToCache(rl);
+      }
       p_119249_.popPush("textures");
       Set<Pair<String, String>> set = Sets.newLinkedHashSet();
       Set<Material> set1 = this.topLevelModels.values().stream().flatMap((p_119340_) -> {
          return p_119340_.getMaterials(this::getModel, set).stream();
       }).collect(Collectors.toSet());
       set1.addAll(UNREFERENCED_TEXTURES);
+      net.minecraftforge.client.ForgeHooksClient.gatherFluidTextures(set1);
       set.stream().filter((p_119357_) -> {
          return !p_119357_.getSecond().equals(MISSING_MODEL_LOCATION_STRING);
       }).forEach((p_119292_) -> {
@@ -234,6 +_,7 @@
          try {
             bakedmodel = this.bake(p_119369_, BlockModelRotation.X0_Y0);
          } catch (Exception exception) {
+            exception.printStackTrace();
             LOGGER.warn("Unable to bake model: '{}': {}", p_119369_, exception);
          }
 
@@ -473,6 +_,13 @@
       this.loadingStack.addAll(p_119354_.getDependencies());
    }
 
+   // Same as loadTopLevel but without restricting to MRL's
+   private void addModelToCache(ResourceLocation p_217843_1_) {
+      UnbakedModel unbakedmodel = this.getModel(p_217843_1_);
+      this.unbakedCache.put(p_217843_1_, unbakedmodel);
+      this.topLevelModels.put(p_217843_1_, unbakedmodel);
+   }
+
    private void loadTopLevel(ModelResourceLocation p_119307_) {
       UnbakedModel unbakedmodel = this.getModel(p_119307_);
       this.unbakedCache.put(p_119307_, unbakedmodel);
@@ -487,7 +_,13 @@
    }
 
    @Nullable
+   @Deprecated
    public BakedModel bake(ResourceLocation p_119350_, ModelState p_119351_) {
+      return bake(p_119350_, p_119351_, this.atlasSet::getSprite);
+   }
+
+   @Nullable
+   public BakedModel bake(ResourceLocation p_119350_, ModelState p_119351_, java.util.function.Function<Material, net.minecraft.client.renderer.texture.TextureAtlasSprite> sprites) {
       Triple<ResourceLocation, Transformation, Boolean> triple = Triple.of(p_119350_, p_119351_.getRotation(), p_119351_.isUvLocked());
       if (this.bakedCache.containsKey(triple)) {
          return this.bakedCache.get(triple);
@@ -498,11 +_,11 @@
          if (unbakedmodel instanceof BlockModel) {
             BlockModel blockmodel = (BlockModel)unbakedmodel;
             if (blockmodel.getRootModel() == GENERATION_MARKER) {
-               return ITEM_MODEL_GENERATOR.generateBlockModel(this.atlasSet::getSprite, blockmodel).bake(this, blockmodel, this.atlasSet::getSprite, p_119351_, p_119350_, false);
+               return ITEM_MODEL_GENERATOR.generateBlockModel(sprites, blockmodel).bake(this, blockmodel, this.atlasSet::getSprite, p_119351_, p_119350_, false);
             }
          }
 
-         BakedModel bakedmodel = unbakedmodel.bake(this, this.atlasSet::getSprite, p_119351_, p_119350_);
+         BakedModel bakedmodel = unbakedmodel.bake(this, sprites, p_119351_, p_119350_);
          this.bakedCache.put(triple, bakedmodel);
          return bakedmodel;
       }
@@ -555,11 +_,19 @@
       return this.modelGroups;
    }
 
+   public Set<ResourceLocation> getSpecialModels() {
+      return java.util.Collections.emptySet();
+   }
+
    @OnlyIn(Dist.CLIENT)
    static class BlockStateDefinitionException extends RuntimeException {
       public BlockStateDefinitionException(String p_119373_) {
          super(p_119373_);
       }
+   }
+
+   public AtlasSet getSpriteMap() {
+      return this.atlasSet;
    }
 
    @OnlyIn(Dist.CLIENT)
