--- a/net/minecraft/client/resources/model/SimpleBakedModel.java
+++ b/net/minecraft/client/resources/model/SimpleBakedModel.java
@@ -81,6 +_,10 @@
       private final boolean isGui3d;
       private final ItemTransforms transforms;
 
+      public Builder(net.minecraftforge.client.model.IModelConfiguration model, ItemOverrides overrides) {
+         this(model.useSmoothLighting(), model.isSideLit(), model.isShadedInGui(), model.getCameraTransforms(), overrides);
+      }
+
       public Builder(BlockModel p_119517_, ItemOverrides p_119518_, boolean p_119519_) {
          this(p_119517_.hasAmbientOcclusion(), p_119517_.getGuiLight().lightLikeBlock(), p_119519_, p_119517_.getTransforms(), p_119518_);
       }
