--- a/net/minecraft/world/entity/monster/ElderGuardian.java
+++ b/net/minecraft/world/entity/monster/ElderGuardian.java
@@ -66,7 +_,8 @@
          for(ServerPlayer serverplayer : list) {
             if (!serverplayer.hasEffect(mobeffect) || serverplayer.getEffect(mobeffect).getAmplifier() < 2 || serverplayer.getEffect(mobeffect).getDuration() < 1200) {
                serverplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.GUARDIAN_ELDER_EFFECT, this.isSilent() ? 0.0F : 1.0F));
-               serverplayer.addEffect(new MobEffectInstance(mobeffect, 6000, 2), this);
+               serverplayer.prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Magma
+               serverplayer.addEffect(new MobEffectInstance(mobeffect, 6000, 2), this); // CraftBukkit
             }
          }
       }
