--- a/net/minecraft/world/entity/monster/Husk.java
+++ b/net/minecraft/world/entity/monster/Husk.java
@@ -15,6 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
 
 public class Husk extends Zombie {
    public Husk(EntityType<? extends Husk> p_32889_, Level p_32890_) {
@@ -49,7 +_,8 @@
       boolean flag = super.m_7327_(p_32892_);
       if (flag && this.m_21205_().m_41619_() && p_32892_ instanceof LivingEntity) {
          float f = this.f_19853_.m_6436_(this.m_142538_()).m_19056_();
-         ((LivingEntity)p_32892_).m_147207_(new MobEffectInstance(MobEffects.f_19612_, 140 * (int)f), this);
+         ((LivingEntity)p_32892_).prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Magma
+         ((LivingEntity)p_32892_).m_147207_(new MobEffectInstance(MobEffects.f_19612_, 140 * (int)f), this); // CraftBukkit
       }
 
       return flag;
