--- a/net/minecraft/world/entity/monster/MagmaCube.java
+++ b/net/minecraft/world/entity/monster/MagmaCube.java
@@ -52,7 +_,7 @@
       return ParticleTypes.f_123744_;
    }
 
-   protected ResourceLocation m_7582_() {
+   public ResourceLocation m_7582_() {
       return this.m_33633_() ? BuiltInLootTables.f_78712_ : this.m_6095_().m_20677_();
    }
 
@@ -72,6 +_,7 @@
       Vec3 vec3 = this.m_20184_();
       this.m_20334_(vec3.f_82479_, (double)(this.m_6118_() + (float)this.m_33632_() * 0.1F), vec3.f_82481_);
       this.f_19812_ = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    protected void m_203347_(TagKey<Fluid> p_204065_) {
