--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -119,7 +_,12 @@
    }
 
    public boolean canBeAffected(MobEffectInstance p_33809_) {
-      return p_33809_.getEffect() == MobEffects.POISON ? false : super.canBeAffected(p_33809_);
+      if (p_33809_.getEffect() == MobEffects.POISON) {
+         net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, p_33809_);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
+      }
+      return super.canBeAffected(p_33809_);
    }
 
    public boolean isClimbing() {
@@ -157,7 +_,7 @@
       if (p_33793_ instanceof Spider.SpiderEffectsGroupData) {
          MobEffect mobeffect = ((Spider.SpiderEffectsGroupData)p_33793_).effect;
          if (mobeffect != null) {
-            this.addEffect(new MobEffectInstance(mobeffect, Integer.MAX_VALUE));
+            this.addEffect(new MobEffectInstance(mobeffect, Integer.MAX_VALUE), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SPIDER_SPAWN); // CraftBukkit
          }
       }
 
