--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -277,7 +_,13 @@
       BlockState blockstate = this.level.getBlockState(this.blockPosition());
       BlockState blockstate1 = this.getBlockStateOn();
       boolean flag = blockstate.is(BlockTags.STRIDER_WARM_BLOCKS) || blockstate1.is(BlockTags.STRIDER_WARM_BLOCKS) || this.getFluidHeight(FluidTags.LAVA) > 0.0D;
-      this.setSuffocating(!flag);
+      // CraftBukkit start
+      if (!flag ^ this.isSuffocating()) {
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, !flag)) {
+            this.setSuffocating(!flag);
+         }
+      }
+      // CraftBukkit end
       super.tick();
       this.floatStrider();
       this.checkInsideBlocks();
