--- a/net/minecraft/world/entity/monster/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/ZombifiedPiglin.java
@@ -136,7 +_,7 @@
       }).filter((p_34456_) -> {
          return !p_34456_.isAlliedTo(this.getTarget());
       }).forEach((p_34440_) -> {
-         p_34440_.setTarget(this.getTarget());
+         p_34440_.setTarget(this.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
       });
    }
 
@@ -144,7 +_,7 @@
       this.playSound(SoundEvents.ZOMBIFIED_PIGLIN_ANGRY, this.getSoundVolume() * 2.0F, this.getVoicePitch() * 1.8F);
    }
 
-   public void setTarget(@Nullable LivingEntity p_34478_) {
+   public boolean setTarget(@Nullable LivingEntity p_34478_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent) { // CraftBukkit - signature
       if (this.getTarget() == null && p_34478_ != null) {
          this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.sample(this.random);
          this.ticksUntilNextAlert = ALERT_INTERVAL.sample(this.random);
@@ -154,11 +_,20 @@
          this.setLastHurtByPlayer((Player)p_34478_);
       }
 
-      super.setTarget(p_34478_);
+      return super.setTarget(p_34478_, reason, fireEvent); // CraftBukkit
    }
 
    public void startPersistentAngerTimer() {
-      this.setRemainingPersistentAngerTime(PERSISTENT_ANGER_TIME.sample(this.random));
+      // CraftBukkit start
+      net.minecraft.world.entity.Entity entity = ((ServerLevel) this.level).getEntity(getPersistentAngerTarget());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), ZombifiedPiglin.PERSISTENT_ANGER_TIME.sample(this.random));
+      this.level.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.setPersistentAngerTarget(null);
+         return;
+      }
+      this.setRemainingPersistentAngerTime(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean checkZombifiedPiglinSpawnRules(EntityType<ZombifiedPiglin> p_34450_, LevelAccessor p_34451_, MobSpawnType p_34452_, BlockPos p_34453_, Random p_34454_) {
