--- a/net/minecraft/network/protocol/status/ClientboundStatusResponsePacket.java
+++ b/net/minecraft/network/protocol/status/ClientboundStatusResponsePacket.java
@@ -22,7 +_,7 @@
    }
 
    public void write(FriendlyByteBuf p_134899_) {
-      p_134899_.writeUtf(GSON.toJson(this.status));
+      p_134899_.writeUtf(this.status.getJson()); //Forge: Let the response cache the JSON
    }
 
    public void handle(ClientStatusPacketListener p_134896_) {
