--- a/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -119,6 +_,7 @@
       int k1 = (int)((double)this.minecraft.getWindow().getGuiScaledHeight() * 0.8325D);
       float f6 = this.reload.getActualProgress();
       this.currentProgress = Mth.clamp(this.currentProgress * 0.95F + f6 * 0.050000012F, 0.0F, 1.0F);
+      net.minecraftforge.client.loading.ClientModLoader.renderProgressText();
       if (f < 1.0F) {
          this.drawProgressBar(p_96178_, i / 2 - j1, k1 - 5, i / 2 + j1, k1 + 5, 1.0F - Mth.clamp(f, 0.0F, 1.0F));
       }
@@ -128,6 +_,7 @@
       }
 
       if (this.fadeOutStart == -1L && this.reload.isDone() && (!this.fadeIn || f1 >= 2.0F)) {
+         this.fadeOutStart = Util.getMillis(); // Moved up to guard against inf loops caused by callback
          try {
             this.reload.checkExceptions();
             this.onFinish.accept(Optional.empty());
@@ -135,7 +_,6 @@
             this.onFinish.accept(Optional.of(throwable));
          }
 
-         this.fadeOutStart = Util.getMillis();
          if (this.minecraft.screen != null) {
             this.minecraft.screen.init(this.minecraft, this.minecraft.getWindow().getGuiScaledWidth(), this.minecraft.getWindow().getGuiScaledHeight());
          }
