--- a/net/minecraft/client/gui/screens/OptionsScreen.java
+++ b/net/minecraft/client/gui/screens/OptionsScreen.java
@@ -141,4 +_,12 @@
       drawCenteredString(p_96249_, this.font, this.title, this.width / 2, 15, 16777215);
       super.render(p_96249_, p_96250_, p_96251_, p_96252_);
    }
+
+    @Override
+    public void onClose() {
+        // We need to consider 2 potential parent screens here:
+        // 1. From the main menu, in which case display the main menu
+        // 2. From the pause menu, in which case exit back to game
+        this.minecraft.setScreen(this.lastScreen instanceof PauseScreen ? null : this.lastScreen);
+    }
 }
