--- a/net/minecraft/world/level/levelgen/DebugLevelSource.java
+++ b/net/minecraft/world/level/levelgen/DebugLevelSource.java
@@ -117,6 +_,12 @@
 
       return blockstate;
    }
+   
+   public static void initValidStates() {
+      ALL_BLOCKS = StreamSupport.stream(Registry.BLOCK.spliterator(), false).flatMap(block -> block.getStateDefinition().getPossibleStates().stream()).collect(Collectors.toList());
+      GRID_WIDTH = Mth.ceil(Mth.sqrt(ALL_BLOCKS.size()));
+      GRID_HEIGHT = Mth.ceil((float) (ALL_BLOCKS.size() / GRID_WIDTH));
+   }
 
    public Climate.Sampler climateSampler() {
       return Climate.empty();
