--- a/net/minecraft/world/level/levelgen/PatrolSpawner.java
+++ b/net/minecraft/world/level/levelgen/PatrolSpawner.java
@@ -104,7 +_,9 @@
             }
 
             patrollingmonster.setPos((double)p_64566_.getX(), (double)p_64566_.getY(), (double)p_64566_.getZ());
+            if(net.minecraftforge.common.ForgeHooks.canEntitySpawn(patrollingmonster, p_64565_, p_64566_.getX(), p_64566_.getY(), p_64566_.getZ(), null, MobSpawnType.PATROL) == -1) return false;
             patrollingmonster.finalizeSpawn(p_64565_, p_64565_.getCurrentDifficultyAt(p_64566_), MobSpawnType.PATROL, (SpawnGroupData)null, (CompoundTag)null);
+            p_64565_.prepareEntity(patrollingmonster, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PATROL); //Magma - add spawn reason
             p_64565_.addFreshEntityWithPassengers(patrollingmonster);
             return true;
          } else {
