--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -15,19 +_,60 @@
    private int f_38699_;
    private int f_38700_ = 20;
 
+   // CraftBukkit start
+   private @org.jetbrains.annotations.Nullable Player entityhuman; //Magma - add @Nullable
+   public boolean saturatedRegenRateModified = false; //Magma - mixin inject
+   public boolean unsaturatedRegenRateModified = false; //Magma - mixin inject
+   public boolean starvationRateModified = false; //Magma - mixin inject
+   public int saturatedRegenRate = 10;
+   public int unsaturatedRegenRate = 80;
+   public int starvationRate = 80;
+   // CraftBukkit end
+
    public FoodData() {
-      this.f_38697_ = 5.0F;
-   }
+      this(null); //Magma - instead of throwing an error we just set it to null
+   }
+
+   // CraftBukkit start - added Player constructor
+   public FoodData(Player player) {
+       this.entityhuman = player;
+       this.f_38697_ = 5.0F;
+   }
+   // CraftBukkit end
 
    public void m_38707_(int p_38708_, float p_38709_) {
       this.f_38696_ = Math.min(p_38708_ + this.f_38696_, 20);
       this.f_38697_ = Math.min(this.f_38697_ + (float)p_38708_ * p_38709_ * 2.0F, (float)this.f_38696_);
    }
 
+   // Use the LivingEntity sensitive version in favour of this.
+   @Deprecated
    public void m_38712_(Item p_38713_, ItemStack p_38714_) {
+      this.eat(p_38713_, p_38714_, null);
+   }
+
+   public void eat(Item p_38713_, ItemStack p_38714_, @javax.annotation.Nullable net.minecraft.world.entity.LivingEntity entity) {
       if (p_38713_.m_41472_()) {
-         FoodProperties foodproperties = p_38713_.m_41473_();
-         this.m_38707_(foodproperties.m_38744_(), foodproperties.m_38745_());
+         FoodProperties foodproperties = p_38714_.getFoodProperties(entity);
+
+         //Magma start - null check
+         if (entityhuman == null) {
+            this.m_38707_(foodproperties.m_38744_(), foodproperties.m_38745_());
+            return;
+         }
+         //Magma end
+
+         // CraftBukkit start
+         int oldFoodLevel = f_38696_;
+         org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callFoodLevelChangeEvent(entityhuman, foodproperties.m_38744_() + oldFoodLevel, p_38714_);
+
+         if (!event.isCancelled()) {
+            this.m_38707_(event.getFoodLevel() - oldFoodLevel, foodproperties.m_38745_());
+         }
+
+         ((net.minecraft.server.level.ServerPlayer) entityhuman).getBukkitEntity().sendHealthUpdate();
+         // CraftBukkit end
+
       }
 
    }
@@ -41,28 +_,48 @@
             this.f_38697_ = Math.max(this.f_38697_ - 1.0F, 0.0F);
          } else if (difficulty != Difficulty.PEACEFUL) {
             this.f_38696_ = Math.max(this.f_38696_ - 1, 0);
+
+            //Magma start - null check
+            if (entityhuman != null) {
+
+               // CraftBukkit start
+               org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callFoodLevelChangeEvent(entityhuman, Math.max(this.f_38696_ - 1, 0));
+
+               if (!event.isCancelled()) {
+                  this.f_38696_ = event.getFoodLevel();
+               }
+
+               ((net.minecraft.server.level.ServerPlayer) entityhuman).f_8906_.m_141995_(new net.minecraft.network.protocol.game.ClientboundSetHealthPacket(((net.minecraft.server.level.ServerPlayer) entityhuman).getBukkitEntity().getScaledHealth(), this.f_38696_, this.f_38697_));
+               // CraftBukkit end
+            }
+            //Magma end
+
          }
       }
 
       boolean flag = p_38711_.f_19853_.m_46469_().m_46207_(GameRules.f_46139_);
       if (flag && this.f_38697_ > 0.0F && p_38711_.m_36325_() && this.f_38696_ >= 20) {
          ++this.f_38699_;
-         if (this.f_38699_ >= 10) {
+         if (this.f_38699_ >= (saturatedRegenRateModified ? saturatedRegenRate : 10)) { // CraftBukkit - add saturatedRegenRate //Magma - mixin inject
             float f = Math.min(this.f_38697_, 6.0F);
-            p_38711_.m_5634_(f / 6.0F);
+            p_38711_.setRegainReason(org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.SATIATED); // Magma - fix mixin inject
+            p_38711_.m_5634_(f / 6.0F); // CraftBukkit - add RegainReason
+            this.prepareExhaustion(p_38711_, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.REGEN); //Magma
             this.m_38703_(f);
             this.f_38699_ = 0;
          }
       } else if (flag && this.f_38696_ >= 18 && p_38711_.m_36325_()) {
          ++this.f_38699_;
-         if (this.f_38699_ >= 80) {
+         if (this.f_38699_ >= (unsaturatedRegenRateModified ? this.unsaturatedRegenRate : 80)) { // CraftBukkit - add unsaturatedRegenRate //Magma - mixin inject
+            p_38711_.setRegainReason(org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.SATIATED); // Magma - fix mixin inject
             p_38711_.m_5634_(1.0F);
-            this.m_38703_(6.0F);
+            this.prepareExhaustion(p_38711_, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.REGEN); //Magma
+            this.m_38703_(entityhuman == null ? 6.0F : entityhuman.f_19853_.spigotConfig.regenExhaustion); //Magma - add null check
             this.f_38699_ = 0;
          }
       } else if (this.f_38696_ <= 0) {
          ++this.f_38699_;
-         if (this.f_38699_ >= 80) {
+         if (this.f_38699_ >= (starvationRateModified ? starvationRate : 80)) { // CraftBukkit - add starvationRate //Magma - mixin inject
             if (p_38711_.m_21223_() > 10.0F || difficulty == Difficulty.HARD || p_38711_.m_21223_() > 1.0F && difficulty == Difficulty.NORMAL) {
                p_38711_.m_6469_(DamageSource.f_19313_, 1.0F);
             }
@@ -104,9 +_,29 @@
       return this.f_38696_ < 20;
    }
 
+   //Magma start - mixin inject
+   private @org.jetbrains.annotations.Nullable Player addExhaustion_player;
+   private @org.jetbrains.annotations.Nullable org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason addExhaustion_reason;
+   public void prepareExhaustion(Player player, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason reason) {
+      this.addExhaustion_player = player;
+      this.addExhaustion_reason = reason;
+   }
+
    public void m_38703_(float p_38704_) {
-      this.f_38698_ = Math.min(this.f_38698_ + p_38704_, 40.0F);
+
+      if (addExhaustion_player != null) {
+         // CraftBukkit start
+         org.bukkit.event.entity.EntityExhaustionEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerExhaustionEvent(addExhaustion_player, addExhaustion_reason, p_38704_);
+         if (!event.isCancelled()) {
+            this.f_38698_ = Math.min(this.f_38698_ + event.getExhaustion(), 40.0F);
+         }
+         // CraftBukkit end
+      } else this.f_38698_ = Math.min(this.f_38698_ + p_38704_, 40.0F);
+
+      this.addExhaustion_player = null;
+      this.addExhaustion_reason = null;
    }
+   //Magma end
 
    public float m_150380_() {
       return this.f_38698_;
