--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -75,6 +_,11 @@
       return false;
    }
 
+   // Magma start - NMS
+   public void setStrengthPublic(int i) {
+      this.setStrength(i);
+   }
+   // Magma end
    public void setStrength(int p_30841_) {
       this.entityData.set(DATA_STRENGTH_ID, Math.max(1, Math.min(5, p_30841_)));
    }
