--- a/net/minecraft/data/loot/BlockLoot.java
+++ b/net/minecraft/data/loot/BlockLoot.java
@@ -250,7 +_,7 @@
       return LootTable.lootTable();
    }
 
-   public void accept(BiConsumer<ResourceLocation, LootTable.Builder> p_124179_) {
+   protected void addTables() {
       this.dropSelf(Blocks.GRANITE);
       this.dropSelf(Blocks.POLISHED_GRANITE);
       this.dropSelf(Blocks.DIORITE);
@@ -1252,9 +_,13 @@
       this.add(Blocks.NETHER_PORTAL, noDrop());
       this.add(Blocks.BUDDING_AMETHYST, noDrop());
       this.add(Blocks.POWDER_SNOW, noDrop());
+   }
+
+   public void accept(BiConsumer<ResourceLocation, LootTable.Builder> p_124179_) {
+      this.addTables();
       Set<ResourceLocation> set = Sets.newHashSet();
 
-      for(Block block : Registry.BLOCK) {
+      for(Block block : getKnownBlocks()) {
          ResourceLocation resourcelocation = block.getLootTable();
          if (resourcelocation != BuiltInLootTables.EMPTY && set.add(resourcelocation)) {
             LootTable.Builder loottable$builder = this.map.remove(resourcelocation);
@@ -1279,6 +_,10 @@
 
    public static LootTable.Builder createDoorTable(Block p_124138_) {
       return createSinglePropConditionTable(p_124138_, DoorBlock.HALF, DoubleBlockHalf.LOWER);
+   }
+
+   protected Iterable<Block> getKnownBlocks() {
+       return Registry.BLOCK;
    }
 
    public void dropPottedContents(Block p_124253_) {
