--- a/net/minecraft/data/loot/EntityLoot.java
+++ b/net/minecraft/data/loot/EntityLoot.java
@@ -49,7 +_,7 @@
       return LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(p_124375_))).withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootTableReference.lootTableReference(EntityType.SHEEP.getDefaultLootTable())));
    }
 
-   public void accept(BiConsumer<ResourceLocation, LootTable.Builder> p_124377_) {
+   protected void addTables() {
       this.add(EntityType.ARMOR_STAND, LootTable.lootTable());
       this.add(EntityType.AXOLOTL, LootTable.lootTable());
       this.add(EntityType.BAT, LootTable.lootTable());
@@ -141,11 +_,15 @@
       this.add(EntityType.PIGLIN, LootTable.lootTable());
       this.add(EntityType.PIGLIN_BRUTE, LootTable.lootTable());
       this.add(EntityType.ZOMBIE_VILLAGER, LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.ROTTEN_FLESH).apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F))).apply(LootingEnchantFunction.lootingMultiplier(UniformGenerator.between(0.0F, 1.0F))))).withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.IRON_INGOT)).add(LootItem.lootTableItem(Items.CARROT)).add(LootItem.lootTableItem(Items.POTATO).apply(SmeltItemFunction.smelted().when(LootItemEntityPropertyCondition.hasProperties(LootContext.EntityTarget.THIS, ENTITY_ON_FIRE)))).when(LootItemKilledByPlayerCondition.killedByPlayer()).when(LootItemRandomChanceWithLootingCondition.randomChanceAndLootingBoost(0.025F, 0.01F))));
+   }
+
+   public void accept(BiConsumer<ResourceLocation, LootTable.Builder> p_124377_) {
+      this.addTables();
       Set<ResourceLocation> set = Sets.newHashSet();
 
-      for(EntityType<?> entitytype : Registry.ENTITY_TYPE) {
+      for(EntityType<?> entitytype : getKnownEntities()) {
          ResourceLocation resourcelocation = entitytype.getDefaultLootTable();
-         if (!SPECIAL_LOOT_TABLE_TYPES.contains(entitytype) && entitytype.getCategory() == MobCategory.MISC) {
+         if (isNonLiving(entitytype)) {
             if (resourcelocation != BuiltInLootTables.EMPTY && this.map.remove(resourcelocation) != null) {
                throw new IllegalStateException(String.format("Weird loottable '%s' for '%s', not a LivingEntity so should not have loot", resourcelocation, Registry.ENTITY_TYPE.getKey(entitytype)));
             }
@@ -160,6 +_,14 @@
       }
 
       this.map.forEach(p_124377_);
+   }
+
+   protected Iterable<EntityType<?>> getKnownEntities() {
+      return Registry.ENTITY_TYPE;
+   }
+
+   protected boolean isNonLiving(EntityType<?> entitytype) {
+      return !SPECIAL_LOOT_TABLE_TYPES.contains(entitytype) && entitytype.getCategory() == MobCategory.MISC;
    }
 
    protected void add(EntityType<?> p_124372_, LootTable.Builder p_124373_) {
