--- a/net/minecraft/server/players/StoredUserList.java
+++ b/net/minecraft/server/players/StoredUserList.java
@@ -72,6 +_,12 @@
       return this.map.keySet().toArray(new String[0]);
    }
 
+   // CraftBukkit start
+   public Collection<V> getValues() {
+      return this.map.values();
+   }
+   // CraftBukkit end
+
    public boolean isEmpty() {
       return this.map.size() < 1;
    }
@@ -147,6 +_,14 @@
                   this.map.put(this.getKeyForUser(storeduserentry.getUser()), (V)storeduserentry);
                }
             }
+         // Spigot Start
+         } catch ( com.google.gson.JsonParseException | NullPointerException ex )
+         {
+            org.bukkit.Bukkit.getLogger().log( java.util.logging.Level.WARNING, "Unable to read file " + this.file + ", backing it up to {0}.backup and creating new copy.", ex );
+            File backup = new File( this.file + ".backup" );
+            this.file.renameTo( backup );
+            this.file.delete();
+            // Spigot End
          } catch (Throwable throwable1) {
             if (bufferedreader != null) {
                try {
