--- a/net/minecraft/commands/synchronization/ArgumentTypes.java
+++ b/net/minecraft/commands/synchronization/ArgumentTypes.java
@@ -64,7 +_,7 @@
 
 public class ArgumentTypes {
    private static final Logger LOGGER = LogUtils.getLogger();
-   private static final Map<Class<?>, ArgumentTypes.Entry<?>> BY_CLASS = Maps.newHashMap();
+   public static final Map<Class<?>, ArgumentTypes.Entry<?>> BY_CLASS = Maps.newHashMap();
    private static final Map<ResourceLocation, ArgumentTypes.Entry<?>> BY_NAME = Maps.newHashMap();
 
    public static <T extends ArgumentType<?>> void register(String p_121602_, Class<T> p_121603_, ArgumentSerializer<T> p_121604_) {
@@ -122,7 +_,7 @@
       register("uuid", UuidArgument.class, new EmptyArgumentSerializer<>(UuidArgument::uuid));
       register("resource", fixClassType(ResourceKeyArgument.class), new ResourceKeyArgument.Serializer());
       register("resource_or_tag", fixClassType(ResourceOrTagLocationArgument.class), new ResourceOrTagLocationArgument.Serializer());
-      if (SharedConstants.IS_RUNNING_IN_IDE) {
+      if (net.minecraftforge.gametest.ForgeGameTestHooks.isGametestEnabled()) {
          register("test_argument", TestFunctionArgument.class, new EmptyArgumentSerializer<>(TestFunctionArgument::testFunctionArgument));
          register("test_class", TestClassNameArgument.class, new EmptyArgumentSerializer<>(TestClassNameArgument::testClassName));
       }
@@ -139,7 +_,7 @@
    }
 
    @Nullable
-   private static ArgumentTypes.Entry<?> get(ArgumentType<?> p_121617_) {
+   public static ArgumentTypes.Entry<?> get(ArgumentType<?> p_121617_) {
       return BY_CLASS.get(p_121617_.getClass());
    }
 
@@ -254,7 +_,7 @@
       }
    }
 
-   static class Entry<T extends ArgumentType<?>> {
+   public static class Entry<T extends ArgumentType<?>> {
       public final ArgumentSerializer<T> serializer;
       public final ResourceLocation name;
 
@@ -262,5 +_,9 @@
          this.serializer = p_211034_;
          this.name = p_211035_;
       }
+   }
+   @javax.annotation.Nullable public static ResourceLocation getId(ArgumentType<?> type) {
+      Entry<?> entry = get(type);
+      return entry == null ? null : entry.name;
    }
 }
