--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -230,7 +_,8 @@
             }
 
             if (i > 0) {
-               ((LivingEntity)p_27722_).addEffect(new MobEffectInstance(MobEffects.POISON, i * 20, 0), this);
+               ((LivingEntity)p_27722_).prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Magma
+               ((LivingEntity)p_27722_).addEffect(new MobEffectInstance(MobEffects.POISON, i * 20, 0), this); // CraftBukkit
             }
          }
 
@@ -471,7 +_,7 @@
          return false;
       } else {
          BlockEntity blockentity = this.level.getBlockEntity(this.hivePos);
-         return blockentity != null && blockentity.getType() == BlockEntityType.BEEHIVE;
+         return blockentity instanceof BeehiveBlockEntity;
       }
    }
 
@@ -601,11 +_,15 @@
       if (this.isInvulnerableTo(p_27762_)) {
          return false;
       } else {
-         if (!this.level.isClientSide) {
+         // CraftBukkit start
+         boolean result = super.hurt(p_27762_, p_27763_);
+
+         if (result && !this.level.isClientSide) {
             this.beePollinateGoal.stopPollinating();
          }
 
-         return super.hurt(p_27762_, p_27763_);
+         return result;
+         // CraftBukkit end
       }
    }
 
@@ -917,7 +_,7 @@
                      ((BonemealableBlock)blockstate.getBlock()).performBonemeal((ServerLevel)Bee.this.level, Bee.this.random, blockpos, blockstate);
                   }
 
-                  if (flag) {
+                  if (flag && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(Bee.this, blockpos, blockstate.setValue(integerproperty, (Integer) blockstate.getValue(integerproperty) + 1)).isCancelled()) { // Spigot
                      Bee.this.level.levelEvent(2005, blockpos, 0);
                      Bee.this.level.setBlockAndUpdate(blockpos, blockstate.setValue(integerproperty, Integer.valueOf(blockstate.getValue(integerproperty) + 1)));
                      Bee.this.incrementNumCropsGrownSincePollination();
@@ -940,7 +_,7 @@
 
       protected void alertOther(Mob p_28035_, LivingEntity p_28036_) {
          if (p_28035_ instanceof Bee && this.mob.hasLineOfSight(p_28036_)) {
-            p_28035_.setTarget(p_28036_);
+            p_28035_.setTarget(p_28036_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
          }
 
       }
