--- a/net/minecraft/world/entity/animal/Parrot.java
+++ b/net/minecraft/world/entity/animal/Parrot.java
@@ -240,7 +_,7 @@
          }
 
          if (!this.level.isClientSide) {
-            if (this.random.nextInt(10) == 0) {
+            if (this.random.nextInt(10) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_29414_) || !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityTameEvent(this, p_29414_).isCancelled()) { // CraftBukkit
                this.tame(p_29414_);
                this.level.broadcastEntityEvent(this, (byte)7);
             } else {
@@ -254,7 +_,7 @@
             itemstack.shrink(1);
          }
 
-         this.addEffect(new MobEffectInstance(MobEffects.POISON, 900));
+         this.addEffect(new MobEffectInstance(MobEffects.POISON, 900), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // CraftBukkit
          if (p_29414_.isCreative() || !this.isInvulnerable()) {
             this.hurt(DamageSource.playerAttack(p_29414_), Float.MAX_VALUE);
          }
@@ -351,7 +_,7 @@
    }
 
    public boolean isPushable() {
-      return true;
+      return super.isPushable(); // CraftBukkit - collidable API
    }
 
    protected void doPush(Entity p_29367_) {
@@ -365,7 +_,7 @@
          return false;
       } else {
          if (!this.level.isClientSide) {
-            this.setOrderedToSit(false);
+            // this.setOrderedToSit(false);  // CraftBukkit - moved into LivingEntity.damageEntity(DamageSource, float)
          }
 
          return super.hurt(p_29378_, p_29379_);
