--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -201,7 +_,7 @@
    }
 
    public void thunderHit(ServerLevel p_29473_, LightningBolt p_29474_) {
-      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL) {
+      if (p_29473_.getDifficulty() != Difficulty.PEACEFUL && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> {})) {
          ZombifiedPiglin zombifiedpiglin = EntityType.ZOMBIFIED_PIGLIN.create(p_29473_);
          zombifiedpiglin.setItemSlot(EquipmentSlot.MAINHAND, new ItemStack(Items.GOLDEN_SWORD));
          zombifiedpiglin.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
@@ -213,7 +_,14 @@
          }
 
          zombifiedpiglin.setPersistenceRequired();
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombifiedpiglin);
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPigZapEvent(this, p_29474_, zombifiedpiglin).isCancelled()) {
+            return;
+         }
+         p_29473_.prepareEntity(zombifiedpiglin, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); //Magma - add spawn reason
          p_29473_.addFreshEntity(zombifiedpiglin);
+         // CraftBukkit end
          this.discard();
       } else {
          super.thunderHit(p_29473_, p_29474_);
