--- a/net/minecraft/world/entity/animal/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/Pufferfish.java
@@ -25,6 +_,6 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 
 public class Pufferfish extends AbstractFish {
    private static final EntityDataAccessor<Integer> f_29596_ = SynchedEntityData.m_135353_(Pufferfish.class, EntityDataSerializers.f_135028_);
@@ -132,7 +_,8 @@
    private void m_29605_(Mob p_29606_) {
       int i = this.m_29631_();
       if (p_29606_.m_6469_(DamageSource.m_19370_(this), (float)(1 + i))) {
-         p_29606_.m_147207_(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this);
+         p_29606_.prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Magma
+         p_29606_.m_147207_(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this); // CraftBukkit
          this.m_5496_(SoundEvents.f_12295_, 1.0F, 1.0F);
       }
 
@@ -145,7 +_,8 @@
             ((ServerPlayer)p_29617_).f_8906_.m_141995_(new ClientboundGameEventPacket(ClientboundGameEventPacket.f_132162_, 0.0F));
          }
 
-         p_29617_.m_147207_(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this);
+         p_29617_.prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Magma
+         p_29617_.m_147207_(new MobEffectInstance(MobEffects.f_19614_, 60 * i, 0), this); // CraftBukkit
       }
 
    }
