--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -40,7 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.Vec3;
 
-public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackMob {
+public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackMob, net.minecraftforge.common.IForgeShearable {
    private static final EntityDataAccessor<Byte> DATA_PUMPKIN_ID = SynchedEntityData.defineId(SnowGolem.class, EntityDataSerializers.BYTE);
    private static final byte PUMPKIN_FLAG = 16;
    private static final float EYE_HEIGHT = 1.7F;
@@ -94,10 +_,10 @@
          BlockPos blockpos = new BlockPos(i, j, k);
          Biome biome = this.level.getBiome(blockpos).value();
          if (biome.shouldSnowGolemBurn(blockpos)) {
-            this.hurt(DamageSource.ON_FIRE, 1.0F);
+            this.hurt(org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.MELTING, 1.0F); // CraftBukkit - DamageSource.BURN -> org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.MELTING
          }
 
-         if (!this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+         if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level, this)) {
             return;
          }
 
@@ -108,8 +_,8 @@
             j = Mth.floor(this.getY());
             k = Mth.floor(this.getZ() + (double)((float)(l / 2 % 2 * 2 - 1) * 0.25F));
             BlockPos blockpos1 = new BlockPos(i, j, k);
-            if (this.level.getBlockState(blockpos1).isAir() && blockstate.canSurvive(this.level, blockpos1)) {
-               this.level.setBlockAndUpdate(blockpos1, blockstate);
+            if (this.level.isEmptyBlock(blockpos1) && blockstate.canSurvive(this.level, blockpos1)) {
+               org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockFormEvent(this.level, blockpos1, blockstate, this); // CraftBukkit
             }
          }
       }
@@ -134,7 +_,7 @@
 
    protected InteractionResult mobInteract(Player p_29920_, InteractionHand p_29921_) {
       ItemStack itemstack = p_29920_.getItemInHand(p_29921_);
-      if (itemstack.is(Items.SHEARS) && this.readyForShearing()) {
+      if (false && itemstack.getItem() == Items.SHEARS && this.readyForShearing()) { //Forge: Moved to onSheared
          this.shear(SoundSource.PLAYERS);
          this.gameEvent(GameEvent.SHEAR, p_29920_);
          if (!this.level.isClientSide) {
@@ -153,7 +_,9 @@
       this.level.playSound((Player)null, this, SoundEvents.SNOW_GOLEM_SHEAR, p_29907_, 1.0F, 1.0F);
       if (!this.level.isClientSide()) {
          this.setPumpkin(false);
+         this.forceDrops = true; // CraftBukkit
          this.spawnAtLocation(new ItemStack(Items.CARVED_PUMPKIN), 1.7F);
+         this.forceDrops = false; // CraftBukkit
       }
 
    }
@@ -193,5 +_,22 @@
 
    public Vec3 getLeashOffset() {
       return new Vec3(0.0D, (double)(0.75F * this.getEyeHeight()), (double)(this.getBbWidth() * 0.4F));
+   }
+
+   @Override
+   public boolean isShearable(@javax.annotation.Nonnull ItemStack item, Level world, BlockPos pos) {
+      return readyForShearing();
+   }
+
+   @javax.annotation.Nonnull
+   @Override
+   public java.util.List<ItemStack> onSheared(@Nullable Player player, @javax.annotation.Nonnull ItemStack item, Level world, BlockPos pos, int fortune) {
+      world.playSound(null, this, SoundEvents.SNOW_GOLEM_SHEAR, player == null ? SoundSource.BLOCKS : SoundSource.PLAYERS, 1.0F, 1.0F);
+      this.gameEvent(GameEvent.SHEAR, player);
+      if (!world.isClientSide()) {
+         setPumpkin(false);
+         return java.util.Collections.singletonList(new ItemStack(Items.CARVED_PUMPKIN));
+      }
+      return java.util.Collections.emptyList();
    }
 }
