--- a/net/minecraft/world/entity/animal/Turtle.java
+++ b/net/minecraft/world/entity/animal/Turtle.java
@@ -284,7 +_,9 @@
    protected void ageBoundaryReached() {
       super.ageBoundaryReached();
       if (!this.isBaby() && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
+         this.forceDrops = true; // CraftBukkit
          this.spawnAtLocation(Items.SCUTE, 1);
+         this.forceDrops = false; // CraftBukkit
       }
 
    }
@@ -308,7 +_,9 @@
    }
 
    public void thunderHit(ServerLevel p_30140_, LightningBolt p_30141_) {
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.entityDamage = p_30141_; // CraftBukkit
       this.hurt(DamageSource.LIGHTNING_BOLT, Float.MAX_VALUE);
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.entityDamage = null; // CraftBukkit
    }
 
    static class TurtleBreedGoal extends BreedGoal {
@@ -467,8 +_,12 @@
                this.turtle.setLayingEgg(true);
             } else if (this.turtle.layEggCounter > this.adjustedTickDelay(200)) {
                Level level = this.turtle.level;
-               level.playSound((Player)null, blockpos, SoundEvents.TURTLE_LAY_EGG, SoundSource.BLOCKS, 0.3F, 0.9F + level.random.nextFloat() * 0.2F);
-               level.setBlock(this.blockPos.above(), Blocks.TURTLE_EGG.defaultBlockState().setValue(TurtleEggBlock.EGGS, Integer.valueOf(this.turtle.random.nextInt(4) + 1)), 3);
+               // CraftBukkit start
+               if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(this.turtle, this.blockPos.above(), (BlockState) Blocks.TURTLE_EGG.defaultBlockState().setValue(TurtleEggBlock.EGGS, this.turtle.random.nextInt(4) + 1)).isCancelled()) {
+                  level.playSound((Player)null, blockpos, SoundEvents.TURTLE_LAY_EGG, SoundSource.BLOCKS, 0.3F, 0.9F + level.random.nextFloat() * 0.2F);
+                  level.setBlock(this.blockPos.above(), Blocks.TURTLE_EGG.defaultBlockState().setValue(TurtleEggBlock.EGGS, Integer.valueOf(this.turtle.random.nextInt(4) + 1)), 3);
+               }
+               // CraftBukkit end
                this.turtle.setHasEgg(false);
                this.turtle.setLayingEgg(false);
                this.turtle.setInLoveTime(600);
