--- a/net/minecraft/world/entity/animal/Wolf.java
+++ b/net/minecraft/world/entity/animal/Wolf.java
@@ -123,6 +_,16 @@
       return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, (double)0.3F).add(Attributes.MAX_HEALTH, 8.0D).add(Attributes.ATTACK_DAMAGE, 2.0D);
    }
 
+   // CraftBukkit - add overriden version
+   @Override
+   public boolean setTarget(LivingEntity entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fire) {
+      if (!super.setTarget(entityliving, reason, fire)) {
+         return false;
+      }
+      entityliving = getTarget();
+      return true;
+   }
+   // CraftBukkit end
    protected void defineSynchedData() {
       super.defineSynchedData();
       this.entityData.define(DATA_INTERESTED_ID, false);
@@ -284,7 +_,7 @@
       } else {
          Entity entity = p_30386_.getEntity();
          if (!this.level.isClientSide) {
-            this.setOrderedToSit(false);
+            // this.setOrderedToSit(false); // CraftBukkit - moved into LivingEntity.damageEntity(DamageSource, float)
          }
 
          if (entity != null && !(entity instanceof Player) && !(entity instanceof AbstractArrow)) {
@@ -308,7 +_,7 @@
       super.setTame(p_30443_);
       if (p_30443_) {
          this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(20.0D);
-         this.setHealth(20.0F);
+         this.setHealth(this.getMaxHealth()); // CraftBukkit - 20.0 -> getMaxHealth()
       } else {
          this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(8.0D);
       }
@@ -325,11 +_,12 @@
       } else {
          if (this.isTame()) {
             if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
+               this.setRegainReason(org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.EATING); //Magma
+               this.heal((float)itemstack.getFoodProperties(this).getNutrition());
                if (!p_30412_.getAbilities().instabuild) {
                   itemstack.shrink(1);
                }
 
-               this.heal((float)item.getFoodProperties().getNutrition());
                this.gameEvent(GameEvent.MOB_INTERACT, this.eyeBlockPosition());
                return InteractionResult.SUCCESS;
             }
@@ -340,7 +_,7 @@
                   this.setOrderedToSit(!this.isOrderedToSit());
                   this.jumping = false;
                   this.navigation.stop();
-                  this.setTarget((LivingEntity)null);
+                  this.setTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit - reason
                   return InteractionResult.SUCCESS;
                }
 
@@ -360,8 +_,9 @@
             if (!p_30412_.getAbilities().instabuild) {
                itemstack.shrink(1);
             }
-
-            if (this.random.nextInt(3) == 0) {
+            
+            // CraftBukkit - added event call and isCancelled check.
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_30412_) || !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityTameEvent(this, p_30412_).isCancelled()) {
                this.tame(p_30412_);
                this.navigation.stop();
                this.setTarget((LivingEntity)null);
@@ -401,7 +_,7 @@
 
    public boolean isFood(ItemStack p_30440_) {
       Item item = p_30440_.getItem();
-      return item.isEdible() && item.getFoodProperties().isMeat();
+      return item.isEdible() && p_30440_.getFoodProperties(this).isMeat();
    }
 
    public int getMaxSpawnClusterSize() {
