--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -1,7 +_,7 @@
 package net.minecraft.commands;
 
 import java.util.UUID;
 import net.minecraft.network.chat.Component;
 
 public interface CommandSource {
    CommandSource f_80164_ = new CommandSource() {
@@ -19,6 +_,14 @@
       public boolean m_6102_() {
          return false;
       }
+
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         org.apache.logging.log4j.LogManager.getLogger("Magma").debug("getBukkitSender is not supported yet");
+         return null; //Magma - Don't throw new UnsupportedOperationException("getBukkitSender is not supported yet"); Exception. Let's just return null :)
+      }
+      // CraftBukkit end
    };
 
    void m_6352_(Component p_80166_, UUID p_80167_);
@@ -32,4 +_,6 @@
    default boolean m_142559_() {
       return false;
    }
+
+   org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper); // CraftBukkit
 }
