--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -1,8 +_,9 @@
 package net.minecraft.commands;
 
-import java.util.UUID;
 import net.minecraft.network.chat.Component;
 
+import java.util.UUID;
+
 public interface CommandSource {
    CommandSource f_80164_ = new CommandSource() {
       public void m_6352_(Component p_80172_, UUID p_80173_) {
@@ -19,6 +_,13 @@
       public boolean m_6102_() {
          return false;
       }
+
+      // CraftBukkit start
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         throw new UnsupportedOperationException("Not supported yet.");
+      }
+      // CraftBukkit end
    };
 
    void m_6352_(Component p_80166_, UUID p_80167_);
@@ -32,4 +_,6 @@
    default boolean m_142559_() {
       return false;
    }
+
+   org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper); // CraftBukkit
 }
