--- a/net/minecraft/commands/CommandFunction.java
+++ b/net/minecraft/commands/CommandFunction.java
@@ -17,6 +_,16 @@
    private final CommandFunction.Entry[] entries;
    final ResourceLocation id;
 
+   // Paper start
+   public co.aikar.timings.Timing timing;
+   public co.aikar.timings.Timing getTiming() {
+      if (timing == null) {
+         timing = co.aikar.timings.MinecraftTimings.getCommandFunctionTiming(this);
+      }
+      return timing;
+   }
+   // Paper end
+
    public CommandFunction(ResourceLocation p_77979_, CommandFunction.Entry[] p_77980_) {
       this.id = p_77979_;
       this.entries = p_77980_;
