--- a/net/minecraft/world/level/levelgen/structure/SwamplandHutPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/SwamplandHutPiece.java
@@ -91,6 +_,7 @@
                witch.setPersistenceRequired();
                witch.moveTo((double)blockpos.getX() + 0.5D, (double)blockpos.getY(), (double)blockpos.getZ() + 0.5D, 0.0F, 0.0F);
                witch.finalizeSpawn(p_192666_, p_192666_.getCurrentDifficultyAt(blockpos), MobSpawnType.STRUCTURE, (SpawnGroupData)null, (CompoundTag)null);
+               p_192666_.prepareEntity(witch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); //Magma - add spawn reason
                p_192666_.addFreshEntityWithPassengers(witch);
             }
          }
@@ -108,6 +_,7 @@
             cat.setPersistenceRequired();
             cat.moveTo((double)blockpos.getX() + 0.5D, (double)blockpos.getY(), (double)blockpos.getZ() + 0.5D, 0.0F, 0.0F);
             cat.finalizeSpawn(p_73644_, p_73644_.getCurrentDifficultyAt(blockpos), MobSpawnType.STRUCTURE, (SpawnGroupData)null, (CompoundTag)null);
+            p_73644_.prepareEntity(cat, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); //Magma - add spawn reason
             p_73644_.addFreshEntityWithPassengers(cat);
          }
       }
