--- a/com/mojang/blaze3d/vertex/BufferBuilder.java
+++ b/com/mojang/blaze3d/vertex/BufferBuilder.java
@@ -332,6 +_,7 @@
       }
 
       ByteBuffer bytebuffer = this.buffer.slice();
+      bytebuffer.order(this.buffer.order()); // FORGE: Fix incorrect byte order
       this.buffer.clear();
       return Pair.of(bufferbuilder$drawstate, bytebuffer);
    }
@@ -444,4 +_,16 @@
          this.sortZ = p_166829_;
       }
    }
+
+   // Forge start
+   public void putBulkData(ByteBuffer buffer) {
+      ensureCapacity(buffer.limit() + this.format.getVertexSize());
+      this.buffer.position(this.vertices * this.format.getVertexSize());
+      this.buffer.put(buffer);
+      this.vertices += buffer.limit() / this.format.getVertexSize();
+      this.nextElementByte += buffer.limit();
+   }
+
+   @Override
+   public VertexFormat getVertexFormat() { return this.format; }
 }
