--- a/com/mojang/blaze3d/vertex/VertexMultiConsumer.java
+++ b/com/mojang/blaze3d/vertex/VertexMultiConsumer.java
@@ -91,6 +_,10 @@
          this.first.unsetDefaultColor();
          this.second.unsetDefaultColor();
       }
+
+      public VertexFormat getVertexFormat() {
+         return this.first.getVertexFormat();
+      }
    }
 
    @OnlyIn(Dist.CLIENT)
@@ -176,6 +_,12 @@
 
       public void unsetDefaultColor() {
          this.forEach(VertexConsumer::unsetDefaultColor);
+      }
+
+      public VertexFormat getVertexFormat() {
+         if(delegates.length > 0)
+            return delegates[0].getVertexFormat();
+         return DefaultVertexFormat.BLOCK;
       }
    }
 }
