--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -11,7 +_,14 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.TagKey;
 
-public interface Holder<T> {
+public interface Holder<T> extends net.minecraftforge.registries.tags.IReverseTag<T> {
+
+   @Override
+   default Stream<TagKey<T>> getTagKeys() {
+      return this.tags();
+   }
+
+
    T value();
 
    boolean isBound();
@@ -43,6 +_,12 @@
    }
 
    public static record Direct<T>(T value) implements Holder<T> {
+
+      @Override
+      public boolean containsTag(TagKey<T> key) {
+         return is(key);
+      }
+
       public boolean isBound() {
          return true;
       }
@@ -98,6 +_,10 @@
    }
 
    public static class Reference<T> implements Holder<T> {
+      @Override
+      public boolean containsTag(TagKey<T> key) {
+         return is(key);
+      }
       private final Registry<T> registry;
       private Set<TagKey<T>> tags = Set.of();
       private final Holder.Reference.Type type;
@@ -192,6 +_,10 @@
 
       public Stream<TagKey<T>> tags() {
          return this.tags.stream();
+      }
+
+      public Type getType() {
+         return this.type;
       }
 
       public String toString() {
