--- a/net/minecraft/core/MappedRegistry.java
+++ b/net/minecraft/core/MappedRegistry.java
@@ -90,7 +_,8 @@
       this.byId.size(Math.max(this.byId.size(), p_205858_ + 1));
       this.toId.put(p_205860_, p_205858_);
       this.holdersInOrder = null;
-      if (p_205862_ && this.byKey.containsKey(p_205859_)) {
+      // Forge: Fix bug where a key is considered a duplicate if getOrCreateHolder was called before the entry was registered
+      if (p_205862_ && this.byKey.containsKey(p_205859_) && this.byKey.get(p_205859_).isBound()) {
          Util.logAndPauseIfInIde("Adding duplicate key '" + p_205859_ + "' to registry");
       }
 
@@ -277,6 +_,14 @@
 
    public boolean containsKey(ResourceKey<T> p_175392_) {
       return this.byKey.containsKey(p_175392_);
+   }
+
+   /** @deprecated Forge: For internal use only. Use the Register events when registering values. */
+   @Deprecated
+   public void unfreeze() {
+      this.frozen = false;
+      if (this.customHolderProvider != null && this.intrusiveHolderCache == null)
+         this.intrusiveHolderCache = new IdentityHashMap<>();
    }
 
    public Registry<T> freeze() {
