--- a/net/minecraft/core/RegistryCodecs.java
+++ b/net/minecraft/core/RegistryCodecs.java
@@ -95,7 +_,8 @@
    }
 
    private static <T> Codec<Map<ResourceKey<T>, T>> directCodec(ResourceKey<? extends Registry<T>> p_206316_, Codec<T> p_206317_) {
-      return Codec.unboundedMap(ResourceLocation.CODEC.xmap(ResourceKey.elementKey(p_206316_), ResourceKey::location), p_206317_);
+      // FORGE: Fix MC-197860
+      return new net.minecraftforge.common.LenientUnboundedMapCodec<>(ResourceLocation.CODEC.xmap(ResourceKey.elementKey(p_206316_), ResourceKey::location), p_206317_);
    }
 
    public static <E> Codec<HolderSet<E>> homogeneousList(ResourceKey<? extends Registry<E>> p_206280_, Codec<E> p_206281_) {
