--- a/net/minecraft/world/entity/ambient/Bat.java
+++ b/net/minecraft/world/entity/ambient/Bat.java
@@ -123,16 +_,24 @@
             }
 
             if (this.level.getNearestPlayer(BAT_RESTING_TARGETING, this) != null) {
+               // CraftBukkit Start - Call BatToggleSleepEvent
+               if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
+                  this.setResting(false);
+                  if (!flag) {
+                     this.level.levelEvent((Player) null, 1025, blockpos, 0);
+                  }
+               }
+               // CraftBukkit End
+            }
+         } else {
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
                this.setResting(false);
                if (!flag) {
                   this.level.levelEvent((Player)null, 1025, blockpos, 0);
                }
             }
-         } else {
-            this.setResting(false);
-            if (!flag) {
-               this.level.levelEvent((Player)null, 1025, blockpos, 0);
-            }
+            // CraftBukkit End - Call BatToggleSleepEvent
          }
       } else {
          if (this.targetPosition != null && (!this.level.isEmptyBlock(this.targetPosition) || this.targetPosition.getY() <= this.level.getMinBuildHeight())) {
@@ -154,7 +_,11 @@
          this.zza = 0.5F;
          this.setYRot(this.getYRot() + f1);
          if (this.random.nextInt(100) == 0 && this.level.getBlockState(blockpos1).isRedstoneConductor(this.level, blockpos1)) {
-            this.setResting(true);
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBatToggleSleepEvent(this, false)) {
+               this.setResting(true);
+            }
+            // CraftBukkit End
          }
       }
 
@@ -180,7 +_,11 @@
          return false;
       } else {
          if (!this.level.isClientSide && this.isResting()) {
-            this.setResting(false);
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
+               this.setResting(false);
+            }
+            // CraftBukkit End - Call BatToggleSleepEvent
          }
 
          return super.hurt(p_27424_, p_27425_);
