--- a/net/minecraft/world/entity/ai/goal/FloatGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FloatGoal.java
@@ -9,6 +_,7 @@
 
    public FloatGoal(Mob p_25230_) {
       this.mob = p_25230_;
+      if (mob.getCommandSenderWorld().paperConfig != null && mob.getCommandSenderWorld().paperConfig.nerfedMobsShouldJump) mob.goalFloat = this; // Paper //Magma - null check
       this.setFlags(EnumSet.of(Goal.Flag.JUMP));
       p_25230_.getNavigation().setCanFloat(true);
    }
