--- a/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FollowOwnerGoal.java
@@ -117,7 +_,19 @@
       } else if (!this.canTeleportTo(new BlockPos(p_25304_, p_25305_, p_25306_))) {
          return false;
       } else {
-         this.tamable.moveTo((double)p_25304_ + 0.5D, (double)p_25305_, (double)p_25306_ + 0.5D, this.tamable.getYRot(), this.tamable.getXRot());
+         // CraftBukkit start
+         org.bukkit.craftbukkit.v1_18_R2.entity.CraftEntity entity = this.tamable.getBukkitEntity();
+         org.bukkit.Location to = new org.bukkit.Location(entity.getWorld(), (double) p_25304_ + 0.5D, (double) p_25305_, (double) p_25306_ + 0.5D, this.tamable.getYRot(), this.tamable.getXRot());
+         org.bukkit.event.entity.EntityTeleportEvent event = new org.bukkit.event.entity.EntityTeleportEvent(entity, entity.getLocation(), to);
+         this.tamable.level.getCraftServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return false;
+         }
+         to = event.getTo();
+
+         this.tamable.moveTo(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+         // CraftBukkit end
+
          this.navigation.stop();
          return true;
       }
