--- a/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RemoveBlockGoal.java
@@ -34,7 +_,7 @@
    }
 
    public boolean canUse() {
-      if (!this.removerMob.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.removerMob.level, this.removerMob)) {
          return false;
       } else if (this.nextStartTick > 0) {
          --this.nextStartTick;
@@ -93,6 +_,14 @@
          }
 
          if (this.ticksSinceReachedGoal > 60) {
+            // CraftBukkit start - Step on eggs
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(this.removerMob.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(level, blockpos1));
+            level.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) event);
+
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             level.removeBlock(blockpos1, false);
             if (!level.isClientSide) {
                for(int i = 0; i < 20; ++i) {
@@ -133,7 +_,9 @@
       if (chunkaccess == null) {
          return false;
       } else {
-         return chunkaccess.getBlockState(p_25851_).is(this.blockToRemove) && chunkaccess.getBlockState(p_25851_.above()).isAir() && chunkaccess.getBlockState(p_25851_.above(2)).isAir();
+         net.minecraft.world.level.block.state.BlockState state = chunkaccess.getBlockState(p_25851_);
+         if (!(state.canEntityDestroy(p_25850_, p_25851_, this.removerMob) && net.minecraftforge.event.ForgeEventFactory.onEntityDestroyBlock(this.removerMob, p_25851_, state))) return false;
+         return state.is(this.blockToRemove) && chunkaccess.getBlockState(p_25851_.above()).isAir() && chunkaccess.getBlockState(p_25851_.above(2)).isAir();
       }
    }
 }
