--- a/net/minecraft/world/Container.java
+++ b/net/minecraft/world/Container.java
@@ -61,4 +_,36 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   //Magma - add a default method for mods that don't override these methods
+   default java.util.List<ItemStack> getContents() {
+      return java.util.Collections.emptyList();
+   }
+
+   void onOpen(org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity who);
+
+   void onClose(org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity who);
+
+   //Magma - add a default value so the server can unload chunks properly when getViewers() is not overridden by a mod
+   default java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+        return java.util.Collections.emptyList();
+   }
+
+   org.bukkit.inventory.InventoryHolder getOwner();
+
+   void setMaxStackSize(int size);
+
+   org.bukkit.Location getLocation();
+
+   default net.minecraft.world.item.crafting.Recipe getCurrentRecipe() {
+      return null;
+   }
+
+   default void setCurrentRecipe(net.minecraft.world.item.crafting.Recipe recipe) {
+   }
+
+   int MAX_STACK = 64;
+// CraftBukkit end
+
 }
