--- a/net/minecraft/nbt/NbtIo.java
+++ b/net/minecraft/nbt/NbtIo.java
@@ -1,24 +_,24 @@
 package net.minecraft.nbt;
 
 import java.io.BufferedOutputStream;
 import java.io.DataInput;
 import java.io.DataInputStream;
 import java.io.DataOutput;
 import java.io.DataOutputStream;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 import javax.annotation.Nullable;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportedException;
 import net.minecraft.util.FastBufferedInputStream;
 
 public class NbtIo {
    public static CompoundTag m_128937_(File p_128938_) throws IOException {
       InputStream inputstream = new FileInputStream(p_128938_);
@@ -222,6 +_,12 @@
    }
 
    public static CompoundTag m_128934_(DataInput p_128935_, NbtAccounter p_128936_) throws IOException {
+      // Spigot start
+      if ( p_128935_ instanceof io.netty.buffer.ByteBufInputStream )
+      {
+         p_128935_ = new DataInputStream(new org.spigotmc.LimitStream((InputStream) p_128935_, p_128936_));
+      }
+      // Spigot end
       Tag tag = m_128930_(p_128935_, 0, p_128936_);
       if (tag instanceof CompoundTag) {
          return (CompoundTag)tag;
@@ -268,10 +_,12 @@
 
    private static Tag m_128930_(DataInput p_128931_, int p_128932_, NbtAccounter p_128933_) throws IOException {
       byte b0 = p_128931_.readByte();
+      p_128933_.m_6800_(8); // Forge: Count everything!
       if (b0 == 0) {
          return EndTag.f_128534_;
       } else {
-         StringTag.m_197563_(p_128931_);
+         p_128933_.readUTF(p_128931_.readUTF()); //Forge: Count this string.
+         p_128933_.m_6800_(32); //Forge: 4 extra bytes for the object allocation.
 
          try {
             return TagTypes.m_129397_(b0).m_7300_(p_128931_, p_128932_, p_128933_);
