--- a/net/minecraft/nbt/StringTag.java
+++ b/net/minecraft/nbt/StringTag.java
@@ -11,7 +_,7 @@
       public StringTag load(DataInput p_129315_, int p_129316_, NbtAccounter p_129317_) throws IOException {
          p_129317_.accountBits(288L);
          String s = p_129315_.readUTF();
-         p_129317_.accountBits((long)(16 * s.length()));
+         p_129317_.readUTF(s);
          return StringTag.valueOf(s);
       }
 
@@ -52,7 +_,7 @@
    }
 
    public static StringTag valueOf(String p_129298_) {
-      return p_129298_.isEmpty() ? EMPTY : new StringTag(p_129298_);
+      return (p_129298_ == null || p_129298_.isEmpty()) ? EMPTY : new StringTag(p_129298_);
    }
 
    public void write(DataOutput p_129296_) throws IOException {
