--- a/net/minecraft/server/packs/repository/PackCompatibility.java
+++ b/net/minecraft/server/packs/repository/PackCompatibility.java
@@ -34,7 +_,7 @@
    }
 
    public static PackCompatibility forMetadata(PackMetadataSection p_143886_, PackType p_143887_) {
-      return forFormat(p_143886_.getPackFormat(), p_143887_);
+      return forFormat(p_143886_.getPackFormat(p_143887_), p_143887_);
    }
 
    public Component getDescription() {
