--- a/net/minecraft/world/item/ArmorItem.java
+++ b/net/minecraft/world/item/ArmorItem.java
@@ -49,7 +_,32 @@
          LivingEntity livingentity = list.get(0);
          EquipmentSlot equipmentslot = Mob.getEquipmentSlotForItem(p_40400_);
          ItemStack itemstack = p_40400_.split(1);
-         livingentity.setItemSlot(equipmentslot, itemstack);
+
+         // CraftBukkit start
+         Level world = p_40399_.getLevel();
+         org.bukkit.block.Block block = world.getWorld().getBlockAt(p_40399_.getPos().getX(), p_40399_.getPos().getY(), p_40399_.getPos().getZ());
+         org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack craftItem = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asCraftMirror(p_40400_);
+         org.bukkit.event.block.BlockDispenseArmorEvent event = new org.bukkit.event.block.BlockDispenseArmorEvent(block, craftItem.clone(), (org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity) livingentity.getBukkitEntity());
+         if (!DispenserBlock.eventFired) {
+            world.getCraftServer().getPluginManager().callEvent(event);
+         }
+         if (event.isCancelled()) {
+            itemstack.grow(1);
+            return false;
+         }
+         if (!event.getItem().equals(craftItem)) {
+            itemstack.grow(1);
+            // Chain to handler for new item
+            ItemStack eventStack = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asNMSCopy(event.getItem());
+            DispenseItemBehavior idispensebehavior = (DispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+            if (idispensebehavior != DispenseItemBehavior.NOOP && idispensebehavior != ArmorItem.DISPENSE_ITEM_BEHAVIOR) {
+               idispensebehavior.dispense(p_40399_, eventStack);
+               return true;
+            }
+         }
+
+         livingentity.setItemSlot(equipmentslot, org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asNMSCopy(event.getItem()));
+         // CraftBukkit end
          if (livingentity instanceof Mob) {
             ((Mob)livingentity).setDropChance(equipmentslot, 2.0F);
             ((Mob)livingentity).setPersistenceRequired();
@@ -71,7 +_,7 @@
       UUID uuid = ARMOR_MODIFIER_UUID_PER_SLOT[p_40387_.getIndex()];
       builder.put(Attributes.ARMOR, new AttributeModifier(uuid, "Armor modifier", (double)this.defense, AttributeModifier.Operation.ADDITION));
       builder.put(Attributes.ARMOR_TOUGHNESS, new AttributeModifier(uuid, "Armor toughness", (double)this.toughness, AttributeModifier.Operation.ADDITION));
-      if (p_40386_ == ArmorMaterials.NETHERITE) {
+      if (this.knockbackResistance > 0) {
          builder.put(Attributes.KNOCKBACK_RESISTANCE, new AttributeModifier(uuid, "Armor knockback resistance", (double)this.knockbackResistance, AttributeModifier.Operation.ADDITION));
       }
 
