--- a/net/minecraft/world/item/GameMasterBlockItem.java
+++ b/net/minecraft/world/item/GameMasterBlockItem.java
@@ -1,10 +_,11 @@
 package net.minecraft.world.item;
 
-import javax.annotation.Nullable;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
+
+import javax.annotation.Nullable;
 
 public class GameMasterBlockItem extends BlockItem {
    public GameMasterBlockItem(Block p_41318_, Item.Properties p_41319_) {
