--- a/net/minecraft/world/item/Items.java
+++ b/net/minecraft/world/item/Items.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.material.Fluids;
 
+@net.minecraftforge.registries.ObjectHolder("minecraft")
 public class Items {
    public static final Item AIR = registerBlock(Blocks.AIR, new AirItem(Blocks.AIR, new Item.Properties()));
    public static final Item STONE = registerBlock(Blocks.STONE, CreativeModeTab.TAB_BUILDING_BLOCKS);
@@ -1138,11 +_,23 @@
    }
 
    private static Item registerBlock(Block p_151090_, CreativeModeTab p_151091_, Block... p_151092_) {
-      BlockItem blockitem = new BlockItem(p_151090_, (new Item.Properties()).tab(p_151091_));
+      BlockItem blockitem = new BlockItem(p_151090_, (new Item.Properties()).tab(p_151091_)) {
+         @Override
+         public void registerBlocks(java.util.Map<Block, Item> map, Item self) {
+            super.registerBlocks(map, self);
+            for (Block b : p_151092_) {
+               map.put(b, self);
+            }
+         }
 
-      for(Block block : p_151092_) {
-         Item.BY_BLOCK.put(block, blockitem);
-      }
+         @Override
+         public void removeFromBlockToItemMap(java.util.Map<Block, Item> map, Item self) {
+            super.removeFromBlockToItemMap(map, self);
+            for (Block b : p_151092_) {
+               map.remove(b);
+            }
+         }
+      };
 
       return registerBlock(blockitem);
    }
