--- a/net/minecraft/world/item/LeadItem.java
+++ b/net/minecraft/world/item/LeadItem.java
@@ -44,8 +_,24 @@
          if (mob.getLeashHolder() == p_42830_) {
             if (leashfenceknotentity == null) {
                leashfenceknotentity = LeashFenceKnotEntity.getOrCreateKnot(p_42831_, p_42832_);
+
+               // CraftBukkit start - fire HangingPlaceEvent
+               org.bukkit.event.hanging.HangingPlaceEvent event = new org.bukkit.event.hanging.HangingPlaceEvent((org.bukkit.entity.Hanging) leashfenceknotentity.getBukkitEntity(), p_42830_ != null ? (org.bukkit.entity.Player) p_42830_.getBukkitEntity() : null, p_42831_.getWorld().getBlockAt(i, j, k), org.bukkit.block.BlockFace.SELF);
+               p_42831_.getCraftServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  leashfenceknotentity.discard();
+                  return InteractionResult.PASS;
+               }
+               // CraftBukkit end
+
                leashfenceknotentity.playPlacementSound();
             }
+
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerLeashEntityEvent(mob, leashfenceknotentity, p_42830_).isCancelled()) {
+               continue;
+            }
+            // CraftBukkit end
 
             mob.setLeashedTo(leashfenceknotentity, true);
             flag = true;
