--- a/net/minecraft/world/level/block/BambooBlock.java
+++ b/net/minecraft/world/level/block/BambooBlock.java
@@ -26,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class BambooBlock extends Block implements BonemealableBlock {
+public class BambooBlock extends Block implements BonemealableBlock, net.minecraftforge.common.IPlantable  {
    protected static final float SMALL_LEAVES_AABB_OFFSET = 3.0F;
    protected static final float LARGE_LEAVES_AABB_OFFSET = 5.0F;
    protected static final float COLLISION_AABB_OFFSET = 1.5F;
@@ -114,10 +_,11 @@
 
    public void randomTick(BlockState p_48936_, ServerLevel p_48937_, BlockPos p_48938_, Random p_48939_) {
       if (p_48936_.getValue(STAGE) == 0) {
-         if (p_48939_.nextInt(3) == 0 && p_48937_.isEmptyBlock(p_48938_.above()) && p_48937_.getRawBrightness(p_48938_.above(), 0) >= 9) {
+         if (p_48937_.isEmptyBlock(p_48938_.above()) && p_48937_.getRawBrightness(p_48938_.above(), 0) >= 9) {
             int i = this.getHeightBelowUpToMax(p_48937_, p_48938_) + 1;
-            if (i < 16) {
+            if (i < 16 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_48937_, p_48938_, p_48936_, p_48939_.nextInt(Math.max(1, (int) (100.0F / p_48937_.spigotConfig.bambooModifier) * 3)) == 0)) { // Spigot
                this.growBamboo(p_48936_, p_48937_, p_48938_, p_48939_, i);
+               net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_48937_, p_48938_, p_48936_);
             }
          }
 
@@ -159,7 +_,7 @@
       for(int i1 = 0; i1 < l; ++i1) {
          BlockPos blockpos = p_48878_.above(i);
          BlockState blockstate = p_48876_.getBlockState(blockpos);
-         if (k >= 16 || blockstate.getValue(STAGE) == 1 || !p_48876_.isEmptyBlock(blockpos.above())) {
+         if (k >= 16 || !blockstate.is(Blocks.BAMBOO) || blockstate.getValue(STAGE) == 1 || !p_48876_.isEmptyBlock(blockpos.above())) { // CraftBukkit - If the BlockSpreadEvent was cancelled, we have no bamboo here
             return;
          }
 
@@ -171,7 +_,7 @@
    }
 
    public float getDestroyProgress(BlockState p_48901_, Player p_48902_, BlockGetter p_48903_, BlockPos p_48904_) {
-      return p_48902_.getMainHandItem().getItem() instanceof SwordItem ? 1.0F : super.getDestroyProgress(p_48901_, p_48902_, p_48903_, p_48904_);
+      return p_48902_.getMainHandItem().canPerformAction(net.minecraftforge.common.ToolActions.SWORD_DIG) ? 1.0F : super.getDestroyProgress(p_48901_, p_48902_, p_48903_, p_48904_);
    }
 
    protected void growBamboo(BlockState p_48911_, Level p_48912_, BlockPos p_48913_, Random p_48914_, int p_48915_) {
@@ -179,13 +_,18 @@
       BlockPos blockpos = p_48913_.below(2);
       BlockState blockstate1 = p_48912_.getBlockState(blockpos);
       BambooLeaves bambooleaves = BambooLeaves.NONE;
+      boolean shouldUpdateOthers = false; // CraftBukkit
+
       if (p_48915_ >= 1) {
          if (blockstate.is(Blocks.BAMBOO) && blockstate.getValue(LEAVES) != BambooLeaves.NONE) {
             if (blockstate.is(Blocks.BAMBOO) && blockstate.getValue(LEAVES) != BambooLeaves.NONE) {
                bambooleaves = BambooLeaves.LARGE;
                if (blockstate1.is(Blocks.BAMBOO)) {
-                  p_48912_.setBlock(p_48913_.below(), blockstate.setValue(LEAVES, BambooLeaves.SMALL), 3);
-                  p_48912_.setBlock(blockpos, blockstate1.setValue(LEAVES, BambooLeaves.NONE), 3);
+                  // CraftBukkit start - moved down
+                  // p_48912_.setBlock(p_48913_.below(), blockstate.setValue(LEAVES, BambooLeaves.SMALL), 3);
+                  // p_48912_.setBlock(blockpos, blockstate1.setValue(LEAVES, BambooLeaves.NONE), 3);
+                  shouldUpdateOthers = true;
+                  // CraftBukkit end
                }
             }
          } else {
@@ -193,9 +_,19 @@
          }
       }
 
-      int i = p_48911_.getValue(AGE) != 1 && !blockstate1.is(Blocks.BAMBOO) ? 0 : 1;
-      int j = (p_48915_ < 11 || !(p_48914_.nextFloat() < 0.25F)) && p_48915_ != 15 ? 0 : 1;
-      p_48912_.setBlock(p_48913_.above(), this.defaultBlockState().setValue(AGE, Integer.valueOf(i)).setValue(LEAVES, bambooleaves).setValue(STAGE, Integer.valueOf(j)), 3);
+      //Magma start - fix bamboo growing into the ground
+      int j = (Integer) p_48911_.getValue(AGE) != 1 && !blockstate1.is(Blocks.BAMBOO) ? 0 : 1;
+      int k = (p_48915_ < 11 || p_48914_.nextFloat() >= 0.25F) && p_48915_ != 15 ? 0 : 1;
+
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_48912_, p_48913_, p_48913_.above(), (BlockState) ((BlockState) ((BlockState) this.defaultBlockState().setValue(AGE, j)).setValue(LEAVES, bambooleaves)).setValue(STAGE, k), 3)) {
+         if (shouldUpdateOthers) {
+            p_48912_.setBlock(p_48913_.below(), blockstate.setValue(LEAVES, BambooLeaves.SMALL), 3);
+            p_48912_.setBlock(blockpos, blockstate1.setValue(LEAVES, BambooLeaves.NONE), 3);
+         }
+      }
+      // CraftBukkit end
+      //Magma end
    }
 
    protected int getHeightAboveUpToMax(BlockGetter p_48883_, BlockPos p_48884_) {
@@ -212,5 +_,12 @@
       }
 
       return i;
+   }
+
+   @Override
+   public BlockState getPlant(BlockGetter world, BlockPos pos) {
+      BlockState state = world.getBlockState(pos);
+      if (state.getBlock() != this) return defaultBlockState();
+      return state;
    }
 }
