--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -39,7 +_,8 @@
 
          if (block != null) {
             BlockState blockstate1 = block.defaultBlockState().setValue(AmethystClusterBlock.FACING, direction).setValue(AmethystClusterBlock.WATERLOGGED, Boolean.valueOf(blockstate.getFluidState().getType() == Fluids.WATER));
-            p_152729_.setBlockAndUpdate(blockpos, blockstate1);
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_152729_, p_152730_, blockpos, blockstate1))
+               p_152729_.setBlockAndUpdate(blockpos, blockstate1);
          }
 
       }
