--- a/net/minecraft/world/level/block/ButtonBlock.java
+++ b/net/minecraft/world/level/block/ButtonBlock.java
@@ -97,6 +_,19 @@
       if (p_51088_.getValue(POWERED)) {
          return InteractionResult.CONSUME;
       } else {
+         // CraftBukkit start
+         boolean powered = ((Boolean) p_51088_.getValue(POWERED));
+         org.bukkit.block.Block block = p_51089_.getWorld().getBlockAt(p_51090_.getX(), p_51090_.getY(), p_51090_.getZ());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(block, old, current);
+         p_51089_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return InteractionResult.SUCCESS;
+         }
+         // CraftBukkit end
          this.press(p_51088_, p_51089_, p_51090_);
          this.playSound(p_51091_, p_51089_, p_51090_, true);
          p_51089_.gameEvent(p_51091_, GameEvent.BLOCK_PRESS, p_51090_);
@@ -143,6 +_,16 @@
          if (this.sensitive) {
             this.checkPressed(p_51073_, p_51074_, p_51075_);
          } else {
+            // CraftBukkit start
+            org.bukkit.block.Block block = p_51074_.getWorld().getBlockAt(p_51075_.getX(), p_51075_.getY(), p_51075_.getZ());
+
+            org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(block, 15, 0);
+            p_51074_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+               return;
+            }
+            // CraftBukkit end
             p_51074_.setBlock(p_51075_, p_51073_.setValue(POWERED, Boolean.valueOf(false)), 3);
             this.updateNeighbours(p_51073_, p_51074_, p_51075_);
             this.playSound((Player)null, p_51074_, p_51075_, false);
@@ -162,7 +_,42 @@
       List<? extends Entity> list = p_51122_.getEntitiesOfClass(AbstractArrow.class, p_51121_.getShape(p_51122_, p_51123_).bounds().move(p_51123_));
       boolean flag = !list.isEmpty();
       boolean flag1 = p_51121_.getValue(POWERED);
+
+      // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+      if (flag1 != flag && flag) {
+         org.bukkit.block.Block block = p_51122_.getWorld().getBlockAt(p_51123_.getX(), p_51123_.getY(), p_51123_.getZ());
+         boolean allowed = false;
+         // If all of the events are cancelled block the button press, else allow
+         for (Object object : list) {
+            if (object != null) {
+               org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+               p_51122_.getCraftServer().getPluginManager().callEvent(event);
+               if (!event.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+         if (!allowed) {
+            return;
+         }
+      }
+      // CraftBukkit end
+
       if (flag != flag1) {
+         // CraftBukkit start
+         boolean powered = flag1;
+         org.bukkit.block.Block block = p_51122_.getWorld().getBlockAt(p_51123_.getX(), p_51123_.getY(), p_51123_.getZ());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(block, old, current);
+         p_51122_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((flag && eventRedstone.getNewCurrent() <= 0) || (!flag && eventRedstone.getNewCurrent() > 0)) {
+            return;
+         }
+         // CraftBukkit end
          p_51122_.setBlock(p_51123_, p_51121_.setValue(POWERED, Boolean.valueOf(flag)), 3);
          this.updateNeighbours(p_51121_, p_51122_, p_51123_);
          this.playSound((Player)null, p_51122_, p_51123_, flag);
