--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -55,17 +_,28 @@
 
    private void trySpawnGolem(Level p_51379_, BlockPos p_51380_) {
       BlockPattern.BlockPatternMatch blockpattern$blockpatternmatch = this.getOrCreateSnowGolemFull().find(p_51379_, p_51380_);
+      org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator(p_51379_); // CraftBukkit - Use BlockStateListPopulator
       if (blockpattern$blockpatternmatch != null) {
          for(int i = 0; i < this.getOrCreateSnowGolemFull().getHeight(); ++i) {
             BlockInWorld blockinworld = blockpattern$blockpatternmatch.getBlock(0, i, 0);
-            p_51379_.setBlock(blockinworld.getPos(), Blocks.AIR.defaultBlockState(), 2);
-            p_51379_.levelEvent(2001, blockinworld.getPos(), Block.getId(blockinworld.getState()));
+            blockList.setBlock(blockinworld.getPos(), Blocks.AIR.defaultBlockState(), 2); // CraftBukkit
+            //p_51379_.levelEvent(2001, blockinworld.getPos(), Block.getId(blockinworld.getState()));
          }
 
          SnowGolem snowgolem = EntityType.SNOW_GOLEM.create(p_51379_);
          BlockPos blockpos1 = blockpattern$blockpatternmatch.getBlock(0, 2, 0).getPos();
          snowgolem.moveTo((double)blockpos1.getX() + 0.5D, (double)blockpos1.getY() + 0.05D, (double)blockpos1.getZ() + 0.5D, 0.0F, 0.0F);
-         p_51379_.addFreshEntity(snowgolem);
+
+         // CraftBukkit start
+         p_51379_.prepareEntity(snowgolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN);
+         if (!p_51379_.addFreshEntity(snowgolem)) {
+            return;
+         }
+         for (BlockPos pos : blockList.getBlocks()) {
+            p_51379_.levelEvent(2001, pos, Block.getId(p_51379_.getBlockState(pos)));
+         }
+         blockList.updateList();
+         // CraftBukkit end
 
          for(ServerPlayer serverplayer : p_51379_.getEntitiesOfClass(ServerPlayer.class, snowgolem.getBoundingBox().inflate(5.0D))) {
             CriteriaTriggers.SUMMONED_ENTITY.trigger(serverplayer, snowgolem);
@@ -81,8 +_,8 @@
             for(int j = 0; j < this.getOrCreateIronGolemFull().getWidth(); ++j) {
                for(int k = 0; k < this.getOrCreateIronGolemFull().getHeight(); ++k) {
                   BlockInWorld blockinworld2 = blockpattern$blockpatternmatch.getBlock(j, k, 0);
-                  p_51379_.setBlock(blockinworld2.getPos(), Blocks.AIR.defaultBlockState(), 2);
-                  p_51379_.levelEvent(2001, blockinworld2.getPos(), Block.getId(blockinworld2.getState()));
+                  blockList.setBlock(blockinworld2.getPos(), Blocks.AIR.defaultBlockState(), 2); // CraftBukkit
+                  //p_51379_.levelEvent(2001, blockinworld2.getPos(), Block.getId(blockinworld2.getState()));
                }
             }
 
@@ -90,7 +_,14 @@
             IronGolem irongolem = EntityType.IRON_GOLEM.create(p_51379_);
             irongolem.setPlayerCreated(true);
             irongolem.moveTo((double)blockpos.getX() + 0.5D, (double)blockpos.getY() + 0.05D, (double)blockpos.getZ() + 0.5D, 0.0F, 0.0F);
-            p_51379_.addFreshEntity(irongolem);
+            // CraftBukkit start
+            p_51379_.prepareEntity(irongolem, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM); //Magma - add spawn reason
+            if (!p_51379_.addFreshEntity(irongolem)) {
+               return;
+            }
+            blockList.getBlocks().forEach(pos -> p_51379_.levelEvent(2001, pos, Block.getId(p_51379_.getBlockState(pos))));
+            blockList.updateList();
+            // CraftBukkit end
 
             for(ServerPlayer serverplayer1 : p_51379_.getEntitiesOfClass(ServerPlayer.class, irongolem.getBoundingBox().inflate(5.0D))) {
                CriteriaTriggers.SUMMONED_ENTITY.trigger(serverplayer1, irongolem);
