--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -52,11 +_,11 @@
 
    protected void receiveStalactiteDrip(BlockState p_152940_, Level p_152941_, BlockPos p_152942_, Fluid p_152943_) {
       if (p_152943_ == Fluids.WATER) {
-         p_152941_.setBlockAndUpdate(p_152942_, Blocks.WATER_CAULDRON.defaultBlockState());
+         LayeredCauldronBlock.changeLevel(p_152940_, p_152941_, p_152942_, Blocks.WATER_CAULDRON.defaultBlockState(), null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
          p_152941_.levelEvent(1047, p_152942_, 0);
          p_152941_.gameEvent((Entity)null, GameEvent.FLUID_PLACE, p_152942_);
       } else if (p_152943_ == Fluids.LAVA) {
-         p_152941_.setBlockAndUpdate(p_152942_, Blocks.LAVA_CAULDRON.defaultBlockState());
+         LayeredCauldronBlock.changeLevel(p_152940_, p_152941_, p_152942_, Blocks.LAVA_CAULDRON.defaultBlockState(), null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
          p_152941_.levelEvent(1046, p_152942_, 0);
          p_152941_.gameEvent((Entity)null, GameEvent.FLUID_PLACE, p_152942_);
       }
