--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -59,7 +_,7 @@
       float f1 = f * f * this.getChanceModifier();
       if (p_153050_.nextFloat() < f1) {
          this.getNext(p_153047_).ifPresent((p_153039_) -> {
-            p_153048_.setBlockAndUpdate(p_153049_, p_153039_);
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockFormEvent(p_153048_, p_153049_, p_153039_);
          });
       }
 
