--- a/net/minecraft/world/level/block/CocoaBlock.java
+++ b/net/minecraft/world/level/block/CocoaBlock.java
@@ -47,10 +_,11 @@
    }
 
    public void randomTick(BlockState p_51782_, ServerLevel p_51783_, BlockPos p_51784_, Random p_51785_) {
-      if (p_51783_.random.nextInt(5) == 0) {
+      if (true) {
          int i = p_51782_.getValue(AGE);
-         if (i < 2) {
-            p_51783_.setBlock(p_51784_, p_51782_.setValue(AGE, Integer.valueOf(i + 1)), 2);
+         if (i < 2 && net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_51783_, p_51784_, p_51782_, p_51783_.random.nextInt(Math.max(1, (int) (100.0F / p_51783_.spigotConfig.cocoaModifier) * 5)) == 0)) { // Spigot
+            org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockGrowEvent(p_51783_, p_51784_, (BlockState) p_51782_.setValue(AGE, i + 1), 2); // CraftBukkkit
+            net.minecraftforge.common.ForgeHooks.onCropsGrowPost(p_51783_, p_51784_, p_51782_);
          }
       }
 
@@ -107,7 +_,7 @@
    }
 
    public void performBonemeal(ServerLevel p_51745_, Random p_51746_, BlockPos p_51747_, BlockState p_51748_) {
-      p_51745_.setBlock(p_51747_, p_51748_.setValue(AGE, Integer.valueOf(p_51748_.getValue(AGE) + 1)), 2);
+      org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockGrowEvent(p_51745_, p_51747_, (BlockState) p_51748_.setValue(AGE, (Integer) p_51748_.getValue(AGE) + 1), 2); // CraftBukkit
    }
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> p_51778_) {
