--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -21,7 +_,7 @@
 
    public void onLand(Level p_52068_, BlockPos p_52069_, BlockState p_52070_, BlockState p_52071_, FallingBlockEntity p_52072_) {
       if (shouldSolidify(p_52068_, p_52069_, p_52071_)) {
-         p_52068_.setBlock(p_52069_, this.concrete, 3);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockFormEvent(p_52068_, p_52069_, this.concrete, 3); // CraftBukkit
       }
 
    }
@@ -30,7 +_,23 @@
       BlockGetter blockgetter = p_52063_.getLevel();
       BlockPos blockpos = p_52063_.getClickedPos();
       BlockState blockstate = blockgetter.getBlockState(blockpos);
-      return shouldSolidify(blockgetter, blockpos, blockstate) ? this.concrete : super.getStateForPlacement(p_52063_);
+      // CraftBukkit start
+      if (!shouldSolidify(blockgetter, blockpos, blockstate)) {
+         return super.getStateForPlacement(p_52063_);
+      }
+
+      // TODO: 30/08/2020 Magma We should be able to make a factory for this
+      org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockState blockState = org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockStates.getBlockState(p_52063_.getLevel(), blockpos);
+      blockState.setData(this.concrete);
+
+      org.bukkit.event.block.BlockFormEvent event = new org.bukkit.event.block.BlockFormEvent(blockState.getBlock(), blockState);
+      p_52063_.getLevel().getCraftServer().getPluginManager().callEvent(event);
+
+      if (!event.isCancelled()) {
+         return blockState.getHandle();
+      }
+      return super.getStateForPlacement(p_52063_);
+      // CraftBukkit end
    }
 
    private static boolean shouldSolidify(BlockGetter p_52081_, BlockPos p_52082_, BlockState p_52083_) {
@@ -61,7 +_,25 @@
    }
 
    public BlockState updateShape(BlockState p_52074_, Direction p_52075_, BlockState p_52076_, LevelAccessor p_52077_, BlockPos p_52078_, BlockPos p_52079_) {
-      return touchesLiquid(p_52077_, p_52078_) ? this.concrete : super.updateShape(p_52074_, p_52075_, p_52076_, p_52077_, p_52078_, p_52079_);
+
+      // CraftBukkit start
+      if (touchesLiquid(p_52077_, p_52078_)) {
+         // Suppress during worldgen
+         if (!(p_52077_ instanceof Level)) {
+            return this.concrete;
+         }
+         org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockState blockState = org.bukkit.craftbukkit.v1_18_R2.block.CraftBlockStates.getBlockState(p_52077_, p_52078_);
+         blockState.setData(this.concrete);
+
+         org.bukkit.event.block.BlockFormEvent event = new org.bukkit.event.block.BlockFormEvent(blockState.getBlock(), blockState);
+         ((Level) p_52077_).getCraftServer().getPluginManager().callEvent(event);
+
+         if (!event.isCancelled()) {
+            return blockState.getHandle();
+         }
+      }
+      return super.updateShape(p_52074_, p_52075_, p_52076_, p_52077_, p_52078_, p_52079_);
+      // CraftBukkit end
    }
 
    public int getDustColor(BlockState p_52085_, BlockGetter p_52086_, BlockPos p_52087_) {
