--- a/net/minecraft/world/level/block/CoralPlantBlock.java
+++ b/net/minecraft/world/level/block/CoralPlantBlock.java
@@ -29,6 +_,11 @@
 
    public void tick(BlockState p_52178_, ServerLevel p_52179_, BlockPos p_52180_, Random p_52181_) {
       if (!scanForWater(p_52178_, p_52179_, p_52180_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_52179_, p_52180_, this.deadBlock.defaultBlockState().setValue(CoralPlantBlock.WATERLOGGED, false)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_52179_.setBlock(p_52180_, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)), 2);
       }
 
