--- a/net/minecraft/world/level/block/CoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/CoralWallFanBlock.java
@@ -24,6 +_,11 @@
 
    public void tick(BlockState p_52205_, ServerLevel p_52206_, BlockPos p_52207_, Random p_52208_) {
       if (!scanForWater(p_52205_, p_52206_, p_52207_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_52206_, p_52207_, this.deadBlock.defaultBlockState().setValue(CoralWallFanBlock.WATERLOGGED, false).setValue(CoralWallFanBlock.FACING, p_52205_.getValue(CoralWallFanBlock.FACING))).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_52206_.setBlock(p_52207_, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)).setValue(FACING, p_52205_.getValue(FACING)), 2);
       }
 
