--- a/net/minecraft/world/level/block/CoralFanBlock.java
+++ b/net/minecraft/world/level/block/CoralFanBlock.java
@@ -24,6 +_,11 @@
 
    public void tick(BlockState p_52154_, ServerLevel p_52155_, BlockPos p_52156_, Random p_52157_) {
       if (!scanForWater(p_52154_, p_52155_, p_52156_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_52155_, p_52156_, this.deadBlock.defaultBlockState().setValue(CoralFanBlock.WATERLOGGED, false)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_52155_.setBlock(p_52156_, this.deadBlock.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)), 2);
       }
 
