--- a/net/minecraft/world/level/block/FungusBlock.java
+++ b/net/minecraft/world/level/block/FungusBlock.java
@@ -44,6 +_,16 @@
    }
 
    public void performBonemeal(ServerLevel p_53603_, Random p_53604_, BlockPos p_53605_, BlockState p_53606_) {
+      // CraftBukkit start
+      if (this == Blocks.WARPED_FUNGUS) {
+         SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+      } else if (this == Blocks.CRIMSON_FUNGUS) {
+         SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+      }
+      // CraftBukkit end
       this.feature.get().value().place(p_53603_, p_53603_.getChunkSource().getGenerator(), p_53604_, p_53605_);
+      net.minecraftforge.event.world.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_53603_, p_53604_, p_53605_, this.feature.get());
+      if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY)) return;
+      event.getFeature().value().place(p_53603_, p_53603_.getChunkSource().getGenerator(), p_53604_, p_53605_);
    }
 }
