--- a/net/minecraft/world/level/block/InfestedBlock.java
+++ b/net/minecraft/world/level/block/InfestedBlock.java
@@ -40,6 +_,7 @@
    private void spawnInfestation(ServerLevel p_54181_, BlockPos p_54182_) {
       Silverfish silverfish = EntityType.SILVERFISH.create(p_54181_);
       silverfish.moveTo((double)p_54182_.getX() + 0.5D, (double)p_54182_.getY(), (double)p_54182_.getZ() + 0.5D, 0.0F, 0.0F);
+      p_54181_.prepareEntity(silverfish, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SILVERFISH_BLOCK); //Magma - add spawn reason
       p_54181_.addFreshEntity(silverfish);
       silverfish.spawnAnim();
    }
