--- a/net/minecraft/world/level/block/JukeboxBlock.java
+++ b/net/minecraft/world/level/block/JukeboxBlock.java
@@ -54,7 +_,13 @@
    public void setRecord(LevelAccessor p_54270_, BlockPos p_54271_, BlockState p_54272_, ItemStack p_54273_) {
       BlockEntity blockentity = p_54270_.getBlockEntity(p_54271_);
       if (blockentity instanceof JukeboxBlockEntity) {
-         ((JukeboxBlockEntity)blockentity).setRecord(p_54273_.copy());
+         // CraftBukkit start - There can only be one
+         p_54273_ = p_54273_.copy();
+         if (!p_54273_.isEmpty()) {
+            p_54273_.setCount(1);
+         }
+         ((JukeboxBlockEntity) blockentity).setRecord(p_54273_);
+         // CraftBukkit end
          p_54270_.setBlock(p_54271_, p_54272_.setValue(HAS_RECORD, Boolean.valueOf(true)), 2);
       }
    }
