--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -19,7 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class LeavesBlock extends Block {
+public class LeavesBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final int DECAY_DISTANCE = 7;
    public static final IntegerProperty DISTANCE = BlockStateProperties.DISTANCE;
    public static final BooleanProperty PERSISTENT = BlockStateProperties.PERSISTENT;
@@ -40,6 +_,14 @@
 
    public void randomTick(BlockState p_54451_, ServerLevel p_54452_, BlockPos p_54453_, Random p_54454_) {
       if (!p_54451_.getValue(PERSISTENT) && p_54451_.getValue(DISTANCE) == 7) {
+         // CraftBukkit start
+         org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(p_54452_.getWorld().getBlockAt(p_54453_.getX(), p_54453_.getY(), p_54453_.getZ()));
+         p_54452_.getCraftServer().getPluginManager().callEvent(event);
+
+         if (event.isCancelled() || p_54452_.getBlockState(p_54453_).getBlock() != this) {
+            return;
+         }
+         // CraftBukkit end
          dropResources(p_54451_, p_54452_, p_54453_);
          p_54452_.removeBlock(p_54453_, false);
       }
