--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -43,7 +_,9 @@
       boolean flag = p_55042_.hasNeighborSignal(p_55043_);
       if (flag != p_55041_.getValue(POWERED)) {
          if (flag) {
+            this.prepareNoteData(p_55041_);
             this.playNote(p_55042_, p_55043_);
+            p_55041_ = p_55042_.getBlockState(p_55043_); // CraftBukkit - SPIGOT-5617: update in case changed in event
          }
 
          p_55042_.setBlock(p_55043_, p_55041_.setValue(POWERED, Boolean.valueOf(flag)), 3);
@@ -51,19 +_,37 @@
 
    }
 
+   //Magma - start
+   private @org.jetbrains.annotations.Nullable BlockState playNote_data;
+   private void prepareNoteData(BlockState p_55041_) {
+      this.playNote_data = p_55041_;
+   }
+
    private void playNote(Level p_55020_, BlockPos p_55021_) {
+      if (playNote_data == null)
+         throw new IllegalStateException("playNote_data is null");
+
       if (p_55020_.getBlockState(p_55021_.above()).isAir()) {
-         p_55020_.blockEvent(p_55021_, this, 0, 0);
+         // CraftBukkit start
+         org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callNotePlayEvent(p_55020_, p_55021_, playNote_data.getValue(INSTRUMENT), playNote_data.getValue(NOTE));
+         if (!event.isCancelled()) {
+            p_55020_.blockEvent(p_55021_, this, 0, 0);
+         }
+         // CraftBukkit end
       }
-
+      playNote_data = null;
    }
+   //Magma - end
 
    public InteractionResult use(BlockState p_55034_, Level p_55035_, BlockPos p_55036_, Player p_55037_, InteractionHand p_55038_, BlockHitResult p_55039_) {
       if (p_55035_.isClientSide) {
          return InteractionResult.SUCCESS;
       } else {
-         p_55034_ = p_55034_.cycle(NOTE);
+         int _new = net.minecraftforge.common.ForgeHooks.onNoteChange(p_55035_, p_55036_, p_55034_, p_55034_.getValue(NOTE), p_55034_.cycle(NOTE).getValue(NOTE));
+         if (_new == -1) return InteractionResult.FAIL;
+         p_55034_ = p_55034_.setValue(NOTE, _new);
          p_55035_.setBlock(p_55036_, p_55034_, 3);
+         this.prepareNoteData(p_55034_);
          this.playNote(p_55035_, p_55036_);
          p_55037_.awardStat(Stats.TUNE_NOTEBLOCK);
          return InteractionResult.CONSUME;
@@ -72,12 +_,16 @@
 
    public void attack(BlockState p_55029_, Level p_55030_, BlockPos p_55031_, Player p_55032_) {
       if (!p_55030_.isClientSide) {
+         this.prepareNoteData(p_55029_);
          this.playNote(p_55030_, p_55031_);
          p_55032_.awardStat(Stats.PLAY_NOTEBLOCK);
       }
    }
 
    public boolean triggerEvent(BlockState p_55023_, Level p_55024_, BlockPos p_55025_, int p_55026_, int p_55027_) {
+      net.minecraftforge.event.world.NoteBlockEvent.Play e = new net.minecraftforge.event.world.NoteBlockEvent.Play(p_55024_, p_55025_, p_55023_, p_55023_.getValue(NOTE), p_55023_.getValue(INSTRUMENT));
+      if (net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(e)) return false;
+      p_55023_ = p_55023_.setValue(NOTE, e.getVanillaNoteId()).setValue(INSTRUMENT, e.getInstrument());
       int i = p_55023_.getValue(NOTE);
       float f = (float)Math.pow(2.0D, (double)(i - 12) / 12.0D);
       p_55024_.playSound((Player)null, p_55025_, p_55023_.getValue(INSTRUMENT).getSoundEvent(), SoundSource.RECORDS, 3.0F, f);
