--- a/net/minecraft/world/level/block/NyliumBlock.java
+++ b/net/minecraft/world/level/block/NyliumBlock.java
@@ -27,6 +_,11 @@
 
    public void randomTick(BlockState p_55074_, ServerLevel p_55075_, BlockPos p_55076_, Random p_55077_) {
       if (!canBeNylium(p_55074_, p_55075_, p_55076_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_55075_, p_55076_, Blocks.NETHERRACK.defaultBlockState()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_55075_.setBlockAndUpdate(p_55076_, Blocks.NETHERRACK.defaultBlockState());
       }
 
