--- a/net/minecraft/world/level/block/OreBlock.java
+++ b/net/minecraft/world/level/block/OreBlock.java
@@ -23,12 +_,10 @@
 
    public void spawnAfterBreak(BlockState p_55142_, ServerLevel p_55143_, BlockPos p_55144_, ItemStack p_55145_) {
       super.spawnAfterBreak(p_55142_, p_55143_, p_55144_, p_55145_);
-      if (EnchantmentHelper.getItemEnchantmentLevel(Enchantments.SILK_TOUCH, p_55145_) == 0) {
-         int i = this.xpRange.sample(p_55143_.random);
-         if (i > 0) {
-            this.popExperience(p_55143_, p_55144_, i);
-         }
-      }
+   }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader reader, BlockPos pos, int fortune, int silktouch) {
+      return silktouch == 0 ? this.xpRange.sample(RANDOM) : 0;
    }
 }
