--- a/net/minecraft/world/level/block/PressurePlateBlock.java
+++ b/net/minecraft/world/level/block/PressurePlateBlock.java
@@ -69,6 +_,27 @@
 
       if (!list.isEmpty()) {
          for(Entity entity : list) {
+
+            // CraftBukkit start - Call interact event when turning on a pressure plate
+            if (this.getSignalForState(p_55264_.getBlockState(p_55265_)) == 0) {
+               org.bukkit.World bworld = p_55264_.getWorld();
+               org.bukkit.plugin.PluginManager manager = p_55264_.getCraftServer().getPluginManager();
+               org.bukkit.event.Cancellable cancellable;
+               
+               if (entity instanceof Player) {
+                  cancellable = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, p_55265_, null, null, null);
+               } else {
+                  cancellable = new org.bukkit.event.entity.EntityInteractEvent(entity.getBukkitEntity(), bworld.getBlockAt(p_55265_.getX(), p_55265_.getY(), p_55265_.getZ()));
+                  manager.callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+               }
+               
+               // We only want to block turning the plate on if all events are cancelled
+               if (cancellable.isCancelled()) {
+                  continue;
+               }
+            }
+            // CraftBukkit end
+            
             if (!entity.isIgnoringBlockTriggers()) {
                return 15;
             }
