--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -62,8 +_,24 @@
          list.remove(0);
       }
 
+      // CraftBukkit start
+      org.bukkit.plugin.PluginManager manager = p_55690_.getCraftServer().getPluginManager();
+      org.bukkit.block.Block block = p_55690_.getWorld().getBlockAt(p_55691_.getX(), p_55691_.getY(), p_55691_.getZ());
+      int oldCurrent = ((Boolean) p_55689_.getValue(LIT)).booleanValue() ? 15 : 0;
+
+      org.bukkit.event.block.BlockRedstoneEvent event = new org.bukkit.event.block.BlockRedstoneEvent(block, oldCurrent, oldCurrent);
+      // CraftBukkit end
       if (p_55689_.getValue(LIT)) {
          if (flag) {
+            // CraftBukkit start
+            if (oldCurrent != 0) {
+               event.setNewCurrent(0);
+               manager.callEvent(event);
+               if (event.getNewCurrent() != 0) {
+                  return;
+               }
+            }
+            // CraftBukkit end
             p_55690_.setBlock(p_55691_, p_55689_.setValue(LIT, Boolean.valueOf(false)), 3);
             if (isToggledTooFrequently(p_55690_, p_55691_, true)) {
                p_55690_.levelEvent(1502, p_55691_, 0);
@@ -71,6 +_,15 @@
             }
          }
       } else if (!flag && !isToggledTooFrequently(p_55690_, p_55691_, false)) {
+         // CraftBukkit start
+         if (oldCurrent != 15) {
+            event.setNewCurrent(15);
+            manager.callEvent(event);
+            if (event.getNewCurrent() != 15) {
+               return;
+            }
+         }
+         // CraftBukkit end
          p_55690_.setBlock(p_55691_, p_55689_.setValue(LIT, Boolean.valueOf(true)), 3);
       }
 
