--- a/net/minecraft/world/level/block/ScaffoldingBlock.java
+++ b/net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -88,7 +_,7 @@
       int i = getDistance(p_56033_, p_56034_);
       BlockState blockstate = p_56032_.setValue(DISTANCE, Integer.valueOf(i)).setValue(BOTTOM, Boolean.valueOf(this.isBottom(p_56033_, p_56034_, i)));
       if (blockstate.getValue(DISTANCE) == 7) {
-         if (p_56032_.getValue(DISTANCE) == 7) {
+         if (p_56032_.getValue(DISTANCE) == 7 && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_56033_, p_56034_, Blocks.AIR.defaultBlockState()).isCancelled()) { // CraftBukkit - BlockFadeEvent
             FallingBlockEntity.fall(p_56033_, p_56034_, blockstate);
          } else {
             p_56033_.destroyBlock(p_56034_, true);
