--- a/net/minecraft/world/level/block/SpawnerBlock.java
+++ b/net/minecraft/world/level/block/SpawnerBlock.java
@@ -29,8 +_,11 @@
 
    public void spawnAfterBreak(BlockState p_56789_, ServerLevel p_56790_, BlockPos p_56791_, ItemStack p_56792_) {
       super.spawnAfterBreak(p_56789_, p_56790_, p_56791_, p_56792_);
-      int i = 15 + p_56790_.random.nextInt(15) + p_56790_.random.nextInt(15);
-      this.popExperience(p_56790_, p_56791_, i);
+   }
+
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.level.LevelReader world, BlockPos pos, int fortune, int silktouch) {
+      return 15 + RANDOM.nextInt(15) + RANDOM.nextInt(15);
    }
 
    public RenderShape getRenderShape(BlockState p_56794_) {
