--- a/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -35,15 +_,22 @@
 
    public void randomTick(BlockState p_56819_, ServerLevel p_56820_, BlockPos p_56821_, Random p_56822_) {
       if (!canBeGrass(p_56819_, p_56820_, p_56821_)) {
+         // CraftBukkit start
+         if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_56820_, p_56821_, Blocks.DIRT.defaultBlockState()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
+         if (!p_56820_.isAreaLoaded(p_56821_, 1)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
          p_56820_.setBlockAndUpdate(p_56821_, Blocks.DIRT.defaultBlockState());
       } else {
+         if (!p_56820_.isAreaLoaded(p_56821_, 3)) return; // Forge: prevent loading unloaded chunks when checking neighbor's light and spreading
          if (p_56820_.getMaxLocalRawBrightness(p_56821_.above()) >= 9) {
             BlockState blockstate = this.defaultBlockState();
 
             for(int i = 0; i < 4; ++i) {
                BlockPos blockpos = p_56821_.offset(p_56822_.nextInt(3) - 1, p_56822_.nextInt(5) - 3, p_56822_.nextInt(3) - 1);
                if (p_56820_.getBlockState(blockpos).is(Blocks.DIRT) && canPropagate(blockstate, p_56820_, blockpos)) {
-                  p_56820_.setBlockAndUpdate(blockpos, blockstate.setValue(SNOWY, Boolean.valueOf(p_56820_.getBlockState(blockpos.above()).is(Blocks.SNOW))));
+                  org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockSpreadEvent(p_56820_, p_56821_, blockpos, blockstate.setValue(SNOWY, Boolean.valueOf(p_56820_.getBlockState(blockpos.above()).is(Blocks.SNOW)))); // CraftBukkit
                }
             }
          }
