--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -18,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SugarCaneBlock extends Block {
+public class SugarCaneBlock extends Block implements net.minecraftforge.common.IPlantable {
    public static final IntegerProperty AGE = BlockStateProperties.AGE_15;
    protected static final float AABB_OFFSET = 6.0F;
    protected static final VoxelShape SHAPE = Block.box(2.0D, 0.0D, 2.0D, 14.0D, 16.0D, 14.0D);
@@ -47,12 +_,14 @@
 
          if (i < 3) {
             int j = p_57188_.getValue(AGE);
-            if (j == 15) {
-               p_57189_.setBlockAndUpdate(p_57190_.above(), this.defaultBlockState());
+            if (net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_57189_, p_57190_, p_57188_, true)) {
+            if (j >= (byte) range(3, ((100.0F / p_57189_.spigotConfig.caneModifier) * 15) + 0.5F, 15)) { // Spigot
+               org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockGrowEvent(p_57189_, p_57190_.above(), this.defaultBlockState());
                p_57189_.setBlock(p_57190_, p_57188_.setValue(AGE, Integer.valueOf(0)), 4);
             } else {
                p_57189_.setBlock(p_57190_, p_57188_.setValue(AGE, Integer.valueOf(j + 1)), 4);
             }
+            }
          }
       }
 
@@ -67,6 +_,8 @@
    }
 
    public boolean canSurvive(BlockState p_57175_, LevelReader p_57176_, BlockPos p_57177_) {
+      BlockState soil = p_57176_.getBlockState(p_57177_.below());
+      if (soil.canSustainPlant(p_57176_, p_57177_.below(), Direction.UP, this)) return true;
       BlockState blockstate = p_57176_.getBlockState(p_57177_.below());
       if (blockstate.is(this)) {
          return true;
@@ -89,5 +_,15 @@
 
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> p_57186_) {
       p_57186_.add(AGE);
+   }
+
+   @Override
+   public net.minecraftforge.common.PlantType getPlantType(BlockGetter world, BlockPos pos) {
+       return net.minecraftforge.common.PlantType.BEACH;
+   }
+
+   @Override
+   public BlockState getPlant(BlockGetter world, BlockPos pos) {
+      return defaultBlockState();
    }
 }
