--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -58,6 +_,19 @@
    private void destroyEgg(Level p_154851_, BlockState p_154852_, BlockPos p_154853_, Entity p_154854_, int p_154855_) {
       if (this.canDestroyEgg(p_154851_, p_154854_)) {
          if (!p_154851_.isClientSide && p_154851_.random.nextInt(p_154855_) == 0 && p_154852_.is(Blocks.TURTLE_EGG)) {
+            // CraftBukkit start - Step on eggs
+            org.bukkit.event.Cancellable cancellable;
+            if (p_154854_ instanceof Player) {
+               cancellable = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPlayerInteractEvent((Player) p_154854_, org.bukkit.event.block.Action.PHYSICAL, p_154853_, null, null, null);
+            } else {
+               cancellable = new org.bukkit.event.entity.EntityInteractEvent(p_154854_.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_154851_, p_154853_));
+               p_154851_.getCraftServer().getPluginManager().callEvent((org.bukkit.event.entity.EntityInteractEvent) cancellable);
+            }
+
+            if (cancellable.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             this.decreaseEggs(p_154851_, p_154853_, p_154852_);
          }
 
@@ -80,9 +_,19 @@
       if (this.shouldUpdateHatchLevel(p_57805_) && onSand(p_57805_, p_57806_)) {
          int i = p_57804_.getValue(HATCH);
          if (i < 2) {
+            // CraftBukkit start - Call BlockGrowEvent
+            if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.handleBlockGrowEvent(p_57805_, p_57806_, p_57804_.setValue(HATCH, i + 1), 2)) {
+               return;
+            }
+            // CraftBukkit end
             p_57805_.playSound((Player)null, p_57806_, SoundEvents.TURTLE_EGG_CRACK, SoundSource.BLOCKS, 0.7F, 0.9F + p_57807_.nextFloat() * 0.2F);
-            p_57805_.setBlock(p_57806_, p_57804_.setValue(HATCH, Integer.valueOf(i + 1)), 2);
+            // p_57805_.setBlock(p_57806_, p_57804_.setValue(HATCH, Integer.valueOf(i + 1)), 2); // CraftBukkit - handled above
          } else {
+            // CraftBukkit start - Call BlockFadeEvent
+            if (org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callBlockFadeEvent(p_57805_, p_57806_, Blocks.AIR.defaultBlockState()).isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             p_57805_.playSound((Player)null, p_57806_, SoundEvents.TURTLE_EGG_HATCH, SoundSource.BLOCKS, 0.7F, 0.9F + p_57807_.nextFloat() * 0.2F);
             p_57805_.removeBlock(p_57806_, false);
 
@@ -92,6 +_,7 @@
                turtle.setAge(-24000);
                turtle.setHomePos(p_57806_);
                turtle.moveTo((double)p_57806_.getX() + 0.3D + (double)j * 0.2D, (double)p_57806_.getY(), (double)p_57806_.getZ() + 0.3D, 0.0F, 0.0F);
+               p_57805_.prepareEntity(turtle, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); //Magma - add spawn reason
                p_57805_.addFreshEntity(turtle);
             }
          }
@@ -151,7 +_,7 @@
          if (!(p_57769_ instanceof LivingEntity)) {
             return false;
          } else {
-            return p_57769_ instanceof Player || p_57768_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+            return p_57769_ instanceof Player || net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_57768_, p_57769_);
          }
       } else {
          return false;
