--- a/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
+++ b/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
@@ -33,7 +_,13 @@
             BlockPos blockpos = new BlockPos(p_114634_.getX(), p_114634_.getBoundingBox().maxY, p_114634_.getZ());
             p_114637_.translate(-0.5D, 0.0D, -0.5D);
             BlockRenderDispatcher blockrenderdispatcher = Minecraft.getInstance().getBlockRenderer();
-            blockrenderdispatcher.getModelRenderer().tesselateBlock(level, blockrenderdispatcher.getBlockModel(blockstate), blockstate, blockpos, p_114637_, p_114638_.getBuffer(ItemBlockRenderTypes.getMovingBlockRenderType(blockstate)), false, new Random(), blockstate.getSeed(p_114634_.getStartPos()), OverlayTexture.NO_OVERLAY);
+            for (net.minecraft.client.renderer.RenderType type : net.minecraft.client.renderer.RenderType.chunkBufferLayers()) {
+               if (ItemBlockRenderTypes.canRenderInLayer(blockstate, type)) {
+                  net.minecraftforge.client.ForgeHooksClient.setRenderType(type);
+                  blockrenderdispatcher.getModelRenderer().tesselateBlock(level, blockrenderdispatcher.getBlockModel(blockstate), blockstate, blockpos, p_114637_, p_114638_.getBuffer(type), false, new Random(), blockstate.getSeed(p_114634_.getStartPos()), OverlayTexture.NO_OVERLAY);
+               }
+            }
+            net.minecraftforge.client.ForgeHooksClient.setRenderType(null);
             p_114637_.popPose();
             super.render(p_114634_, p_114635_, p_114636_, p_114637_, p_114638_, p_114639_);
          }
