--- a/net/minecraft/client/renderer/entity/ItemRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemRenderer.java
@@ -76,8 +_,8 @@
 
    public ItemRenderer(TextureManager p_174225_, ModelManager p_174226_, ItemColors p_174227_, BlockEntityWithoutLevelRenderer p_174228_) {
       this.textureManager = p_174225_;
-      this.itemModelShaper = new ItemModelShaper(p_174226_);
       this.blockEntityRenderer = p_174228_;
+      this.itemModelShaper = new net.minecraftforge.client.ItemModelMesherForge(p_174226_);
 
       for(Item item : Registry.ITEM) {
          if (!IGNORED.contains(item)) {
@@ -117,7 +_,7 @@
             }
          }
 
-         p_115151_.getTransforms().getTransform(p_115145_).apply(p_115146_, p_115147_);
+         p_115151_ = net.minecraftforge.client.ForgeHooksClient.handleCameraTransforms(p_115147_, p_115151_, p_115145_, p_115146_);
          p_115147_.translate(-0.5D, -0.5D, -0.5D);
          if (!p_115151_.isCustomRenderer() && (!p_115144_.is(Items.TRIDENT) || flag)) {
             boolean flag1;
@@ -127,7 +_,8 @@
             } else {
                flag1 = true;
             }
-
+            if (p_115151_.isLayered()) { net.minecraftforge.client.ForgeHooksClient.drawItemLayered(this, p_115151_, p_115144_, p_115147_, p_115148_, p_115149_, p_115150_, flag1); }
+            else {
             RenderType rendertype = ItemBlockRenderTypes.getRenderType(p_115144_, flag1);
             VertexConsumer vertexconsumer;
             if (p_115144_.is(Items.COMPASS) && p_115144_.hasFoil()) {
@@ -153,8 +_,9 @@
             }
 
             this.renderModelLists(p_115151_, p_115144_, p_115149_, p_115150_, p_115147_, vertexconsumer);
+            }
          } else {
-            this.blockEntityRenderer.renderByItem(p_115144_, p_115145_, p_115147_, p_115148_, p_115149_, p_115150_);
+            net.minecraftforge.client.RenderProperties.get(p_115144_).getItemStackRenderer().renderByItem(p_115144_, p_115145_, p_115147_, p_115148_, p_115149_, p_115150_);
          }
 
          p_115147_.popPose();
@@ -198,7 +_,7 @@
          float f = (float)(i >> 16 & 255) / 255.0F;
          float f1 = (float)(i >> 8 & 255) / 255.0F;
          float f2 = (float)(i & 255) / 255.0F;
-         p_115164_.putBulkData(posestack$pose, bakedquad, f, f1, f2, p_115167_, p_115168_);
+         p_115164_.putBulkData(posestack$pose, bakedquad, f, f1, f2, p_115167_, p_115168_, true);
       }
 
    }
@@ -301,6 +_,7 @@
             crashreportcategory.setDetail("Item Type", () -> {
                return String.valueOf((Object)p_174237_.getItem());
             });
+            crashreportcategory.setDetail("Registry Name", () -> String.valueOf(p_174237_.getItem().getRegistryName()));
             crashreportcategory.setDetail("Item Damage", () -> {
                return String.valueOf(p_174237_.getDamageValue());
             });
@@ -377,5 +_,9 @@
 
    public void onResourceManagerReload(ResourceManager p_115105_) {
       this.itemModelShaper.rebuildCache();
+   }
+
+   public BlockEntityWithoutLevelRenderer getBlockEntityRenderer() {
+       return blockEntityRenderer;
    }
 }
