--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -18,6 +_,8 @@
    public double f_36813_;
    public double f_36814_;
    public double f_36815_;
+   public float bukkitYield = 1; // CraftBukkit
+   public boolean isIncendiary = true; // CraftBukkit
 
    protected AbstractHurtingProjectile(EntityType<? extends AbstractHurtingProjectile> p_36833_, Level p_36834_) {
       super(p_36833_, p_36834_);
@@ -42,6 +_,17 @@
       this.m_19915_(p_36827_.m_146908_(), p_36827_.m_146909_());
    }
 
+   // CraftBukkit start
+   public void setDirection(double p_36821_, double p_36822_, double p_36823_) {
+      double d0 = Math.sqrt(p_36821_ * p_36821_ + p_36822_ * p_36822_ + p_36823_ * p_36823_);
+      if (d0 != 0.0D) {
+         this.f_36813_ = p_36821_ / d0 * 0.1D;
+         this.f_36814_ = p_36822_ / d0 * 0.1D;
+         this.f_36815_ = p_36823_ / d0 * 0.1D;
+      }
+   }
+   // CraftBukkit end
+
    protected void m_8097_() {
    }
 
@@ -64,7 +_,7 @@
          }
 
          HitResult hitresult = ProjectileUtil.m_37294_(this, this::m_5603_);
-         if (hitresult.m_6662_() != HitResult.Type.MISS) {
+         if (hitresult.m_6662_() != HitResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
             this.m_6532_(hitresult);
          }
 
