--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -64,6 +_,16 @@
    @Nullable
    private List<Entity> piercedAndKilledEntities;
 
+   // Spigot Start
+   @Override
+   public void inactiveTick() {
+      if (this.inGround) {
+         this.life += 1;
+      }
+      super.inactiveTick();
+   }
+   // Spigot End
+
    protected AbstractArrow(EntityType<? extends AbstractArrow> p_36721_, Level p_36722_) {
       super(p_36721_, p_36722_);
    }
@@ -184,8 +_,8 @@
                }
             }
 
-            if (hitresult != null && !flag) {
-               this.onHit(hitresult);
+            if (hitresult != null && hitresult.getType() != HitResult.Type.MISS && !flag && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, hitresult)) {
+               this.preOnHit(hitresult); // CraftBukkit - projectile hit event
                this.hasImpulse = true;
             }
 
@@ -262,7 +_,7 @@
 
    protected void tickDespawn() {
       ++this.life;
-      if (this.life >= 1200) {
+      if (this.life >= ((this instanceof ThrownTrident) ? level.spigotConfig.tridentDespawnRate : level.spigotConfig.arrowDespawnRate)) { // Spigot
          this.discard();
       }
 
@@ -320,7 +_,13 @@
       boolean flag = entity.getType() == EntityType.ENDERMAN;
       int k = entity.getRemainingFireTicks();
       if (this.isOnFire() && !flag) {
-         entity.setSecondsOnFire(5);
+         // CraftBukkit start
+         org.bukkit.event.entity.EntityCombustByEntityEvent combustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+         org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+         if (!combustEvent.isCancelled()) {
+            entity.setSecondsOnFire(combustEvent.getDuration(), false);
+         }
+         // CraftBukkit end
       }
 
       if (entity.hurt(damagesource, (float)i)) {
@@ -472,7 +_,21 @@
 
    public void playerTouch(Player p_36766_) {
       if (!this.level.isClientSide && (this.inGround || this.isNoPhysics()) && this.shakeTime <= 0) {
-         if (this.tryPickup(p_36766_)) {
+         // CraftBukkit start
+         ItemStack itemstack = this.getPickupItem();
+         if (this.pickup == Pickup.ALLOWED && !itemstack.isEmpty() && p_36766_.getInventory().canHold(itemstack) > 0) {
+            net.minecraft.world.entity.item.ItemEntity item = new net.minecraft.world.entity.item.ItemEntity(this.level, this.getX(), this.getY(), this.getZ(), itemstack);
+            org.bukkit.event.player.PlayerPickupArrowEvent event = new org.bukkit.event.player.PlayerPickupArrowEvent((org.bukkit.entity.Player) p_36766_.getBukkitEntity(), new org.bukkit.craftbukkit.v1_18_R2.entity.CraftItem(this.level.getCraftServer(), this, item), (org.bukkit.entity.AbstractArrow) this.getBukkitEntity());
+            // event.setCancelled(!p_36766_.canPickUpLoot); TODO
+            this.level.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+               return;
+            }
+            itemstack = item.getItem();
+         }
+         if ((this.pickup == AbstractArrow.Pickup.ALLOWED && p_36766_.getInventory().add(itemstack)) || (this.pickup == AbstractArrow.Pickup.CREATIVE_ONLY && p_36766_.getAbilities().instabuild)) {
+            // CraftBukkit end
             p_36766_.take(this, 1);
             this.discard();
          }
