--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -124,6 +_,25 @@
       }
    }
 
+   // CraftBukkit start accessor methods
+   public void refreshEffects() {
+      this.getEntityData().set(Arrow.ID_EFFECT_COLOR, PotionUtils.getColor((Collection) PotionUtils.getAllEffects(this.potion, (Collection) this.effects)));
+   }
+
+   public String getPotionType() {
+      return Registry.POTION.getKey(this.potion).toString();
+   }
+
+   public void setPotionType(String string) {
+      this.potion = Registry.POTION.get(new net.minecraft.resources.ResourceLocation(string));
+      this.getEntityData().set(Arrow.ID_EFFECT_COLOR, PotionUtils.getColor((Collection) PotionUtils.getAllEffects(this.potion, (Collection) this.effects)));
+   }
+
+   public boolean isTipped() {
+      return !(this.effects.isEmpty() && this.potion == Potions.EMPTY);
+   }
+   // CraftBukkit end
+
    public int getColor() {
       return this.entityData.get(ID_EFFECT_COLOR);
    }
@@ -178,12 +_,14 @@
       Entity entity = this.getEffectSource();
 
       for(MobEffectInstance mobeffectinstance : this.potion.getEffects()) {
-         p_36873_.addEffect(new MobEffectInstance(mobeffectinstance.getEffect(), Math.max(mobeffectinstance.getDuration() / 8, 1), mobeffectinstance.getAmplifier(), mobeffectinstance.isAmbient(), mobeffectinstance.isVisible()), entity);
+         p_36873_.prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW);
+         p_36873_.addEffect(new MobEffectInstance(mobeffectinstance.getEffect(), Math.max(mobeffectinstance.getDuration() / 8, 1), mobeffectinstance.getAmplifier(), mobeffectinstance.isAmbient(), mobeffectinstance.isVisible()), entity); // CraftBukkit
       }
 
       if (!this.effects.isEmpty()) {
          for(MobEffectInstance mobeffectinstance1 : this.effects) {
-            p_36873_.addEffect(mobeffectinstance1, entity);
+            p_36873_.prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW);
+            p_36873_.addEffect(mobeffectinstance1, entity); // CraftBukkit
          }
       }
 
