--- a/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -40,22 +_,72 @@
       super.onHitEntity(p_37486_);
       p_37486_.getEntity().hurt(DamageSource.thrown(this, this.getOwner()), 0.0F);
    }
+   
+   // Magma - Start
+   private boolean hatching;
+   private org.bukkit.entity.EntityType hatchingType;
+   private net.minecraft.world.entity.Entity shooter;
+   private org.bukkit.event.player.PlayerEggThrowEvent event;
+   private java.util.concurrent.atomic.AtomicReference<net.minecraft.world.entity.Entity> craftBukkitEntity = new java.util.concurrent.atomic.AtomicReference<>();
+   // Magma end
 
    protected void onHit(HitResult p_37488_) {
       super.onHit(p_37488_);
       if (!this.level.isClientSide) {
-         if (this.random.nextInt(8) == 0) {
+         // CraftBukkit start
+         hatching = this.random.nextInt(8) == 0;
+         if (true) {
+            // CraftBukkit end
             int i = 1;
             if (this.random.nextInt(32) == 0) {
                i = 4;
             }
 
-            for(int j = 0; j < i; ++j) {
+            // CraftBukkit start
+            hatchingType = org.bukkit.entity.EntityType.CHICKEN;
+
+            shooter = this.getOwner();
+            if (!hatching) {
+               i = 0;
+            }
+
+            if (shooter instanceof net.minecraft.server.level.ServerPlayer) {
+               event = new org.bukkit.event.player.PlayerEggThrowEvent((org.bukkit.entity.Player) shooter.getBukkitEntity(), (org.bukkit.entity.Egg) this.getBukkitEntity(), hatching, (byte) i, hatchingType);
+               this.level.getCraftServer().getPluginManager().callEvent(event);
+
+               i = event.getNumHatches();
+               hatching = event.isHatching();
+               hatchingType = event.getHatchingType();
+               // If hatching is set to false, ensure child count is 0
+               if (!hatching) {
+                  i = 0;
+               }
+            }
+            // CraftBukkit end
+
+            for (int j = 0; j < i; ++j) {
                Chicken chicken = EntityType.CHICKEN.create(this.level);
-               chicken.setAge(-24000);
-               chicken.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F);
-               this.level.addFreshEntity(chicken);
-            }
+
+               // Magma start - Fix mixin problems
+               if (hatchingType == org.bukkit.entity.EntityType.CHICKEN) {
+                  craftBukkitEntity.set(chicken);
+               } else {
+                  craftBukkitEntity.set(this.level.getWorld().createEntity(new org.bukkit.Location(this.level.getWorld(), this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F), hatchingType.getEntityClass())); // CraftBukkit
+               }
+
+               if (craftBukkitEntity.get() != null) {
+                  // CraftBukkit start
+                  if (craftBukkitEntity.get().getBukkitEntity() instanceof org.bukkit.entity.Ageable) {
+                     ((org.bukkit.entity.Ageable) craftBukkitEntity.get().getBukkitEntity()).setBaby();
+                  }
+                  
+                  craftBukkitEntity.get().moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F);
+                  this.level.prepareAddEntity(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); // Magma - Add SpawnReason
+                  this.level.addFreshEntity(craftBukkitEntity.get());
+               }
+            } 
+
+            // CraftBukkit end
          }
 
          this.level.broadcastEntityEvent(this, (byte)3);
