--- a/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
@@ -29,13 +_,19 @@
    public void setItem(ItemStack p_37447_) {
       if (!p_37447_.is(this.getDefaultItem()) || p_37447_.hasTag()) {
          this.getEntityData().set(DATA_ITEM_STACK, Util.make(p_37447_.copy(), (p_37451_) -> {
-            p_37451_.setCount(1);
+            if (!p_37447_.isEmpty()) p_37447_.setCount(1); // CraftBukkit
          }));
       }
 
    }
 
    protected abstract Item getDefaultItem();
+
+   // CraftBukkit start
+   public Item getDefaultItemPublic() {
+      return getDefaultItem();
+   }
+   // CraftBukkit end
 
    public ItemStack getItemRaw() {
       return this.getEntityData().get(DATA_ITEM_STACK);
