--- a/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -41,7 +_,7 @@
    }
 
    public float getBlockExplosionResistance(Explosion p_37619_, BlockGetter p_37620_, BlockPos p_37621_, BlockState p_37622_, FluidState p_37623_, float p_37624_) {
-      return this.isDangerous() && WitherBoss.canDestroy(p_37622_) ? Math.min(0.8F, p_37624_) : p_37624_;
+      return this.isDangerous() && p_37622_.canEntityDestroy(p_37620_, p_37621_, this) ? Math.min(0.8F, p_37624_) : p_37624_;
    }
 
    protected void onHitEntity(EntityHitResult p_37626_) {
@@ -57,6 +_,7 @@
                if (entity.isAlive()) {
                   this.doEnchantDamageEffects(livingentity, entity);
                } else {
+                  livingentity.setRegainReason(org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.WITHER);
                   livingentity.heal(5.0F);
                }
             }
@@ -73,7 +_,8 @@
             }
 
             if (i > 0) {
-               ((LivingEntity)entity).addEffect(new MobEffectInstance(MobEffects.WITHER, 20 * i, 1), this.getEffectSource());
+               ((LivingEntity)entity).prepare(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); //Magma
+               ((LivingEntity)entity).addEffect(new MobEffectInstance(MobEffects.WITHER, 20 * i, 1), this.getEffectSource()); // CraftBukkit
             }
          }
 
@@ -83,8 +_,14 @@
    protected void onHit(HitResult p_37628_) {
       super.onHit(p_37628_);
       if (!this.level.isClientSide) {
-         Explosion.BlockInteraction explosion$blockinteraction = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE;
-         this.level.explode(this, this.getX(), this.getY(), this.getZ(), 1.0F, false, explosion$blockinteraction);
+         Explosion.BlockInteraction explosion$blockinteraction = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level, this.getOwner()) ? Explosion.BlockInteraction.DESTROY : Explosion.BlockInteraction.NONE;
+         // CraftBukkit start
+         org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), 1.0F, false);
+         this.level.getCraftServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            this.level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), explosion$blockinteraction);
+         }
+         // CraftBukkit end
          this.discard();
       }
 
