--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -26,8 +_,23 @@
    }
 
    public ClientboundInitializeBorderPacket(WorldBorder p_178877_) {
-      this.f_178868_ = p_178877_.m_6347_();
-      this.f_178869_ = p_178877_.m_6345_();
+      //Magma start
+      if (p_178877_.world == null)
+         p_178877_.world = org.magmafoundation.magma.helpers.BorderHelper.getLevelFromPlayer(this);
+
+      double newCenterX = p_178877_.m_6347_();
+      double newCenterZ = p_178877_.m_6345_();
+
+      if (p_178877_.world != null) {
+         // CraftBukkit start - multiply out nether border
+         newCenterX = newCenterX * p_178877_.world.m_6042_().m_63959_();
+         newCenterZ = newCenterZ * p_178877_.world.m_6042_().m_63959_();
+         // CraftBukkit end
+      }
+
+      this.f_178868_ = newCenterX;
+      this.f_178869_ = newCenterZ;
+      //Magma end
       this.f_178870_ = p_178877_.m_61959_();
       this.f_178871_ = p_178877_.m_61961_();
       this.f_178872_ = p_178877_.m_61960_();
