--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -26,8 +_,23 @@
    }
 
    public ClientboundInitializeBorderPacket(WorldBorder p_178877_) {
-      this.newCenterX = p_178877_.getCenterX();
-      this.newCenterZ = p_178877_.getCenterZ();
+      //Magma start
+      if (p_178877_.world == null)
+         p_178877_.world = org.magmafoundation.magma.helpers.BorderHelper.getLevelFromPlayer(this);
+
+      double newCenterX = p_178877_.getCenterX();
+      double newCenterZ = p_178877_.getCenterZ();
+
+      if (p_178877_.world != null) {
+         // CraftBukkit start - multiply out nether border
+         newCenterX = newCenterX * p_178877_.world.dimensionType().coordinateScale();
+         newCenterZ = newCenterZ * p_178877_.world.dimensionType().coordinateScale();
+         // CraftBukkit end
+      }
+
+      this.newCenterX = newCenterX;
+      this.newCenterZ = newCenterZ;
+      //Magma end
       this.oldSize = p_178877_.getSize();
       this.newSize = p_178877_.getLerpTarget();
       this.lerpTime = p_178877_.getLerpRemainingTime();
