--- a/net/minecraft/network/protocol/game/ClientboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundChatPacket.java
@@ -8,6 +_,7 @@
 
 public class ClientboundChatPacket implements Packet<ClientGamePacketListener> {
    private final Component message;
+   public net.md_5.bungee.api.chat.BaseComponent[] components; // Spigot
    private final ChatType type;
    private final UUID sender;
 
@@ -24,7 +_,13 @@
    }
 
    public void write(FriendlyByteBuf p_131838_) {
-      p_131838_.writeComponent(this.message);
+      // Spigot start
+      if (components != null) {
+         p_131838_.writeUtf(net.md_5.bungee.chat.ComponentSerializer.toString(components));
+      } else {
+         p_131838_.writeComponent(this.message);
+      }
+      // Spigot end
       p_131838_.writeByte(this.type.getIndex());
       p_131838_.writeUUID(this.sender);
    }
