--- a/net/minecraft/network/protocol/game/ClientboundPlayerPositionPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundPlayerPositionPacket.java
@@ -1,9 +_,10 @@
 package net.minecraft.network.protocol.game;
 
-import java.util.EnumSet;
-import java.util.Set;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
+
+import java.util.EnumSet;
+import java.util.Set;
 
 public class ClientboundPlayerPositionPacket implements Packet<ClientGamePacketListener> {
    private final double f_132796_;
