--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -16,14 +_,22 @@
    }
 
    public ServerboundChatPacket(FriendlyByteBuf p_179545_) {
-      this.message = p_179545_.readUtf(256);
+      this.message = org.apache.commons.lang3.StringUtils.normalizeSpace(p_179545_.readUtf(256)); // CraftBukkit - see PlayerConnection
    }
 
    public void write(FriendlyByteBuf p_133839_) {
       p_133839_.writeUtf(this.message);
    }
 
+   // Spigot Start
+   private static final java.util.concurrent.ExecutorService executors = java.util.concurrent.Executors.newCachedThreadPool(new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Chat Thread - #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER)).build());
+
    public void handle(ServerGamePacketListener p_133836_) {
+      if (!message.startsWith("/")) {
+         executors.submit(() -> p_133836_.handleChat(ServerboundChatPacket.this));
+         return;
+      }
+      // Spigot End
       p_133836_.handleChat(this);
    }
 
