--- a/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
@@ -6,12 +_,14 @@
 
 public class ServerboundUseItemPacket implements Packet<ServerGamePacketListener> {
    private final InteractionHand hand;
+   public long timestamp; // Spigot
 
    public ServerboundUseItemPacket(InteractionHand p_134710_) {
       this.hand = p_134710_;
    }
 
    public ServerboundUseItemPacket(FriendlyByteBuf p_179798_) {
+      this.timestamp = System.currentTimeMillis(); // Spigot
       this.hand = p_179798_.readEnum(InteractionHand.class);
    }
 
