--- a/net/minecraft/network/protocol/game/ServerboundCustomPayloadPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundCustomPayloadPacket.java
@@ -5,7 +_,7 @@
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.resources.ResourceLocation;
 
-public class ServerboundCustomPayloadPacket implements Packet<ServerGamePacketListener> {
+public class ServerboundCustomPayloadPacket implements Packet<ServerGamePacketListener>, net.minecraftforge.network.ICustomPacket<ServerboundCustomPayloadPacket> {
    private static final int MAX_PAYLOAD_SIZE = 32767;
    public static final ResourceLocation BRAND = new ResourceLocation("brand");
    public final ResourceLocation identifier;
@@ -28,7 +_,7 @@
 
    public void write(FriendlyByteBuf p_133994_) {
       p_133994_.writeResourceLocation(this.identifier);
-      p_133994_.writeBytes((ByteBuf)this.data);
+      p_133994_.writeBytes((ByteBuf)this.data.copy()); //This may be access multiple times, from multiple threads, lets be safe like the S->C packet
    }
 
    public void handle(ServerGamePacketListener p_133992_) {
