--- a/net/minecraft/world/level/portal/PortalInfo.java
+++ b/net/minecraft/world/level/portal/PortalInfo.java
@@ -1,12 +_,24 @@
 package net.minecraft.world.level.portal;
 
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent;
 
 public class PortalInfo {
    public final Vec3 f_77676_;
    public final Vec3 f_77677_;
    public final float f_77678_;
    public final float f_77679_;
+
+   // CraftBukkit start
+   public ServerLevel world;
+   public CraftPortalEvent portalEventInfo;
+
+   public void setCBPortalEvent(ServerLevel world, CraftPortalEvent portalEventInfo) {
+      this.world = world;
+      this.portalEventInfo = portalEventInfo;
+   }
+   // CraftBukkit end
 
    public PortalInfo(Vec3 p_77681_, Vec3 p_77682_, float p_77683_, float p_77684_) {
       this.f_77676_ = p_77681_;
