--- a/net/minecraft/world/level/portal/PortalInfo.java
+++ b/net/minecraft/world/level/portal/PortalInfo.java
@@ -8,6 +_,16 @@
    public final float f_77678_;
    public final float f_77679_;
 
+   // CraftBukkit start
+   public net.minecraft.server.level.ServerLevel world;
+   public org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent portalEventInfo;
+
+   public void setCBPortalEvent(net.minecraft.server.level.ServerLevel world, org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent portalEventInfo) {
+      this.world = world;
+      this.portalEventInfo = portalEventInfo;
+   }
+   // CraftBukkit end
+
    public PortalInfo(Vec3 p_77681_, Vec3 p_77682_, float p_77683_, float p_77684_) {
       this.f_77676_ = p_77681_;
       this.f_77677_ = p_77682_;
