--- a/net/minecraft/world/level/portal/PortalShape.java
+++ b/net/minecraft/world/level/portal/PortalShape.java
@@ -24,7 +_,7 @@
    private static final int MIN_HEIGHT = 3;
    public static final int MAX_HEIGHT = 21;
    private static final BlockBehaviour.StatePredicate FRAME = (p_77720_, p_77721_, p_77722_) -> {
-      return p_77720_.is(Blocks.OBSIDIAN);
+      return p_77720_.isPortalFrame(p_77721_, p_77722_);
    };
    private final LevelAccessor level;
    private final Direction.Axis axis;
@@ -34,6 +_,7 @@
    private BlockPos bottomLeft;
    private int height;
    private final int width;
+    org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator blocks; // CraftBukkit - add field
 
    public static Optional<PortalShape> findEmptyPortalShape(LevelAccessor p_77709_, BlockPos p_77710_, Direction.Axis p_77711_) {
       return findPortalShape(p_77709_, p_77710_, (p_77727_) -> {
@@ -52,6 +_,7 @@
    }
 
    public PortalShape(LevelAccessor p_77695_, BlockPos p_77696_, Direction.Axis p_77697_) {
+      this.blocks = new org.bukkit.craftbukkit.v1_18_R2.util.BlockStateListPopulator( p_77695_.getMinecraftWorld() ); // CraftBukkit - add field
       this.level = p_77695_;
       this.axis = p_77697_;
       this.rightDir = p_77697_ == Direction.Axis.X ? Direction.WEST : Direction.SOUTH;
@@ -92,6 +_,7 @@
          BlockState blockstate = this.level.getBlockState(blockpos$mutableblockpos);
          if (!isEmpty(blockstate)) {
             if (FRAME.test(blockstate, this.level, blockpos$mutableblockpos)) {
+                    blocks.setBlock(blockpos$mutableblockpos, blockstate, 18 ); // CraftBukkit - lower left / right
                return i;
             }
             break;
@@ -101,6 +_,7 @@
          if (!FRAME.test(blockstate1, this.level, blockpos$mutableblockpos)) {
             break;
          }
+            blocks.setBlock(blockpos$mutableblockpos, blockstate1, 18); // CraftBukkit - bottom row
       }
 
       return 0;
@@ -118,6 +_,7 @@
          if (!FRAME.test(this.level.getBlockState(blockpos$mutableblockpos), this.level, blockpos$mutableblockpos)) {
             return false;
          }
+            blocks.setBlock(blockpos$mutableblockpos, this.level.getBlockState(blockpos$mutableblockpos), 18); // CraftBukkit - upper row
       }
 
       return true;
@@ -146,6 +_,10 @@
                ++this.numPortalBlocks;
             }
          }
+            // CraftBukkit start - left and right
+            blocks.setBlock(p_77729_.set(this.bottomLeft).move(Direction.UP, i).move(this.rightDir, -1), this.level.getBlockState(p_77729_), 18);
+            blocks.setBlock(p_77729_.set(this.bottomLeft).move(Direction.UP, i).move(this.rightDir, this.width), this.level.getBlockState(p_77729_), 18);
+            // CraftBukkit end
       }
 
       return 21;
@@ -160,10 +_,30 @@
    }
 
    public void createPortalBlocks() {
+        CBcreatePortalBlocks();
+    }
+
+    // CraftBukkit start - return boolean
+    public boolean CBcreatePortalBlocks() {
+        org.bukkit.World bworld = this.level.getMinecraftWorld().getWorld();
+
+        // Copy below for loop
       BlockState blockstate = Blocks.NETHER_PORTAL.defaultBlockState().setValue(NetherPortalBlock.AXIS, this.axis);
       BlockPos.betweenClosed(this.bottomLeft, this.bottomLeft.relative(Direction.UP, this.height - 1).relative(this.rightDir, this.width - 1)).forEach((p_77725_) -> {
+            blocks.setBlock( p_77725_, blockstate, 18 );
+        } );
+
+        org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks.getList(), bworld, null, org.bukkit.event.world.PortalCreateEvent.CreateReason.FIRE);
+        this.level.getMinecraftWorld().getServer().server.getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return false;
+        }
+        // CraftBukkit end
+        BlockPos.betweenClosed(this.bottomLeft, this.bottomLeft.relative(Direction.UP, this.height - 1).relative(this.rightDir, this.width - 1)).forEach((p_77725_) -> {
          this.level.setBlock(p_77725_, blockstate, 18);
       });
+        return true; // CraftBukkit
    }
 
    public boolean isComplete() {
@@ -196,6 +_,10 @@
    }
 
    public static PortalInfo createPortalInfo(ServerLevel p_77700_, BlockUtil.FoundRectangle p_77701_, Direction.Axis p_77702_, Vec3 p_77703_, EntityDimensions p_77704_, Vec3 p_77705_, float p_77706_, float p_77707_) {
+        return createPortalInfo( p_77700_, p_77701_, p_77702_, p_77703_, p_77704_, p_77705_, p_77706_, p_77707_, null );
+    }
+
+    public static PortalInfo createPortalInfo(ServerLevel p_77700_, BlockUtil.FoundRectangle p_77701_, Direction.Axis p_77702_, Vec3 p_77703_, EntityDimensions p_77704_, Vec3 p_77705_, float p_77706_, float p_77707_, org.bukkit.craftbukkit.v1_18_R2.event.CraftPortalEvent event) {
       BlockPos blockpos = p_77701_.minCorner;
       BlockState blockstate = p_77700_.getBlockState(blockpos);
       Direction.Axis direction$axis = blockstate.getOptionalValue(BlockStateProperties.HORIZONTAL_AXIS).orElse(Direction.Axis.X);
@@ -208,6 +_,8 @@
       double d4 = 0.5D + p_77703_.z();
       boolean flag = direction$axis == Direction.Axis.X;
       Vec3 vec31 = new Vec3((double)blockpos.getX() + (flag ? d2 : d4), (double)blockpos.getY() + d3, (double)blockpos.getZ() + (flag ? d4 : d2));
-      return new PortalInfo(vec31, vec3, p_77706_ + (float)i, p_77707_);
+        PortalInfo info = new PortalInfo( vec31, vec3, p_77706_ + (float) i, p_77707_ );
+        info.setCBPortalEvent( p_77700_, event );
+        return info;
    }
 }
