--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -18,13 +_,28 @@
    final String group;
    final ItemStack result;
    final NonNullList<Ingredient> ingredients;
+   private final boolean isSimple;
 
    public ShapelessRecipe(ResourceLocation p_44246_, String p_44247_, ItemStack p_44248_, NonNullList<Ingredient> p_44249_) {
       this.id = p_44246_;
       this.group = p_44247_;
       this.result = p_44248_;
       this.ingredients = p_44249_;
-   }
+      this.isSimple = p_44249_.stream().allMatch(Ingredient::isSimple);
+   }
+
+   // CraftBukkit start
+   @SuppressWarnings("unchecked")
+   public org.bukkit.inventory.ShapelessRecipe toBukkitRecipe() {
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack result = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asCraftMirror(this.result);
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftShapelessRecipe recipe = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftShapelessRecipe(result, this);
+      recipe.setGroup(this.group);
+      for (Ingredient list : this.ingredients) {
+         recipe.addIngredient(org.bukkit.craftbukkit.v1_18_R2.inventory.CraftRecipe.toBukkit(list));
+      }
+      return recipe;
+   }
+   // CraftBukkit end
 
    public ResourceLocation getId() {
       return this.id;
@@ -48,17 +_,20 @@
 
    public boolean matches(CraftingContainer p_44262_, Level p_44263_) {
       StackedContents stackedcontents = new StackedContents();
+      java.util.List<ItemStack> inputs = new java.util.ArrayList<>();
       int i = 0;
 
       for(int j = 0; j < p_44262_.getContainerSize(); ++j) {
          ItemStack itemstack = p_44262_.getItem(j);
          if (!itemstack.isEmpty()) {
             ++i;
+            if (isSimple)
             stackedcontents.accountStack(itemstack, 1);
+            else inputs.add(itemstack);
          }
       }
 
-      return i == this.ingredients.size() && stackedcontents.canCraft(this, (IntList)null);
+      return i == this.ingredients.size() && (isSimple ? stackedcontents.canCraft(this, (IntList)null) : net.minecraftforge.common.util.RecipeMatcher.findMatches(inputs,  this.ingredients) != null);
    }
 
    public ItemStack assemble(CraftingContainer p_44260_) {
@@ -69,14 +_,15 @@
       return p_44252_ * p_44253_ >= this.ingredients.size();
    }
 
-   public static class Serializer implements RecipeSerializer<ShapelessRecipe> {
+   public static class Serializer extends net.minecraftforge.registries.ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<ShapelessRecipe> {
+      private static final ResourceLocation NAME = new ResourceLocation("minecraft", "crafting_shapeless");
       public ShapelessRecipe fromJson(ResourceLocation p_44290_, JsonObject p_44291_) {
          String s = GsonHelper.getAsString(p_44291_, "group", "");
          NonNullList<Ingredient> nonnulllist = itemsFromJson(GsonHelper.getAsJsonArray(p_44291_, "ingredients"));
          if (nonnulllist.isEmpty()) {
             throw new JsonParseException("No ingredients for shapeless recipe");
-         } else if (nonnulllist.size() > 9) {
-            throw new JsonParseException("Too many ingredients for shapeless recipe");
+         } else if (nonnulllist.size() > ShapedRecipe.MAX_WIDTH * ShapedRecipe.MAX_HEIGHT) {
+            throw new JsonParseException("Too many ingredients for shapeless recipe. The maximum is " + (ShapedRecipe.MAX_WIDTH * ShapedRecipe.MAX_HEIGHT));
          } else {
             ItemStack itemstack = ShapedRecipe.itemStackFromJson(GsonHelper.getAsJsonObject(p_44291_, "result"));
             return new ShapelessRecipe(p_44290_, s, itemstack, nonnulllist);
@@ -88,7 +_,7 @@
 
          for(int i = 0; i < p_44276_.size(); ++i) {
             Ingredient ingredient = Ingredient.fromJson(p_44276_.get(i));
-            if (!ingredient.isEmpty()) {
+            if (net.minecraftforge.common.ForgeConfig.SERVER.skipEmptyShapelessCheck.get() || !ingredient.isEmpty()) {
                nonnulllist.add(ingredient);
             }
          }
