--- a/net/minecraft/world/item/crafting/SmeltingRecipe.java
+++ b/net/minecraft/world/item/crafting/SmeltingRecipe.java
@@ -16,4 +_,14 @@
    public RecipeSerializer<?> getSerializer() {
       return RecipeSerializer.SMELTING_RECIPE;
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack result = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asCraftMirror(this.result);
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftFurnaceRecipe recipe = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftFurnaceRecipe(org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey.fromMinecraft(this.id), result, org.bukkit.craftbukkit.v1_18_R2.inventory.CraftRecipe.toBukkit(this.ingredient), this.experience, this.cookingTime);
+      recipe.setGroup(this.group);
+      return recipe;
+   }
+   // CraftBukkit end
 }
