--- a/net/minecraft/world/item/crafting/StonecutterRecipe.java
+++ b/net/minecraft/world/item/crafting/StonecutterRecipe.java
@@ -18,4 +_,15 @@
    public ItemStack m_8042_() {
       return new ItemStack(Blocks.f_50679_);
    }
+
+   // CraftBukkit start
+   @Override
+   public org.bukkit.inventory.Recipe toBukkitRecipe() {
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack result = org.bukkit.craftbukkit.v1_18_R2.inventory.CraftItemStack.asCraftMirror(this.f_44410_);
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftStonecuttingRecipe recipe = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftStonecuttingRecipe(org.bukkit.craftbukkit.v1_18_R2.util.CraftNamespacedKey.fromMinecraft(this.f_44411_), result, org.bukkit.craftbukkit.v1_18_R2.inventory.CraftRecipe.toBukkit(this.f_44409_));
+      recipe.setGroup(this.f_44412_);
+      return recipe;
+   }
+   // CraftBukkit end
+
 }
