--- a/net/minecraft/world/entity/schedule/Schedule.java
+++ b/net/minecraft/world/entity/schedule/Schedule.java
@@ -8,7 +_,7 @@
 import java.util.stream.Collectors;
 import net.minecraft.core.Registry;
 
-public class Schedule {
+public class Schedule extends net.minecraftforge.registries.ForgeRegistryEntry<Schedule> {
    public static final int WORK_START_TIME = 2000;
    public static final int TOTAL_WORK_TIME = 7000;
    public static final Schedule EMPTY = register("empty").changeActivityAt(0, Activity.IDLE).build();
