--- a/net/minecraft/client/gui/screens/worldselection/WorldGenSettingsComponent.java
+++ b/net/minecraft/client/gui/screens/worldselection/WorldGenSettingsComponent.java
@@ -107,6 +_,7 @@
       this.customWorldDummyButton.visible = false;
       this.customizeTypeButton = p_101430_.addRenderableWidget(new Button(j, 120, 150, 20, new TranslatableComponent("selectWorld.customizeType"), (p_170248_) -> {
          WorldPreset.PresetEditor worldpreset$preseteditor = WorldPreset.EDITORS.get(this.preset);
+         worldpreset$preseteditor = net.minecraftforge.client.ForgeHooksClient.getPresetEditor(this.preset, worldpreset$preseteditor);
          if (worldpreset$preseteditor != null) {
             p_101431_.setScreen(worldpreset$preseteditor.createEditScreen(p_101430_, this.settings));
          }
@@ -273,7 +_,7 @@
       } else {
          this.featuresButton.visible = p_170288_;
          this.bonusItemsButton.visible = p_170288_;
-         this.customizeTypeButton.visible = p_170288_ && WorldPreset.EDITORS.containsKey(this.preset);
+         this.customizeTypeButton.visible = p_170288_ && (WorldPreset.EDITORS.containsKey(this.preset) || net.minecraftforge.client.ForgeHooksClient.hasPresetEditor(this.preset));
          this.importSettingsButton.visible = p_170288_;
       }
 
