--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 
 public class DedicatedServerProperties extends Settings<DedicatedServerProperties> {
+   public final boolean debug = this.get("debug", false); // CraftBukkit
    public final boolean onlineMode = this.get("online-mode", true);
    public final boolean preventProxyConnections = this.get("prevent-proxy-connections", false);
    public final String serverIp = this.get("server-ip", "");
