--- a/net/minecraft/server/dedicated/Settings.java
+++ b/net/minecraft/server/dedicated/Settings.java
@@ -60,7 +_,7 @@
          OutputStream outputstream = Files.newOutputStream(p_139877_);
 
          try {
-            this.properties.store(outputstream, "Minecraft server properties");
+            net.minecraftforge.common.util.SortedProperties.store(properties, outputstream, "Minecraft server properties");
          } catch (Throwable throwable1) {
             if (outputstream != null) {
                try {
