--- a/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
@@ -36,7 +_,13 @@
                return p_147412_.getJobPoiType() == p_22457_;
             }).findFirst();
          }).ifPresent((p_22464_) -> {
-            p_22454_.setVillagerData(p_22454_.getVillagerData().setProfession(p_22464_));
+            // CraftBukkit start - Fire org.bukkit.event.entity.VillagerCareerChangeEvent where Villager gets employed
+            org.bukkit.event.entity.VillagerCareerChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callVillagerCareerChangeEvent(p_22454_, org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager.nmsToBukkitProfession(p_22464_), org.bukkit.event.entity.VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+            if (event.isCancelled()) {
+               return;
+            }
+            p_22454_.setVillagerData(p_22454_.getVillagerData().setProfession(org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager.bukkitToNmsProfession(event.getProfession())));
+            // CraftBukkit end
             p_22454_.refreshBrain(p_22453_);
          });
       }
