--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -81,6 +_,7 @@
    }
 
    public static void throwItem(LivingEntity p_22614_, ItemStack p_22615_, Vec3 p_22616_) {
+      if (p_22615_.isEmpty()) return; // CraftBukkit - SPIGOT-4940: no empty loot
       double d0 = p_22614_.getEyeY() - (double)0.3F;
       ItemEntity itementity = new ItemEntity(p_22614_.level, p_22614_.getX(), d0, p_22614_.getZ(), p_22615_);
       float f = 0.3F;
@@ -88,6 +_,13 @@
       vec3 = vec3.normalize().scale((double)0.3F);
       itementity.setDeltaMovement(vec3);
       itementity.setDefaultPickUpDelay();
+      // CraftBukkit start
+      org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(p_22614_.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+      itementity.level.getCraftServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         return;
+      }
+      // CraftBukkit end
       p_22614_.level.addFreshEntity(itementity);
    }
 
