--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -38,7 +_,7 @@
    }
 
    protected boolean checkExtraStartConditions(ServerLevel p_23174_, Villager p_23175_) {
-      if (!p_23174_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_23174_, p_23175_)) {
          return false;
       } else if (p_23175_.getVillagerData().getProfession() != VillagerProfession.FARMER) {
          return false;
@@ -96,7 +_,11 @@
             Block block = blockstate.getBlock();
             Block block1 = p_23196_.getBlockState(this.aboveFarmlandPos.below()).getBlock();
             if (block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockstate)) {
-               p_23196_.destroyBlock(this.aboveFarmlandPos, true, p_23197_);
+               // CraftBukkit start
+               if (!org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.aboveFarmlandPos, Blocks.AIR.defaultBlockState()).isCancelled()) {
+                  p_23196_.destroyBlock(this.aboveFarmlandPos, true, p_23197_);
+               }
+               // CraftBukkit end
             }
 
             if (blockstate.isAir() && block1 instanceof FarmBlock && p_23197_.hasFarmSeeds()) {
@@ -106,19 +_,33 @@
                   ItemStack itemstack = simplecontainer.getItem(i);
                   boolean flag = false;
                   if (!itemstack.isEmpty()) {
+                     // CraftBukkit start
+                     Block planted = null;
                      if (itemstack.is(Items.WHEAT_SEEDS)) {
-                        p_23196_.setBlock(this.aboveFarmlandPos, Blocks.WHEAT.defaultBlockState(), 3);
+                        planted = Blocks.WHEAT;
                         flag = true;
                      } else if (itemstack.is(Items.POTATO)) {
-                        p_23196_.setBlock(this.aboveFarmlandPos, Blocks.POTATOES.defaultBlockState(), 3);
+                        planted = Blocks.POTATOES;
                         flag = true;
                      } else if (itemstack.is(Items.CARROT)) {
-                        p_23196_.setBlock(this.aboveFarmlandPos, Blocks.CARROTS.defaultBlockState(), 3);
+                        planted = Blocks.CARROTS;
                         flag = true;
                      } else if (itemstack.is(Items.BEETROOT_SEEDS)) {
-                        p_23196_.setBlock(this.aboveFarmlandPos, Blocks.BEETROOTS.defaultBlockState(), 3);
+                        planted = Blocks.BEETROOTS;
                         flag = true;
-                     }
+                     } else if (itemstack.getItem() instanceof net.minecraftforge.common.IPlantable) {
+                        if (((net.minecraftforge.common.IPlantable)itemstack.getItem()).getPlantType(p_23196_, aboveFarmlandPos) == net.minecraftforge.common.PlantType.CROP) {
+                           p_23196_.setBlock(aboveFarmlandPos, ((net.minecraftforge.common.IPlantable)itemstack.getItem()).getPlant(p_23196_, aboveFarmlandPos), 3);
+                           flag = true;
+                        }
+                     }
+
+                     if (planted != null && !org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityChangeBlockEvent(p_23197_, this.aboveFarmlandPos, planted.defaultBlockState()).isCancelled()) {
+                        p_23196_.setBlock(this.aboveFarmlandPos, planted.defaultBlockState(), 3);
+                     } else {
+                        flag = false;
+                     }
+                     // CraftBukkit end
                   }
 
                   if (flag) {
