--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -60,6 +_,13 @@
       })) {
          DoorBlock doorblock = (DoorBlock)blockstate.getBlock();
          if (!doorblock.isOpen(blockstate)) {
+            // CraftBukkit start - entities opening doors
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(p_23296_.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_23296_.level, blockpos));
+            p_23296_.level.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             doorblock.setOpen(p_23296_, p_23295_, blockstate, blockpos, true);
          }
 
@@ -73,6 +_,13 @@
       })) {
          DoorBlock doorblock1 = (DoorBlock)blockstate1.getBlock();
          if (!doorblock1.isOpen(blockstate1)) {
+            // CraftBukkit start - entities opening doors
+            org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(p_23296_.getBukkitEntity(), org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock.at(p_23296_.level, blockpos1));
+            p_23296_.level.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            // CraftBukkit end
             doorblock1.setOpen(p_23296_, p_23295_, blockstate1, blockpos1, true);
             this.rememberDoorToClose(p_23295_, p_23296_, blockpos1);
          }
