--- a/net/minecraft/world/entity/ai/behavior/ResetProfession.java
+++ b/net/minecraft/world/entity/ai/behavior/ResetProfession.java
@@ -19,7 +_,13 @@
    }
 
    protected void m_6735_(ServerLevel p_23776_, Villager p_23777_, long p_23778_) {
-      p_23777_.m_141967_(p_23777_.m_7141_().m_35565_(VillagerProfession.f_35585_));
+      // CraftBukkit start
+      org.bukkit.event.entity.VillagerCareerChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callVillagerCareerChangeEvent(p_23777_, org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager.nmsToBukkitProfession(VillagerProfession.f_35585_), org.bukkit.event.entity.VillagerCareerChangeEvent.ChangeReason.LOSING_JOB);
+      if (event.isCancelled()) {
+         return;
+      }
+      p_23777_.m_141967_(p_23777_.m_7141_().m_35565_(org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager.bukkitToNmsProfession(event.getProfession())));
+      // CraftBukkit end
       p_23777_.m_35483_(p_23776_);
    }
 }
