--- a/net/minecraft/world/entity/ai/behavior/ResetProfession.java
+++ b/net/minecraft/world/entity/ai/behavior/ResetProfession.java
@@ -19,7 +_,13 @@
    }
 
    protected void start(ServerLevel p_23776_, Villager p_23777_, long p_23778_) {
-      p_23777_.setVillagerData(p_23777_.getVillagerData().setProfession(VillagerProfession.NONE));
+      // CraftBukkit start
+      org.bukkit.event.entity.VillagerCareerChangeEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callVillagerCareerChangeEvent(p_23777_, org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager.nmsToBukkitProfession(VillagerProfession.NONE), org.bukkit.event.entity.VillagerCareerChangeEvent.ChangeReason.LOSING_JOB);
+      if (event.isCancelled()) {
+         return;
+      }
+      p_23777_.setVillagerData(p_23777_.getVillagerData().setProfession(org.bukkit.craftbukkit.v1_18_R2.entity.CraftVillager.bukkitToNmsProfession(event.getProfession())));
+      // CraftBukkit end
       p_23777_.refreshBrain(p_23776_);
    }
 }
