--- a/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -73,6 +_,17 @@
    }
 
    protected void m_24255_(E p_24256_) {
+      // CraftBukkit start
+      LivingEntity old = p_24256_.m_6274_().m_21952_(MemoryModuleType.f_26372_).orElse(null);
+      org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityTargetLivingEvent(p_24256_, null, (old != null && !old.m_6084_()) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED : org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET);
+      if (event.isCancelled()) {
+         return;
+      }
+      if (event.getTarget() != null) {
+         p_24256_.m_6274_().m_21879_(MemoryModuleType.f_26372_, ((org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity) event.getTarget()).getHandle());
+         return;
+      }
+      // CraftBukkit end
       this.f_147979_.accept(p_24256_);
       p_24256_.m_6274_().m_21936_(MemoryModuleType.f_26372_);
    }
