--- a/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -73,6 +_,17 @@
    }
 
    protected void clearAttackTarget(E p_24256_) {
+      // CraftBukkit start
+      LivingEntity old = p_24256_.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null);
+      org.bukkit.event.entity.EntityTargetEvent event = org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callEntityTargetLivingEvent(p_24256_, null, (old != null && !old.isAlive()) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED : org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET);
+      if (event.isCancelled()) {
+         return;
+      }
+      if (event.getTarget() != null) {
+         p_24256_.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, ((org.bukkit.craftbukkit.v1_18_R2.entity.CraftLivingEntity) event.getTarget()).getHandle());
+         return;
+      }
+      // CraftBukkit end
       this.onTargetErased.accept(p_24256_);
       p_24256_.getBrain().eraseMemory(MemoryModuleType.ATTACK_TARGET);
    }
