--- a/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
+++ b/net/minecraft/world/entity/ai/behavior/WorkAtComposter.java
@@ -35,7 +_,7 @@
    private void compostItems(ServerLevel p_24793_, Villager p_24794_, GlobalPos p_24795_, BlockState p_24796_) {
       BlockPos blockpos = p_24795_.pos();
       if (p_24796_.getValue(ComposterBlock.LEVEL) == 8) {
-         p_24796_ = ComposterBlock.extractProduce(p_24796_, p_24793_, blockpos);
+         p_24796_ = ComposterBlock.extractProduce(p_24796_, p_24793_, blockpos, p_24794_); // CraftBukkit
       }
 
       int i = 20;
@@ -57,7 +_,7 @@
                i -= l1;
 
                for(int i2 = 0; i2 < l1; ++i2) {
-                  blockstate = ComposterBlock.insertItem(blockstate, p_24793_, itemstack, blockpos);
+                  blockstate = ComposterBlock.insertItem(blockstate, p_24793_, itemstack, blockpos, p_24794_); // CraftBukkit
                   if (blockstate.getValue(ComposterBlock.LEVEL) == 7) {
                      this.spawnComposterFillEffects(p_24793_, p_24796_, blockpos, blockstate);
                      return;
