--- a/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
+++ b/net/minecraft/client/renderer/chunk/ChunkRenderDispatcher.java
@@ -73,12 +_,15 @@
    private Vec3 camera = Vec3.ZERO;
 
    public ChunkRenderDispatcher(ClientLevel p_194405_, LevelRenderer p_194406_, Executor p_194407_, boolean p_194408_, ChunkBufferBuilderPack p_194409_) {
+      this(p_194405_, p_194406_, p_194407_, p_194408_, p_194409_, -1);
+   }
+   public ChunkRenderDispatcher(ClientLevel p_194405_, LevelRenderer p_194406_, Executor p_194407_, boolean p_194408_, ChunkBufferBuilderPack p_194409_, int countRenderBuilders) {
       this.level = p_194405_;
       this.renderer = p_194406_;
       int i = Math.max(1, (int)((double)Runtime.getRuntime().maxMemory() * 0.3D) / (RenderType.chunkBufferLayers().stream().mapToInt(RenderType::bufferSize).sum() * 4) - 1);
       int j = Runtime.getRuntime().availableProcessors();
       int k = p_194408_ ? j : Math.min(j, 4);
-      int l = Math.max(1, Math.min(k, i));
+      int l = countRenderBuilders < 0 ? Math.max(1, Math.min(k, i)) : countRenderBuilders;
       this.fixedBuffers = p_194409_;
       List<ChunkBufferBuilderPack> list = Lists.newArrayListWithExpectedSize(l);
 
@@ -420,7 +_,7 @@
          if (!chunkrenderdispatcher$compiledchunk.hasLayer.contains(p_112810_)) {
             return false;
          } else {
-            this.lastResortTransparencyTask = new ChunkRenderDispatcher.RenderChunk.ResortTransparencyTask(this.getDistToPlayerSqr(), chunkrenderdispatcher$compiledchunk);
+            this.lastResortTransparencyTask = new ChunkRenderDispatcher.RenderChunk.ResortTransparencyTask(new net.minecraft.world.level.ChunkPos(getOrigin()), this.getDistToPlayerSqr(), chunkrenderdispatcher$compiledchunk);
             p_112811_.schedule(this.lastResortTransparencyTask);
             return true;
          }
@@ -452,7 +_,7 @@
             this.initialCompilationCancelCount.incrementAndGet();
          }
 
-         this.lastRebuildTask = new ChunkRenderDispatcher.RenderChunk.RebuildTask(this.getDistToPlayerSqr(), renderchunkregion, !flag1 || this.initialCompilationCancelCount.get() > 2);
+         this.lastRebuildTask = new ChunkRenderDispatcher.RenderChunk.RebuildTask(new net.minecraft.world.level.ChunkPos(getOrigin()), this.getDistToPlayerSqr(), renderchunkregion, flag || this.compiled.get() != ChunkRenderDispatcher.CompiledChunk.UNCOMPILED);
          return this.lastRebuildTask;
       }
 
@@ -485,10 +_,20 @@
          protected final double distAtCreation;
          protected final AtomicBoolean isCancelled = new AtomicBoolean(false);
          protected final boolean isHighPriority;
+         protected java.util.Map<net.minecraft.core.BlockPos, net.minecraftforge.client.model.data.IModelData> modelData;
 
          public ChunkCompileTask(double p_194423_, boolean p_194424_) {
+            this(null, p_194423_, p_194424_);
+         }
+
+         public ChunkCompileTask(@Nullable net.minecraft.world.level.ChunkPos pos, double p_194423_, boolean p_194424_) {
             this.distAtCreation = p_194423_;
             this.isHighPriority = p_194424_;
+            if (pos == null) {
+               this.modelData = java.util.Collections.emptyMap();
+            } else {
+               this.modelData = net.minecraftforge.client.model.ModelDataManager.getModelData(net.minecraft.client.Minecraft.getInstance().level, pos);
+            }
          }
 
          public abstract CompletableFuture<ChunkRenderDispatcher.ChunkTaskResult> doTask(ChunkBufferBuilderPack p_112853_);
@@ -500,6 +_,10 @@
          public int compareTo(ChunkRenderDispatcher.RenderChunk.ChunkCompileTask p_112855_) {
             return Doubles.compare(this.distAtCreation, p_112855_.distAtCreation);
          }
+
+         public net.minecraftforge.client.model.data.IModelData getModelData(net.minecraft.core.BlockPos pos) {
+            return modelData.getOrDefault(pos, net.minecraftforge.client.model.data.EmptyModelData.INSTANCE);
+         }
       }
 
       @OnlyIn(Dist.CLIENT)
@@ -507,8 +_,13 @@
          @Nullable
          protected RenderChunkRegion region;
 
+         @Deprecated
          public RebuildTask(@Nullable double p_194427_, RenderChunkRegion p_194428_, boolean p_194429_) {
-            super(p_194427_, p_194429_);
+            this(null, p_194427_, p_194428_, p_194429_);
+         }
+
+         public RebuildTask(@Nullable net.minecraft.world.level.ChunkPos pos, double p_194427_, @Nullable RenderChunkRegion p_194428_, boolean p_194429_) {
+            super(pos, p_194427_, p_194429_);
             this.region = p_194428_;
          }
 
@@ -591,8 +_,10 @@
 
                   BlockState blockstate1 = renderchunkregion.getBlockState(blockpos2);
                   FluidState fluidstate = blockstate1.getFluidState();
-                  if (!fluidstate.isEmpty()) {
-                     RenderType rendertype = ItemBlockRenderTypes.getRenderLayer(fluidstate);
+                  net.minecraftforge.client.model.data.IModelData modelData = getModelData(blockpos2);
+                  for (RenderType rendertype : RenderType.chunkBufferLayers()) {
+                     net.minecraftforge.client.ForgeHooksClient.setRenderType(rendertype);
+                  if (!fluidstate.isEmpty() && ItemBlockRenderTypes.canRenderInLayer(fluidstate, rendertype)) {
                      BufferBuilder bufferbuilder = p_112870_.builder(rendertype);
                      if (p_112869_.hasLayer.add(rendertype)) {
                         RenderChunk.this.beginLayer(bufferbuilder);
@@ -604,8 +_,8 @@
                      }
                   }
 
-                  if (blockstate.getRenderShape() != RenderShape.INVISIBLE) {
-                     RenderType rendertype1 = ItemBlockRenderTypes.getChunkRenderType(blockstate);
+                  if (blockstate.getRenderShape() != RenderShape.INVISIBLE && ItemBlockRenderTypes.canRenderInLayer(blockstate, rendertype)) {
+                     RenderType rendertype1 = rendertype;
                      BufferBuilder bufferbuilder2 = p_112870_.builder(rendertype1);
                      if (p_112869_.hasLayer.add(rendertype1)) {
                         RenderChunk.this.beginLayer(bufferbuilder2);
@@ -613,14 +_,16 @@
 
                      posestack.pushPose();
                      posestack.translate((double)(blockpos2.getX() & 15), (double)(blockpos2.getY() & 15), (double)(blockpos2.getZ() & 15));
-                     if (blockrenderdispatcher.renderBatched(blockstate, blockpos2, renderchunkregion, posestack, bufferbuilder2, true, random)) {
+                     if (blockrenderdispatcher.renderBatched(blockstate, blockpos2, renderchunkregion, posestack, bufferbuilder2, true, random, modelData)) {
                         p_112869_.isCompletelyEmpty = false;
                         p_112869_.hasBlocks.add(rendertype1);
                      }
 
                      posestack.popPose();
                   }
+                  }
                }
+               net.minecraftforge.client.ForgeHooksClient.setRenderType(null);
 
                if (p_112869_.hasBlocks.contains(RenderType.translucent())) {
                   BufferBuilder bufferbuilder1 = p_112870_.builder(RenderType.translucent());
@@ -639,10 +_,10 @@
          private <E extends BlockEntity> void handleBlockEntity(ChunkRenderDispatcher.CompiledChunk p_112878_, Set<BlockEntity> p_112879_, E p_112880_) {
             BlockEntityRenderer<E> blockentityrenderer = Minecraft.getInstance().getBlockEntityRenderDispatcher().getRenderer(p_112880_);
             if (blockentityrenderer != null) {
-               p_112878_.renderableBlockEntities.add(p_112880_);
                if (blockentityrenderer.shouldRenderOffScreen(p_112880_)) {
                   p_112879_.add(p_112880_);
                }
+               else p_112878_.renderableBlockEntities.add(p_112880_); //FORGE: Fix MC-112730
             }
 
          }
@@ -660,8 +_,13 @@
       class ResortTransparencyTask extends ChunkRenderDispatcher.RenderChunk.ChunkCompileTask {
          private final ChunkRenderDispatcher.CompiledChunk compiledChunk;
 
+         @Deprecated
          public ResortTransparencyTask(double p_112889_, ChunkRenderDispatcher.CompiledChunk p_112890_) {
-            super(p_112889_, true);
+            this(null, p_112889_, p_112890_);
+         }
+
+         public ResortTransparencyTask(@Nullable net.minecraft.world.level.ChunkPos pos, double p_112889_, ChunkRenderDispatcher.CompiledChunk p_112890_) {
+            super(pos, p_112889_, true);
             this.compiledChunk = p_112890_;
          }
 
