--- a/net/minecraft/client/gui/screens/inventory/LoomScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/LoomScreen.java
@@ -125,20 +_,20 @@
          int l2 = j + 13;
          int l = this.startIndex + 16;
 
-         for(int i1 = this.startIndex; i1 < l && i1 < BannerPattern.COUNT - BannerPattern.PATTERN_ITEM_COUNT; ++i1) {
+         for(int i1 = this.startIndex; i1 < l && i1 < net.minecraftforge.common.ForgeHooks.getNonPatternItemCount(); ++i1) {
             int j1 = i1 - this.startIndex;
             int k1 = j2 + j1 % 4 * 14;
             int l1 = l2 + j1 / 4 * 14;
             RenderSystem.setShaderTexture(0, BG_LOCATION);
             int i2 = this.imageHeight;
-            if (i1 == this.menu.getSelectedBannerPatternIndex()) {
+            if (net.minecraftforge.common.ForgeHooks.getActualPatternIndex(i1) == this.menu.getSelectedBannerPatternIndex()) {
                i2 += 14;
             } else if (p_99101_ >= k1 && p_99102_ >= l1 && p_99101_ < k1 + 14 && p_99102_ < l1 + 14) {
                i2 += 28;
             }
 
             this.blit(p_99099_, k1, l1, 0, i2, 14, 14);
-            this.renderPattern(i1, k1, l1);
+            this.renderPattern(net.minecraftforge.common.ForgeHooks.getActualPatternIndex(i1), k1, l1);
          }
       } else if (this.displaySpecialPattern) {
          int k2 = i + 60;
@@ -209,7 +_,7 @@
          int j = i + 56;
          this.scrollOffs = ((float)p_99088_ - (float)i - 7.5F) / ((float)(j - i) - 15.0F);
          this.scrollOffs = Mth.clamp(this.scrollOffs, 0.0F, 1.0F);
-         int k = TOTAL_PATTERN_ROWS - 4;
+         int k = net.minecraftforge.common.ForgeHooks.getTotalPatternRows() - 4;
          int l = (int)((double)(this.scrollOffs * (float)k) + 0.5D);
          if (l < 0) {
             l = 0;
@@ -224,7 +_,7 @@
 
    public boolean mouseScrolled(double p_99079_, double p_99080_, double p_99081_) {
       if (this.displayPatterns) {
-         int i = TOTAL_PATTERN_ROWS - 4;
+         int i = net.minecraftforge.common.ForgeHooks.getTotalPatternRows() - 4;
          float f = (float)p_99081_ / (float)i;
          this.scrollOffs = Mth.clamp(this.scrollOffs - f, 0.0F, 1.0F);
          this.startIndex = 1 + (int)(this.scrollOffs * (float)i + 0.5F) * 4;
