--- a/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
+++ b/net/minecraft/client/gui/screens/recipebook/RecipeBookComponent.java
@@ -107,7 +_,7 @@
       this.initFilterButtonTextures();
       this.tabButtons.clear();
 
-      for(RecipeBookCategories recipebookcategories : RecipeBookCategories.getCategories(this.menu.getRecipeBookType())) {
+      for(RecipeBookCategories recipebookcategories : this.menu.getRecipeBookCategories()) {
          this.tabButtons.add(new RecipeBookTabButton(recipebookcategories));
       }
 
@@ -319,7 +_,7 @@
       }
 
       if (itemstack != null && this.minecraft.screen != null) {
-         this.minecraft.screen.renderComponentTooltip(p_100375_, this.minecraft.screen.getTooltipFromItem(itemstack), p_100378_, p_100379_);
+         this.minecraft.screen.renderComponentTooltip(p_100375_, this.minecraft.screen.getTooltipFromItem(itemstack), p_100378_, p_100379_, itemstack);
       }
 
    }
