--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -360,6 +_,7 @@
 
       this.minecraft.debugRenderer.clear();
       this.minecraft.player.resetPos();
+      net.minecraftforge.client.ForgeHooksClient.firePlayerLogin(this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().getConnection());
       int i = p_105030_.playerId();
       this.minecraft.player.setId(i);
       this.level.addPlayer(i, this.minecraft.player);
@@ -371,6 +_,7 @@
       this.minecraft.player.setShowDeathScreen(p_105030_.showDeathScreen());
       this.minecraft.gameMode.setLocalMode(p_105030_.gameType(), p_105030_.previousGameType());
       this.minecraft.options.setServerRenderDistance(p_105030_.chunkRadius());
+      net.minecraftforge.network.NetworkHooks.sendMCRegistryPackets(connection, "PLAY_TO_SERVER");
       this.minecraft.options.broadcastOptions();
       this.connection.send(new ServerboundCustomPayloadPacket(ServerboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf(ClientBrandRetriever.getClientModName())));
       this.minecraft.getGame().onStartGameSession();
@@ -712,7 +_,9 @@
 
    public void handleChat(ClientboundChatPacket p_104986_) {
       PacketUtils.ensureRunningOnSameThread(p_104986_, this, this.minecraft);
-      this.minecraft.gui.handleChat(p_104986_.getType(), p_104986_.getMessage(), p_104986_.getSender());
+      net.minecraft.network.chat.Component message = net.minecraftforge.event.ForgeEventFactory.onClientChat(p_104986_.getType(), p_104986_.getMessage(), p_104986_.getSender());
+      if (message == null) return;
+      this.minecraft.gui.handleChat(p_104986_.getType(), message, p_104986_.getSender());
    }
 
    public void handleAnimate(ClientboundAnimatePacket p_104968_) {
@@ -894,8 +_,10 @@
          localplayer1.getAttributes().assignValues(localplayer.getAttributes());
       }
 
+      localplayer1.updateSyncFields(localplayer); // Forge: fix MC-10657
       localplayer1.resetPos();
       localplayer1.setServerBrand(s);
+      net.minecraftforge.client.ForgeHooksClient.firePlayerRespawn(this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.getConnection());
       this.level.addPlayer(i, localplayer1);
       localplayer1.setYRot(-180.0F);
       localplayer1.input = new KeyboardInput(this.minecraft.options);
@@ -998,10 +_,7 @@
       PacketUtils.ensureRunningOnSameThread(p_104976_, this, this.minecraft);
       BlockPos blockpos = p_104976_.getPos();
       this.minecraft.level.getBlockEntity(blockpos, p_104976_.getType()).ifPresent((p_205557_) -> {
-         CompoundTag compoundtag = p_104976_.getTag();
-         if (compoundtag != null) {
-            p_205557_.load(compoundtag);
-         }
+         p_205557_.onDataPacket(connection, p_104976_);
 
          if (p_205557_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
             ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
@@ -1147,6 +_,7 @@
    public void handleCommands(ClientboundCommandsPacket p_104990_) {
       PacketUtils.ensureRunningOnSameThread(p_104990_, this, this.minecraft);
       this.commands = new CommandDispatcher<>(p_104990_.getRoot());
+      this.commands = net.minecraftforge.client.ClientCommandHandler.mergeServerCommands(this.commands);
    }
 
    public void handleStopSoundEvent(ClientboundStopSoundPacket p_105116_) {
@@ -1168,6 +_,7 @@
       clientrecipebook.setupCollections(this.recipeManager.getRecipes());
       clientrecipebook.getCollections().forEach(mutablesearchtree::add);
       mutablesearchtree.refresh();
+      net.minecraftforge.client.ForgeHooksClient.onRecipesUpdated(this.recipeManager);
    }
 
    public void handleLookAt(ClientboundPlayerLookAtPacket p_105054_) {
@@ -1245,7 +_,7 @@
       PacketUtils.ensureRunningOnSameThread(p_105130_, this, this.minecraft);
       Entity entity = this.level.getEntity(p_105130_.getEntityId());
       if (entity instanceof LivingEntity) {
-         MobEffect mobeffect = MobEffect.byId(p_105130_.getEffectId());
+         MobEffect mobeffect = MobEffect.byId(p_105130_.getEffectId() & 0xFF);
          if (mobeffect != null) {
             MobEffectInstance mobeffectinstance = new MobEffectInstance(mobeffect, p_105130_.getEffectDurationTicks(), p_105130_.getEffectAmplifier(), p_105130_.isEffectAmbient(), p_105130_.isEffectVisible(), p_105130_.effectShowsIcon());
             mobeffectinstance.setNoCounter(p_105130_.isSuperLongDuration());
@@ -1262,6 +_,7 @@
       }
 
       this.minecraft.getSearchTree(SearchRegistry.CREATIVE_TAGS).refresh();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent(this.registryAccess, true, connection.isMemoryConnection()));
    }
 
    private <T> void updateTagsForRegistry(ResourceKey<? extends Registry<? extends T>> p_205561_, TagNetworkSerialization.NetworkPayload p_205562_) {
@@ -1815,7 +_,8 @@
             int i5 = friendlybytebuf.readVarInt();
             this.minecraft.debugRenderer.gameEventListenerRenderer.trackListener(positionsource, i5);
          } else {
-            LOGGER.warn("Unknown custom packed identifier: {}", (Object)resourcelocation);
+            if (!net.minecraftforge.network.NetworkHooks.onCustomPayload(p_105004_, this.connection))
+            LOGGER.warn("Unknown custom packet identifier: {}", (Object)resourcelocation);
          }
       } finally {
          if (friendlybytebuf != null) {
