--- a/net/minecraft/world/damagesource/IndirectEntityDamageSource.java
+++ b/net/minecraft/world/damagesource/IndirectEntityDamageSource.java
@@ -33,4 +_,11 @@
       String s1 = s + ".item";
       return !itemstack.isEmpty() && itemstack.hasCustomHoverName() ? new TranslatableComponent(s1, p_19410_.getDisplayName(), component, itemstack.getDisplayName()) : new TranslatableComponent(s, p_19410_.getDisplayName(), component);
    }
+
+   // CraftBukkit start
+   public Entity getProximateDamageSource() {
+      return super.getEntity();
+   }
+   // CraftBukkit end
+
 }
