--- a/net/minecraft/world/effect/MobEffectInstance.java
+++ b/net/minecraft/world/effect/MobEffectInstance.java
@@ -7,7 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import org.slf4j.Logger;
 
-public class MobEffectInstance implements Comparable<MobEffectInstance> {
+public class MobEffectInstance implements Comparable<MobEffectInstance>, net.minecraftforge.common.extensions.IForgeMobEffectInstance {
    private static final Logger LOGGER = LogUtils.getLogger();
    private final MobEffect effect;
    private int duration;
@@ -60,6 +_,7 @@
       this.ambient = p_19549_.ambient;
       this.visible = p_19549_.visible;
       this.showIcon = p_19549_.showIcon;
+      this.curativeItems = p_19549_.curativeItems == null ? null : new java.util.ArrayList<net.minecraft.world.item.ItemStack>(p_19549_.curativeItems);
    }
 
    public boolean update(MobEffectInstance p_19559_) {
@@ -108,7 +_,7 @@
    }
 
    public MobEffect getEffect() {
-      return this.effect;
+      return this.effect == null ? null : this.effect.delegate.get();
    }
 
    public int getDuration() {
@@ -206,6 +_,7 @@
 
    public CompoundTag save(CompoundTag p_19556_) {
       p_19556_.putByte("Id", (byte)MobEffect.getId(this.getEffect()));
+      net.minecraftforge.common.ForgeHooks.saveMobEffect(p_19556_, "forge:id", this.getEffect());
       this.writeDetailsTo(p_19556_);
       return p_19556_;
    }
@@ -221,13 +_,15 @@
          this.hiddenEffect.save(compoundtag);
          p_19568_.put("HiddenEffect", compoundtag);
       }
+      writeCurativeItems(p_19568_);
 
    }
 
    @Nullable
    public static MobEffectInstance load(CompoundTag p_19561_) {
-      int i = p_19561_.getByte("Id");
+      int i = p_19561_.getByte("Id") & 0xFF;
       MobEffect mobeffect = MobEffect.byId(i);
+      mobeffect = net.minecraftforge.common.ForgeHooks.loadMobEffect(p_19561_, "forge:id", mobeffect);
       return mobeffect == null ? null : loadSpecifiedEffect(mobeffect, p_19561_);
    }
 
@@ -250,7 +_,7 @@
          mobeffectinstance = loadSpecifiedEffect(p_19546_, p_19547_.getCompound("HiddenEffect"));
       }
 
-      return new MobEffectInstance(p_19546_, j, i < 0 ? 0 : i, flag, flag1, flag2, mobeffectinstance);
+      return readCurativeItems(new MobEffectInstance(p_19546_, j, i < 0 ? 0 : i, flag, flag1, flag2, mobeffectinstance), p_19547_);
    }
 
    public void setNoCounter(boolean p_19563_) {
@@ -263,6 +_,32 @@
 
    public int compareTo(MobEffectInstance p_19566_) {
       int i = 32147;
-      return (this.getDuration() <= 32147 || p_19566_.getDuration() <= 32147) && (!this.isAmbient() || !p_19566_.isAmbient()) ? ComparisonChain.start().compare(this.isAmbient(), p_19566_.isAmbient()).compare(this.getDuration(), p_19566_.getDuration()).compare(this.getEffect().getColor(), p_19566_.getEffect().getColor()).result() : ComparisonChain.start().compare(this.isAmbient(), p_19566_.isAmbient()).compare(this.getEffect().getColor(), p_19566_.getEffect().getColor()).result();
+      return (this.getDuration() <= 32147 || p_19566_.getDuration() <= 32147) && (!this.isAmbient() || !p_19566_.isAmbient()) ? ComparisonChain.start().compare(this.isAmbient(), p_19566_.isAmbient()).compare(this.getDuration(), p_19566_.getDuration()).compare(this.getEffect().getSortOrder(this), p_19566_.getEffect().getSortOrder(this)).result() : ComparisonChain.start().compare(this.isAmbient(), p_19566_.isAmbient()).compare(this.getEffect().getSortOrder(this), p_19566_.getEffect().getSortOrder(this)).result();
+   }
+
+   //======================= FORGE START ===========================
+   private java.util.List<net.minecraft.world.item.ItemStack> curativeItems;
+
+   @Override
+   public java.util.List<net.minecraft.world.item.ItemStack> getCurativeItems() {
+      if (this.curativeItems == null) //Lazy load this so that we don't create a circular dep on Items.
+         this.curativeItems = getEffect().getCurativeItems();
+      return this.curativeItems;
+   }
+   @Override
+   public void setCurativeItems(java.util.List<net.minecraft.world.item.ItemStack> curativeItems) {
+      this.curativeItems = curativeItems;
+   }
+   private static MobEffectInstance readCurativeItems(MobEffectInstance effect, CompoundTag nbt) {
+      if (nbt.contains("CurativeItems", net.minecraft.nbt.Tag.TAG_LIST)) {
+         java.util.List<net.minecraft.world.item.ItemStack> items = new java.util.ArrayList<net.minecraft.world.item.ItemStack>();
+         net.minecraft.nbt.ListTag list = nbt.getList("CurativeItems", net.minecraft.nbt.Tag.TAG_COMPOUND);
+         for (int i = 0; i < list.size(); i++) {
+            items.add(net.minecraft.world.item.ItemStack.of(list.getCompound(i)));
+         }
+         effect.setCurativeItems(items);
+      }
+
+      return effect;
    }
 }
