--- a/com/mojang/math/Matrix4f.java
+++ b/com/mojang/math/Matrix4f.java
@@ -559,4 +_,40 @@
       matrix4f.m23 = p_27656_;
       return matrix4f;
    }
+
+    // Forge start
+    public Matrix4f(float[] values) {
+        m00 = values[0];
+        m01 = values[1];
+        m02 = values[2];
+        m03 = values[3];
+        m10 = values[4];
+        m11 = values[5];
+        m12 = values[6];
+        m13 = values[7];
+        m20 = values[8];
+        m21 = values[9];
+        m22 = values[10];
+        m23 = values[11];
+        m30 = values[12];
+        m31 = values[13];
+        m32 = values[14];
+        m33 = values[15];
+    }
+
+    public void multiplyBackward(Matrix4f other) {
+        Matrix4f copy = other.copy();
+        copy.multiply(this);
+        this.load(copy);
+    }
+
+    public void setTranslation(float x, float y, float z) {
+        this.m00 = 1.0F;
+        this.m11 = 1.0F;
+        this.m22 = 1.0F;
+        this.m33 = 1.0F;
+        this.m03 = x;
+        this.m13 = y;
+        this.m23 = z;
+    }
 }
