--- a/com/mojang/math/Vector4f.java
+++ b/com/mojang/math/Vector4f.java
@@ -147,4 +_,15 @@
    public String toString() {
       return "[" + this.x + ", " + this.y + ", " + this.z + ", " + this.w + "]";
    }
+
+    public void set(float[] values) {
+        this.x = values[0];
+        this.y = values[1];
+        this.z = values[2];
+        this.w = values[3];
+    }
+    public void setX(float x) { this.x = x; }
+    public void setY(float y) { this.y = y; }
+    public void setZ(float z) { this.z = z; }
+    public void setW(float w) { this.w = w; }
 }
