--- a/com/mojang/math/Vector3f.java
+++ b/com/mojang/math/Vector3f.java
@@ -138,7 +_,7 @@
 
    public boolean normalize() {
       float f = this.x * this.x + this.y * this.y + this.z * this.z;
-      if ((double)f < 1.0E-5D) {
+      if (f < Float.MIN_NORMAL) { //Forge: Fix MC-239212
          return false;
       } else {
          float f1 = Mth.fastInvSqrt(f);
@@ -207,4 +_,16 @@
    public String toString() {
       return "[" + this.x + ", " + this.y + ", " + this.z + "]";
    }
+
+    public Vector3f(float[] values) {
+        set(values);
+    }
+    public void set(float[] values) {
+        this.x = values[0];
+        this.y = values[1];
+        this.z = values[2];
+    }
+    public void setX(float x) { this.x = x; }
+    public void setY(float y) { this.y = y; }
+    public void setZ(float z) { this.z = z; }
 }
