--- a/net/minecraft/world/level/block/grower/AbstractMegaTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractMegaTreeGrower.java
@@ -30,10 +_,12 @@
 
    public boolean placeMega(ServerLevel p_59991_, ChunkGenerator p_59992_, BlockPos p_59993_, BlockState p_59994_, Random p_59995_, int p_59996_, int p_59997_) {
       Holder<? extends ConfiguredFeature<?, ?>> holder = this.getConfiguredMegaFeature(p_59995_);
-      if (holder == null) {
+      net.minecraftforge.event.world.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_59991_, p_59995_, p_59993_, holder);
+      if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY) || holder == null) {
          return false;
       } else {
-         ConfiguredFeature<?, ?> configuredfeature = holder.value();
+         setTreeType(holder); // CraftBukkit
+         ConfiguredFeature<?, ?> configuredfeature = event.getFeature().value();
          BlockState blockstate = Blocks.AIR.defaultBlockState();
          p_59991_.setBlock(p_59993_.offset(p_59996_, 0, p_59997_), blockstate, 4);
          p_59991_.setBlock(p_59993_.offset(p_59996_ + 1, 0, p_59997_), blockstate, 4);
