--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -18,10 +_,12 @@
 
    public boolean growTree(ServerLevel p_60006_, ChunkGenerator p_60007_, BlockPos p_60008_, BlockState p_60009_, Random p_60010_) {
       Holder<? extends ConfiguredFeature<?, ?>> holder = this.getConfiguredFeature(p_60010_, this.hasFlowers(p_60006_, p_60008_));
-      if (holder == null) {
+      net.minecraftforge.event.world.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_60006_, p_60010_, p_60008_, holder);
+      if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY) || holder == null) {
          return false;
       } else {
-         ConfiguredFeature<?, ?> configuredfeature = holder.value();
+         setTreeType(holder); // CraftBukkit
+         ConfiguredFeature<?, ?> configuredfeature = event.getFeature().value();
          p_60006_.setBlock(p_60008_, Blocks.AIR.defaultBlockState(), 4);
          if (configuredfeature.place(p_60006_, p_60007_, p_60010_, p_60008_)) {
             return true;
@@ -41,4 +_,48 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void setTreeType(Holder<? extends ConfiguredFeature<?, ?>> worldgentreeabstract) {
+      if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.OAK || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.OAK_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.HUGE_RED_MUSHROOM) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.HUGE_BROWN_MUSHROOM) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.JUNGLE_TREE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.JUNGLE_TREE_NO_VINE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.PINE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.SPRUCE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.ACACIA) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.BIRCH || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.BIRCH_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.SUPER_BIRCH_BEES_0002) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.SWAMP_OAK) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.FANCY_OAK || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.FANCY_OAK_BEES_005) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.JUNGLE_BUSH) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.DARK_OAK) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MEGA_SPRUCE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MEGA_PINE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.MEGA_JUNGLE_TREE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.AZALEA_TREE) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+      } else {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MODDED; //Magma - default to modded
+      }
+   }
+// CraftBukkit end
 }
