--- a/net/minecraft/world/level/block/grower/AbstractMegaTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractMegaTreeGrower.java
@@ -33,6 +_,7 @@
       if (holder == null) {
          return false;
       } else {
+         setTreeType(holder); // CraftBukkit
          ConfiguredFeature<?, ?> configuredfeature = holder.m_203334_();
          BlockState blockstate = Blocks.f_50016_.m_49966_();
          p_59991_.m_7731_(p_59993_.m_142082_(p_59996_, 0, p_59997_), blockstate, 4);
