--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -18,10 +_,12 @@
 
    public boolean m_6334_(ServerLevel p_60006_, ChunkGenerator p_60007_, BlockPos p_60008_, BlockState p_60009_, Random p_60010_) {
       Holder<? extends ConfiguredFeature<?, ?>> holder = this.m_203525_(p_60010_, this.m_60011_(p_60006_, p_60008_));
-      if (holder == null) {
+      net.minecraftforge.event.world.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_60006_, p_60010_, p_60008_, holder);
+      if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY) || holder == null) {
          return false;
       } else {
-         ConfiguredFeature<?, ?> configuredfeature = holder.m_203334_();
+         setTreeType(holder); // CraftBukkit
+         ConfiguredFeature<?, ?> configuredfeature = event.getFeature().m_203334_();
          p_60006_.m_7731_(p_60008_, Blocks.f_50016_.m_49966_(), 4);
          if (configuredfeature.m_65385_(p_60006_, p_60007_, p_60010_, p_60008_)) {
             return true;
@@ -41,4 +_,48 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void setTreeType(Holder<? extends ConfiguredFeature<?, ?>> worldgentreeabstract) {
+      if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195123_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195142_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195122_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195121_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195129_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195131_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195128_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195127_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195126_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195125_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195108_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195135_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195137_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195130_ || worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195111_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195138_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195124_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195133_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195134_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195132_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+      } else if (worldgentreeabstract == net.minecraft.data.worldgen.features.TreeFeatures.f_195139_) {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+      } else {
+         net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MODDED; //Magma - default to modded
+      }
+   }
+// CraftBukkit end
 }
