--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -113,7 +_,7 @@
       BlockPathTypes blockpathtypes = this.getCachedBlockType(this.mob, p_77641_.x, p_77641_.y + 1, p_77641_.z);
       BlockPathTypes blockpathtypes1 = this.getCachedBlockType(this.mob, p_77641_.x, p_77641_.y, p_77641_.z);
       if (this.mob.getPathfindingMalus(blockpathtypes) >= 0.0F && blockpathtypes1 != BlockPathTypes.STICKY_HONEY) {
-         j = Mth.floor(Math.max(1.0F, this.mob.maxUpStep));
+         j = Mth.floor(Math.max(1.0F, this.mob.getStepHeight()));
       }
 
       double d0 = this.getFloorLevel(new BlockPos(p_77641_.x, p_77641_.y, p_77641_.z));
@@ -476,6 +_,8 @@
 
    protected static BlockPathTypes getBlockPathTypeRaw(BlockGetter p_77644_, BlockPos p_77645_) {
       BlockState blockstate = p_77644_.getBlockState(p_77645_);
+      BlockPathTypes type = blockstate.getBlockPathType(p_77644_, p_77645_);
+      if (type != null) return type;
       Block block = blockstate.getBlock();
       Material material = blockstate.getMaterial();
       if (blockstate.isAir()) {
