--- a/net/minecraft/server/commands/BanIpCommands.java
+++ b/net/minecraft/server/commands/BanIpCommands.java
@@ -26,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_136528_) {
       p_136528_.register(Commands.literal("ban-ip").requires((p_136532_) -> {
-         return p_136532_.hasPermission(3);
+         return p_136532_.hasPermission(3,"minecraft.command.ban-ip");
       }).then(Commands.argument("target", StringArgumentType.word()).executes((p_136538_) -> {
          return banIpOrName(p_136538_.getSource(), StringArgumentType.getString(p_136538_, "target"), (Component)null);
       }).then(Commands.argument("reason", MessageArgument.message()).executes((p_136530_) -> {
