--- a/net/minecraft/server/commands/BanListCommands.java
+++ b/net/minecraft/server/commands/BanListCommands.java
@@ -13,7 +_,7 @@
 public class BanListCommands {
    public static void register(CommandDispatcher<CommandSourceStack> p_136544_) {
       p_136544_.register(Commands.literal("banlist").requires((p_136548_) -> {
-         return p_136548_.hasPermission(3);
+         return p_136548_.hasPermission(3,"minecraft.command.banlist");
       }).executes((p_136555_) -> {
          PlayerList playerlist = p_136555_.getSource().getServer().getPlayerList();
          return showList(p_136555_.getSource(), Lists.newArrayList(Iterables.concat(playerlist.getBans().getEntries(), playerlist.getIpBans().getEntries())));
