--- a/net/minecraft/server/commands/BanPlayerCommands.java
+++ b/net/minecraft/server/commands/BanPlayerCommands.java
@@ -23,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_136559_) {
       p_136559_.register(Commands.literal("ban").requires((p_136563_) -> {
-         return p_136563_.hasPermission(3);
+         return p_136563_.hasPermission(3,"minecraft.command.ban");
       }).then(Commands.argument("targets", GameProfileArgument.gameProfile()).executes((p_136569_) -> {
          return banPlayers(p_136569_.getSource(), GameProfileArgument.getGameProfiles(p_136569_, "targets"), (Component)null);
       }).then(Commands.argument("reason", MessageArgument.message()).executes((p_136561_) -> {
