--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -38,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_136906_) {
       p_136906_.register(Commands.literal("debug").requires((p_180073_) -> {
-         return p_180073_.hasPermission(3);
+         return p_180073_.hasPermission(3,"minecraft.command.debug");
       }).then(Commands.literal("start").executes((p_180069_) -> {
          return start(p_180069_.getSource());
       })).then(Commands.literal("stop").executes((p_136918_) -> {
@@ -216,6 +_,11 @@
 
       public boolean alwaysAccepts() {
          return true;
+      }
+
+      @Override
+      public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+         return null;
       }
    }
 }
