--- a/net/minecraft/server/commands/DefaultGameModeCommands.java
+++ b/net/minecraft/server/commands/DefaultGameModeCommands.java
@@ -12,7 +_,7 @@
 public class DefaultGameModeCommands {
    public static void register(CommandDispatcher<CommandSourceStack> p_136927_) {
       LiteralArgumentBuilder<CommandSourceStack> literalargumentbuilder = Commands.literal("defaultgamemode").requires((p_136929_) -> {
-         return p_136929_.hasPermission(2);
+         return p_136929_.hasPermission(2,"minecraft.command.defaultgamemode");
       });
 
       for(GameType gametype : GameType.values()) {
