--- a/net/minecraft/server/commands/DeOpCommands.java
+++ b/net/minecraft/server/commands/DeOpCommands.java
@@ -17,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_136889_) {
       p_136889_.register(Commands.literal("deop").requires((p_136896_) -> {
-         return p_136896_.hasPermission(3);
+         return p_136896_.hasPermission(3,"minecraft.command.deop");
       }).then(Commands.argument("targets", GameProfileArgument.gameProfile()).suggests((p_136893_, p_136894_) -> {
          return SharedSuggestionProvider.suggest(p_136893_.getSource().getServer().getPlayerList().getOpNames(), p_136894_);
       }).executes((p_136891_) -> {
