--- a/net/minecraft/server/commands/EmoteCommands.java
+++ b/net/minecraft/server/commands/EmoteCommands.java
@@ -15,7 +_,9 @@
 
 public class EmoteCommands {
    public static void register(CommandDispatcher<CommandSourceStack> p_136986_) {
-      p_136986_.register(Commands.literal("me").then(Commands.argument("action", StringArgumentType.greedyString()).executes((p_136988_) -> {
+      p_136986_.register(Commands.literal("me")
+              .requires(commandSourceStack -> commandSourceStack.hasPermission(0,"minecraft.command.me"))
+              .then(Commands.argument("action", StringArgumentType.greedyString()).executes((p_136988_) -> {
          String s = StringArgumentType.getString(p_136988_, "action");
          Entity entity = p_136988_.getSource().getEntity();
          MinecraftServer minecraftserver = p_136988_.getSource().getServer();
