--- a/net/minecraft/server/commands/EnchantCommand.java
+++ b/net/minecraft/server/commands/EnchantCommand.java
@@ -35,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_137009_) {
       p_137009_.register(Commands.literal("enchant").requires((p_137013_) -> {
-         return p_137013_.hasPermission(2);
+         return p_137013_.hasPermission(2,"minecraft.command.enchant");
       }).then(Commands.argument("targets", EntityArgument.entities()).then(Commands.argument("enchantment", ItemEnchantmentArgument.enchantment()).executes((p_137025_) -> {
          return enchant(p_137025_.getSource(), EntityArgument.getEntities(p_137025_, "targets"), ItemEnchantmentArgument.getEnchantment(p_137025_, "enchantment"), 1);
       }).then(Commands.argument("level", IntegerArgumentType.integer(0)).executes((p_137011_) -> {
