--- a/net/minecraft/server/commands/ExperienceCommand.java
+++ b/net/minecraft/server/commands/ExperienceCommand.java
@@ -22,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_137307_) {
       LiteralCommandNode<CommandSourceStack> literalcommandnode = p_137307_.register(Commands.literal("experience").requires((p_137324_) -> {
-         return p_137324_.hasPermission(2);
+         return p_137324_.hasPermission(2,"minecraft.command.xp");
       }).then(Commands.literal("add").then(Commands.argument("targets", EntityArgument.players()).then(Commands.argument("amount", IntegerArgumentType.integer()).executes((p_137341_) -> {
          return addExperience(p_137341_.getSource(), EntityArgument.getPlayers(p_137341_, "targets"), IntegerArgumentType.getInteger(p_137341_, "amount"), ExperienceCommand.Type.POINTS);
       }).then(Commands.literal("points").executes((p_137339_) -> {
@@ -41,7 +_,7 @@
          return queryExperience(p_137309_.getSource(), EntityArgument.getPlayer(p_137309_, "targets"), ExperienceCommand.Type.LEVELS);
       })))));
       p_137307_.register(Commands.literal("xp").requires((p_137311_) -> {
-         return p_137311_.hasPermission(2);
+         return p_137311_.hasPermission(2,"minecraft.command.xp");
       }).redirect(literalcommandnode));
    }
 
