--- a/net/minecraft/server/commands/FillCommand.java
+++ b/net/minecraft/server/commands/FillCommand.java
@@ -36,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_137380_) {
       p_137380_.register(Commands.literal("fill").requires((p_137384_) -> {
-         return p_137384_.hasPermission(2);
+         return p_137384_.hasPermission(2,"minecraft.command.fill");
       }).then(Commands.argument("from", BlockPosArgument.blockPos()).then(Commands.argument("to", BlockPosArgument.blockPos()).then(Commands.argument("block", BlockStateArgument.block()).executes((p_137405_) -> {
          return fillBlocks(p_137405_.getSource(), BoundingBox.fromCorners(BlockPosArgument.getLoadedBlockPos(p_137405_, "from"), BlockPosArgument.getLoadedBlockPos(p_137405_, "to")), BlockStateArgument.getBlock(p_137405_, "block"), FillCommand.Mode.REPLACE, (Predicate<BlockInWorld>)null);
       }).then(Commands.literal("replace").executes((p_137403_) -> {
