--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -22,7 +_,7 @@
 
    public static void m_137772_(CommandDispatcher<CommandSourceStack> p_137773_) {
       p_137773_.register(Commands.m_82127_("give").requires((p_137777_) -> {
-         return p_137777_.m_6761_(2);
+         return p_137777_.hasPermission(2,"minecraft.command.give");
       }).then(Commands.m_82129_("targets", EntityArgument.m_91470_()).then(Commands.m_82129_("item", ItemArgument.m_120960_()).executes((p_137784_) -> {
          return m_137778_(p_137784_.getSource(), ItemArgument.m_120963_(p_137784_, "item"), EntityArgument.m_91477_(p_137784_, "targets"), 1);
       }).then(Commands.m_82129_("count", IntegerArgumentType.integer(1)).executes((p_137775_) -> {
@@ -47,7 +_,7 @@
                boolean flag = serverplayer.m_150109_().m_36054_(itemstack);
                if (flag && itemstack.m_41619_()) {
                   itemstack.m_41764_(1);
-                  ItemEntity itementity1 = serverplayer.m_36176_(itemstack, false);
+                  ItemEntity itementity1 = serverplayer.drop(itemstack, false, false, false); // SPIGOT-2942: Add boolean to call event
                   if (itementity1 != null) {
                      itementity1.m_32065_();
                   }
