--- a/net/minecraft/server/commands/KickCommand.java
+++ b/net/minecraft/server/commands/KickCommand.java
@@ -13,7 +_,7 @@
 public class KickCommand {
    public static void m_137795_(CommandDispatcher<CommandSourceStack> p_137796_) {
       p_137796_.register(Commands.m_82127_("kick").requires((p_137800_) -> {
-         return p_137800_.m_6761_(3);
+         return p_137800_.hasPermission(3,"minecraft.command.kick");
       }).then(Commands.m_82129_("targets", EntityArgument.m_91470_()).executes((p_137806_) -> {
          return m_137801_(p_137806_.getSource(), EntityArgument.m_91477_(p_137806_, "targets"), new TranslatableComponent("multiplayer.disconnect.kicked"));
       }).then(Commands.m_82129_("reason", MessageArgument.m_96832_()).executes((p_137798_) -> {
