--- a/net/minecraft/server/commands/KickCommand.java
+++ b/net/minecraft/server/commands/KickCommand.java
@@ -13,7 +_,7 @@
 public class KickCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_137796_) {
       p_137796_.register(Commands.literal("kick").requires((p_137800_) -> {
-         return p_137800_.hasPermission(3);
+         return p_137800_.hasPermission(3,"minecraft.command.kick");
       }).then(Commands.argument("targets", EntityArgument.players()).executes((p_137806_) -> {
          return kickPlayers(p_137806_.getSource(), EntityArgument.getPlayers(p_137806_, "targets"), new TranslatableComponent("multiplayer.disconnect.kicked"));
       }).then(Commands.argument("reason", MessageArgument.message()).executes((p_137798_) -> {
