--- a/net/minecraft/server/commands/KillCommand.java
+++ b/net/minecraft/server/commands/KillCommand.java
@@ -12,7 +_,7 @@
 public class KillCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_137808_) {
       p_137808_.register(Commands.literal("kill").requires((p_137812_) -> {
-         return p_137812_.hasPermission(2);
+         return p_137812_.hasPermission(2,"minecraft.command.kill");
       }).executes((p_137817_) -> {
          return kill(p_137817_.getSource(), ImmutableList.of(p_137817_.getSource().getEntityOrException()));
       }).then(Commands.argument("targets", EntityArgument.entities()).executes((p_137810_) -> {
