--- a/net/minecraft/server/commands/MsgCommand.java
+++ b/net/minecraft/server/commands/MsgCommand.java
@@ -18,11 +_,17 @@
 
 public class MsgCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_138061_) {
-      LiteralCommandNode<CommandSourceStack> literalcommandnode = p_138061_.register(Commands.literal("msg").then(Commands.argument("targets", EntityArgument.players()).then(Commands.argument("message", MessageArgument.message()).executes((p_138063_) -> {
+      LiteralCommandNode<CommandSourceStack> literalcommandnode = p_138061_.register(Commands.literal("msg")
+                      .requires(commandSourceStack -> commandSourceStack.hasPermission(0,"minecraft.command.msg"))
+              .then(Commands.argument("targets", EntityArgument.players()).then(Commands.argument("message", MessageArgument.message()).executes((p_138063_) -> {
          return sendMessage(p_138063_.getSource(), EntityArgument.getPlayers(p_138063_, "targets"), MessageArgument.getMessage(p_138063_, "message"));
       }))));
-      p_138061_.register(Commands.literal("tell").redirect(literalcommandnode));
-      p_138061_.register(Commands.literal("w").redirect(literalcommandnode));
+      p_138061_.register(Commands.literal("tell")
+              .requires(commandSourceStack -> commandSourceStack.hasPermission(0,"minecraft.command.msg"))
+              .redirect(literalcommandnode));
+      p_138061_.register(Commands.literal("w")
+              .requires(commandSourceStack -> commandSourceStack.hasPermission(0,"minecraft.command.msg"))
+              .redirect(literalcommandnode));
    }
 
    private static int sendMessage(CommandSourceStack p_138065_, Collection<ServerPlayer> p_138066_, Component p_138067_) {
