--- a/net/minecraft/server/commands/OpCommand.java
+++ b/net/minecraft/server/commands/OpCommand.java
@@ -17,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_138080_) {
       p_138080_.register(Commands.literal("op").requires((p_138087_) -> {
-         return p_138087_.hasPermission(3);
+         return p_138087_.hasPermission(3,"minecraft.command.op");
       }).then(Commands.argument("targets", GameProfileArgument.gameProfile()).suggests((p_138084_, p_138085_) -> {
          PlayerList playerlist = p_138084_.getSource().getServer().getPlayerList();
          return SharedSuggestionProvider.suggest(playerlist.getPlayers().stream().filter((p_180428_) -> {
