--- a/net/minecraft/server/commands/PardonIpCommand.java
+++ b/net/minecraft/server/commands/PardonIpCommand.java
@@ -17,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_138109_) {
       p_138109_.register(Commands.literal("pardon-ip").requires((p_138116_) -> {
-         return p_138116_.hasPermission(3);
+         return p_138116_.hasPermission(3,"minecraft.command.pardon-ip");
       }).then(Commands.argument("target", StringArgumentType.word()).suggests((p_138113_, p_138114_) -> {
          return SharedSuggestionProvider.suggest(p_138113_.getSource().getServer().getPlayerList().getIpBans().getUserList(), p_138114_);
       }).executes((p_138111_) -> {
