--- a/net/minecraft/server/commands/PardonCommand.java
+++ b/net/minecraft/server/commands/PardonCommand.java
@@ -18,7 +_,7 @@
 
    public static void m_138093_(CommandDispatcher<CommandSourceStack> p_138094_) {
       p_138094_.register(Commands.m_82127_("pardon").requires((p_138101_) -> {
-         return p_138101_.m_6761_(3);
+         return p_138101_.hasPermission(3,"minecraft.command.pardon");
       }).then(Commands.m_82129_("targets", GameProfileArgument.m_94584_()).suggests((p_138098_, p_138099_) -> {
          return SharedSuggestionProvider.m_82967_(p_138098_.getSource().m_81377_().m_6846_().m_11295_().m_5875_(), p_138099_);
       }).executes((p_138096_) -> {
