--- a/net/minecraft/server/commands/SayCommand.java
+++ b/net/minecraft/server/commands/SayCommand.java
@@ -13,7 +_,7 @@
 public class SayCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_138410_) {
       p_138410_.register(Commands.literal("say").requires((p_138414_) -> {
-         return p_138414_.hasPermission(2);
+         return p_138414_.hasPermission(2,"minecraft.command.say");
       }).then(Commands.argument("message", MessageArgument.message()).executes((p_138412_) -> {
          Component component = MessageArgument.getMessage(p_138412_, "message");
          Component component1 = new TranslatableComponent("chat.type.announcement", p_138412_.getSource().getDisplayName(), component);
