--- a/net/minecraft/server/commands/ScoreboardCommand.java
+++ b/net/minecraft/server/commands/ScoreboardCommand.java
@@ -45,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_138469_) {
       p_138469_.register(Commands.literal("scoreboard").requires((p_138552_) -> {
-         return p_138552_.hasPermission(2);
+         return p_138552_.hasPermission(2,"minecraft.command.scoreboard");
       }).then(Commands.literal("objectives").then(Commands.literal("list").executes((p_138585_) -> {
          return listObjectives(p_138585_.getSource());
       })).then(Commands.literal("add").then(Commands.argument("objective", StringArgumentType.word()).then(Commands.argument("criteria", ObjectiveCriteriaArgument.criteria()).executes((p_138583_) -> {
