--- a/net/minecraft/server/commands/SeedCommand.java
+++ b/net/minecraft/server/commands/SeedCommand.java
@@ -14,7 +_,7 @@
 public class SeedCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_138590_, boolean p_138591_) {
       p_138590_.register(Commands.literal("seed").requires((p_138596_) -> {
-         return !p_138591_ || p_138596_.hasPermission(2);
+         return !p_138591_ || p_138596_.hasPermission(2,"minecraft.command.seed");
       }).executes((p_138593_) -> {
          long i = p_138593_.getSource().getLevel().getSeed();
          Component component = ComponentUtils.wrapInSquareBrackets((new TextComponent(String.valueOf(i))).withStyle((p_180514_) -> {
