--- a/net/minecraft/server/commands/SetBlockCommand.java
+++ b/net/minecraft/server/commands/SetBlockCommand.java
@@ -23,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_138602_) {
       p_138602_.register(Commands.literal("setblock").requires((p_138606_) -> {
-         return p_138606_.hasPermission(2);
+         return p_138606_.hasPermission(2,"minecraft.command.setblock");
       }).then(Commands.argument("pos", BlockPosArgument.blockPos()).then(Commands.argument("block", BlockStateArgument.block()).executes((p_138618_) -> {
          return setBlock(p_138618_.getSource(), BlockPosArgument.getLoadedBlockPos(p_138618_, "pos"), BlockStateArgument.getBlock(p_138618_, "block"), SetBlockCommand.Mode.REPLACE, (Predicate<BlockInWorld>)null);
       }).then(Commands.literal("destroy").executes((p_138616_) -> {
