--- a/net/minecraft/server/commands/SetSpawnCommand.java
+++ b/net/minecraft/server/commands/SetSpawnCommand.java
@@ -17,7 +_,7 @@
 public class SetSpawnCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_138644_) {
       p_138644_.register(Commands.literal("spawnpoint").requires((p_138648_) -> {
-         return p_138648_.hasPermission(2);
+         return p_138648_.hasPermission(2,"minecraft.command.spawnpoint");
       }).executes((p_138659_) -> {
          return setSpawn(p_138659_.getSource(), Collections.singleton(p_138659_.getSource().getPlayerOrException()), new BlockPos(p_138659_.getSource().getPosition()), 0.0F);
       }).then(Commands.argument("targets", EntityArgument.players()).executes((p_138657_) -> {
