--- a/net/minecraft/server/commands/SetWorldSpawnCommand.java
+++ b/net/minecraft/server/commands/SetWorldSpawnCommand.java
@@ -11,7 +_,7 @@
 public class SetWorldSpawnCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_138661_) {
       p_138661_.register(Commands.literal("setworldspawn").requires((p_138665_) -> {
-         return p_138665_.hasPermission(2);
+         return p_138665_.hasPermission(2,"minecraft.command.setworldspawn");
       }).executes((p_138673_) -> {
          return setSpawn(p_138673_.getSource(), new BlockPos(p_138673_.getSource().getPosition()), 0.0F);
       }).then(Commands.argument("pos", BlockPosArgument.blockPos()).executes((p_138671_) -> {
