--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -29,7 +_,7 @@
 
    public static void register(CommandDispatcher<CommandSourceStack> p_138815_) {
       p_138815_.register(Commands.literal("summon").requires((p_138819_) -> {
-         return p_138819_.hasPermission(2);
+         return p_138819_.hasPermission(2,"minecraft.command.summon");
       }).then(Commands.argument("entity", EntitySummonArgument.id()).suggests(SuggestionProviders.SUMMONABLE_ENTITIES).executes((p_138832_) -> {
          return spawnEntity(p_138832_.getSource(), EntitySummonArgument.getSummonableEntity(p_138832_, "entity"), p_138832_.getSource().getPosition(), new CompoundTag(), true);
       }).then(Commands.argument("pos", Vec3Argument.vec3()).executes((p_138830_) -> {
@@ -58,6 +_,7 @@
                ((Mob)entity).finalizeSpawn(p_138821_.getLevel(), p_138821_.getLevel().getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.COMMAND, (SpawnGroupData)null, (CompoundTag)null);
             }
 
+            serverlevel.prepareEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND); // Magma - add spawn reason
             if (!serverlevel.tryAddFreshEntityWithPassengers(entity)) {
                throw ERROR_DUPLICATE_UUID.create();
             } else {
