--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -21,6 +_,6 @@
 import net.minecraft.world.entity.SpawnGroupData;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
 
 public class SummonCommand {
    private static final SimpleCommandExceptionType f_138810_ = new SimpleCommandExceptionType(new TranslatableComponent("commands.summon.failed"));
@@ -29,7 +_,7 @@
 
    public static void m_138814_(CommandDispatcher<CommandSourceStack> p_138815_) {
       p_138815_.register(Commands.m_82127_("summon").requires((p_138819_) -> {
-         return p_138819_.m_6761_(2);
+         return p_138819_.hasPermission(2,"minecraft.command.summon");
       }).then(Commands.m_82129_("entity", EntitySummonArgument.m_93335_()).suggests(SuggestionProviders.f_121645_).executes((p_138832_) -> {
          return m_138820_(p_138832_.getSource(), EntitySummonArgument.m_93338_(p_138832_, "entity"), p_138832_.getSource().m_81371_(), new CompoundTag(), true);
       }).then(Commands.m_82129_("pos", Vec3Argument.m_120841_()).executes((p_138830_) -> {
@@ -58,6 +_,7 @@
                ((Mob)entity).m_6518_(p_138821_.m_81372_(), p_138821_.m_81372_().m_6436_(entity.m_142538_()), MobSpawnType.COMMAND, (SpawnGroupData)null, (CompoundTag)null);
             }
 
+            serverlevel.prepareEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND); // Magma - add spawn reason
             if (!serverlevel.m_8860_(entity)) {
                throw f_138811_.create();
             } else {
