--- a/net/minecraft/server/commands/TellRawCommand.java
+++ b/net/minecraft/server/commands/TellRawCommand.java
@@ -12,7 +_,7 @@
 public class TellRawCommand {
    public static void m_139063_(CommandDispatcher<CommandSourceStack> p_139064_) {
       p_139064_.register(Commands.m_82127_("tellraw").requires((p_139068_) -> {
-         return p_139068_.m_6761_(2);
+         return p_139068_.hasPermission(2,"minecraft.command.tellraw");
       }).then(Commands.m_82129_("targets", EntityArgument.m_91470_()).then(Commands.m_82129_("message", ComponentArgument.m_87114_()).executes((p_139066_) -> {
          int i = 0;
 
