--- a/net/minecraft/server/commands/TellRawCommand.java
+++ b/net/minecraft/server/commands/TellRawCommand.java
@@ -12,7 +_,7 @@
 public class TellRawCommand {
    public static void register(CommandDispatcher<CommandSourceStack> p_139064_) {
       p_139064_.register(Commands.literal("tellraw").requires((p_139068_) -> {
-         return p_139068_.hasPermission(2);
+         return p_139068_.hasPermission(2,"minecraft.command.tellraw");
       }).then(Commands.argument("targets", EntityArgument.players()).then(Commands.argument("message", ComponentArgument.textComponent()).executes((p_139066_) -> {
          int i = 0;
 
