--- a/net/minecraft/world/level/levelgen/feature/Feature.java
+++ b/net/minecraft/world/level/levelgen/feature/Feature.java
@@ -52,7 +_,7 @@
 import net.minecraft.world.level.levelgen.feature.configurations.UnderwaterMagmaConfiguration;
 import net.minecraft.world.level.levelgen.feature.configurations.VegetationPatchConfiguration;
 
-public abstract class Feature<FC extends FeatureConfiguration> {
+public abstract class Feature<FC extends FeatureConfiguration> extends net.minecraftforge.registries.ForgeRegistryEntry<Feature<?>> {
    public static final Feature<NoneFeatureConfiguration> NO_OP = register("no_op", new NoOpFeature(NoneFeatureConfiguration.CODEC));
    public static final Feature<TreeConfiguration> TREE = register("tree", new TreeFeature(TreeConfiguration.CODEC));
    public static final Feature<RandomPatchConfiguration> FLOWER = register("flower", new RandomPatchFeature(RandomPatchConfiguration.CODEC));
@@ -154,7 +_,7 @@
    }
 
    protected static boolean isStone(BlockState p_159748_) {
-      return p_159748_.is(BlockTags.BASE_STONE_OVERWORLD);
+      return p_159748_.is(net.minecraftforge.common.Tags.Blocks.STONE);
    }
 
    public static boolean isDirt(BlockState p_159760_) {
