--- a/net/minecraft/client/gui/Gui.java
+++ b/net/minecraft/client/gui/Gui.java
@@ -411,6 +_,7 @@
                   if (this.minecraft.crosshairPickEntity != null && this.minecraft.crosshairPickEntity instanceof LivingEntity && f >= 1.0F) {
                      flag = this.minecraft.player.getCurrentItemAttackStrengthDelay() > 5.0F;
                      flag &= this.minecraft.crosshairPickEntity.isAlive();
+                     flag &= this.minecraft.player.canHit(this.minecraft.crosshairPickEntity, 0);
                   }
 
                   int j = this.screenHeight / 2 - 7 + 16;
@@ -463,6 +_,10 @@
 
          for(MobEffectInstance mobeffectinstance : Ordering.natural().reverse().sortedCopy(collection)) {
             MobEffect mobeffect = mobeffectinstance.getEffect();
+            net.minecraftforge.client.EffectRenderer renderer = net.minecraftforge.client.RenderProperties.getEffectRenderer(mobeffectinstance);
+            if (!renderer.shouldRenderHUD(mobeffectinstance)) continue;
+            // Rebind in case previous renderHUDEffect changed texture
+            RenderSystem.setShaderTexture(0, AbstractContainerScreen.INVENTORY_LOCATION);
             if (mobeffectinstance.showIcon()) {
                int i = this.screenWidth;
                int j = 1;
@@ -500,6 +_,7 @@
                   RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, f1);
                   blit(p_93029_, l + 3, i1 + 3, this.getBlitOffset(), 18, 18, textureatlassprite);
                });
+               renderer.renderHUDEffect(mobeffectinstance,this, p_93029_, i, j, this.getBlitOffset(), f);
             }
          }
 
@@ -620,12 +_,13 @@
    public void renderSelectedItemName(PoseStack p_93070_) {
       this.minecraft.getProfiler().push("selectedItemName");
       if (this.toolHighlightTimer > 0 && !this.lastToolHighlight.isEmpty()) {
-         MutableComponent mutablecomponent = (new TextComponent("")).append(this.lastToolHighlight.getHoverName()).withStyle(this.lastToolHighlight.getRarity().color);
+         MutableComponent mutablecomponent = (new TextComponent("")).append(this.lastToolHighlight.getHoverName()).withStyle(this.lastToolHighlight.getRarity().getStyleModifier());
          if (this.lastToolHighlight.hasCustomHoverName()) {
             mutablecomponent.withStyle(ChatFormatting.ITALIC);
          }
 
-         int i = this.getFont().width(mutablecomponent);
+         Component highlightTip = this.lastToolHighlight.getHighlightTip(mutablecomponent);
+         int i = this.getFont().width(highlightTip);
          int j = (this.screenWidth - i) / 2;
          int k = this.screenHeight - 59;
          if (!this.minecraft.gameMode.canHurtPlayer()) {
@@ -641,7 +_,13 @@
             RenderSystem.enableBlend();
             RenderSystem.defaultBlendFunc();
             fill(p_93070_, j - 2, k - 2, j + i + 2, k + 9 + 2, this.minecraft.options.getBackgroundColor(0));
-            this.getFont().drawShadow(p_93070_, mutablecomponent, (float)j, (float)k, 16777215 + (l << 24));
+            Font font = net.minecraftforge.client.RenderProperties.get(lastToolHighlight).getFont(lastToolHighlight);
+            if (font == null) {
+               this.getFont().drawShadow(p_93070_, highlightTip, (float)j, (float)k, 16777215 + (l << 24));
+            } else {
+               j = (this.screenWidth - font.width(highlightTip)) / 2;
+               font.drawShadow(p_93070_, highlightTip, (float)j, (float)k, 16777215 + (l << 24));
+            }
             RenderSystem.disableBlend();
          }
       }
@@ -1153,7 +_,7 @@
          ItemStack itemstack = this.minecraft.player.getInventory().getSelected();
          if (itemstack.isEmpty()) {
             this.toolHighlightTimer = 0;
-         } else if (!this.lastToolHighlight.isEmpty() && itemstack.is(this.lastToolHighlight.getItem()) && itemstack.getHoverName().equals(this.lastToolHighlight.getHoverName())) {
+         } else if (!this.lastToolHighlight.isEmpty() && itemstack.getItem() == this.lastToolHighlight.getItem() && (itemstack.getHoverName().equals(this.lastToolHighlight.getHoverName()) && itemstack.getHighlightTip(itemstack.getHoverName()).equals(lastToolHighlight.getHighlightTip(lastToolHighlight.getHoverName())))) {
             if (this.toolHighlightTimer > 0) {
                --this.toolHighlightTimer;
             }
