--- a/net/minecraft/world/entity/decoration/Motive.java
+++ b/net/minecraft/world/entity/decoration/Motive.java
@@ -2,7 +_,7 @@
 
 import net.minecraft.core.Registry;
 
-public class Motive {
+public class Motive extends net.minecraftforge.registries.ForgeRegistryEntry<Motive> {
    public static final Motive KEBAB = register("kebab", 16, 16);
    public static final Motive AZTEC = register("aztec", 16, 16);
    public static final Motive ALBAN = register("alban", 16, 16);
