--- a/net/minecraft/tags/TagManager.java
+++ b/net/minecraft/tags/TagManager.java
@@ -29,7 +_,8 @@
 
    public static String getTagDir(ResourceKey<? extends Registry<?>> p_203919_) {
       String s = CUSTOM_REGISTRY_DIRECTORIES.get(p_203919_);
-      return s != null ? s : "tags/" + p_203919_.location().getPath();
+      ResourceLocation registryName = p_203919_.location();
+      return s != null ? s : "tags/" + (registryName.getNamespace().equals("minecraft") ? "" : registryName.getNamespace() + "/") + registryName.getPath();
    }
 
    public CompletableFuture<Void> reload(PreparableReloadListener.PreparationBarrier p_13482_, ResourceManager p_13483_, ProfilerFiller p_13484_, ProfilerFiller p_13485_, Executor p_13486_, Executor p_13487_) {
