--- a/net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SmeltItemFunction.java
@@ -32,7 +_,7 @@
             ItemStack itemstack = optional.get().getResultItem();
             if (!itemstack.isEmpty()) {
                ItemStack itemstack1 = itemstack.copy();
-               itemstack1.setCount(p_81268_.getCount());
+               itemstack1.setCount(p_81268_.getCount() * itemstack.getCount()); //Forge: Support smelting returning multiple
                return itemstack1;
             }
          }
