--- a/net/minecraft/world/inventory/CraftingContainer.java
+++ b/net/minecraft/world/inventory/CraftingContainer.java
@@ -13,6 +_,73 @@
    private final int height;
    private final AbstractContainerMenu menu;
 
+   // CraftBukkit start - add fields
+   public java.util.List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<org.bukkit.entity.HumanEntity>();
+   private net.minecraft.world.item.crafting.Recipe currentRecipe;
+   public Container resultInventory;
+   public Player owner;
+   private int maxStack = MAX_STACK;
+
+   public java.util.List<ItemStack> getContents() {
+      return this.items;
+   }
+
+   public void onOpen(org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity who) {
+      transaction.add(who);
+   }
+
+   public org.bukkit.event.inventory.InventoryType getInvType() {
+      return items.size() == 4 ? org.bukkit.event.inventory.InventoryType.CRAFTING : org.bukkit.event.inventory.InventoryType.WORKBENCH;
+   }
+
+   public void onClose(org.bukkit.craftbukkit.v1_18_R2.entity.CraftHumanEntity who) {
+      transaction.remove(who);
+   }
+
+   public java.util.List<org.bukkit.entity.HumanEntity> getViewers() {
+      return transaction;
+   }
+
+   public org.bukkit.inventory.InventoryHolder getOwner() {
+      return (owner == null) ? null : owner.getBukkitEntity();
+   }
+
+   public void setOwner(Player owner) {
+      this.owner = owner;
+   }
+
+
+   @Override
+   public int getMaxStackSize() {
+      return maxStack;
+   }
+
+   public void setMaxStackSize(int size) {
+      maxStack = size;
+      resultInventory.setMaxStackSize(size);
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return menu instanceof CraftingMenu ? ((CraftingMenu) menu).access.getLocation() : owner.getBukkitEntity().getLocation();
+   }
+
+   @Override
+   public net.minecraft.world.item.crafting.Recipe getCurrentRecipe() {
+      return currentRecipe;
+   }
+
+   @Override
+   public void setCurrentRecipe(net.minecraft.world.item.crafting.Recipe currentRecipe) {
+      this.currentRecipe = currentRecipe;
+   }
+
+   public CraftingContainer(AbstractContainerMenu container, int i, int j, Player player) {
+      this(container, i, j);
+      this.owner = player;
+   }
+   // CraftBukkit end
+
    public CraftingContainer(AbstractContainerMenu p_39325_, int p_39326_, int p_39327_) {
       this.items = NonNullList.withSize(p_39326_ * p_39327_, ItemStack.EMPTY);
       this.menu = p_39325_;
