--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -40,9 +_,10 @@
    protected void checkTakeAchievements(ItemStack p_39558_) {
       p_39558_.onCraftedBy(this.player.level, this.player, this.removeCount);
       if (this.player instanceof ServerPlayer && this.container instanceof AbstractFurnaceBlockEntity) {
-         ((AbstractFurnaceBlockEntity)this.container).awardUsedRecipesAndPopExperience((ServerPlayer)this.player);
+         ((AbstractFurnaceBlockEntity)this.container).awardUsedRecipesAndPopExperience((ServerPlayer)this.player, p_39558_, this.removeCount); // CraftBukkit
       }
 
       this.removeCount = 0;
+      net.minecraftforge.event.ForgeEventFactory.firePlayerSmeltedEvent(this.player, p_39558_);
    }
 }
