--- a/net/minecraft/world/inventory/ItemCombinerMenu.java
+++ b/net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -16,13 +_,8 @@
    private static final int INV_SLOT_END = 30;
    private static final int USE_ROW_SLOT_START = 30;
    private static final int USE_ROW_SLOT_END = 39;
-   protected final ResultContainer resultSlots = new ResultContainer();
-   protected final Container inputSlots = new SimpleContainer(2) {
-      public void setChanged() {
-         super.setChanged();
-         ItemCombinerMenu.this.slotsChanged(this);
-      }
-   };
+   protected final ResultContainer resultSlots;
+   protected final Container inputSlots;
    protected final ContainerLevelAccess access;
    protected final Player player;
 
@@ -34,6 +_,22 @@
 
    public ItemCombinerMenu(@Nullable MenuType<?> p_39773_, int p_39774_, Inventory p_39775_, ContainerLevelAccess p_39776_) {
       super(p_39773_, p_39774_);
+      //Magma start - add location
+      this.resultSlots = new ResultContainer();
+      this.inputSlots = new SimpleContainer(2) {
+         public void setChanged() {
+            super.setChanged();
+            ItemCombinerMenu.this.slotsChanged(this);
+         }
+
+         // CraftBukkit start
+         @Override
+         public org.bukkit.Location getLocation() {
+            return p_39776_.getLocation();
+         }
+         // CraftBukkit end
+      };
+      //Magma end
       this.access = p_39776_;
       this.player = p_39775_.player;
       this.addSlot(new Slot(this.inputSlots, 0, 27, 47));
@@ -82,6 +_,7 @@
    }
 
    public boolean stillValid(Player p_39780_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.access.evaluate((p_39785_, p_39786_) -> {
          return !this.isValidBlock(p_39785_.getBlockState(p_39786_)) ? false : p_39780_.distanceToSqr((double)p_39786_.getX() + 0.5D, (double)p_39786_.getY() + 0.5D, (double)p_39786_.getZ() + 0.5D) <= 64.0D;
       }, true);
