--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -12,8 +_,22 @@
    @Nullable
    private EnderChestBlockEntity f_40101_;
 
-   public PlayerEnderChestContainer() {
+   // CraftBukkit start
+   private final Player owner;
+
+   public org.bukkit.inventory.InventoryHolder getBukkitOwner() {
+      return owner.getBukkitEntity();
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return this.f_40101_ != null ? new org.bukkit.Location(this.f_40101_.m_58904_().getWorld(), this.f_40101_.m_58899_().m_123341_(), this.f_40101_.m_58899_().m_123342_(), this.f_40101_.m_58899_().m_123343_()) : null;
+   }
+
+   public PlayerEnderChestContainer(Player owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void m_40105_(EnderChestBlockEntity p_40106_) {
