--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -12,8 +_,22 @@
    @Nullable
    private EnderChestBlockEntity activeChest;
 
-   public PlayerEnderChestContainer() {
+   // CraftBukkit start
+   private final Player owner;
+
+   public org.bukkit.inventory.InventoryHolder getBukkitOwner() {
+      return owner.getBukkitEntity();
+   }
+
+   @Override
+   public org.bukkit.Location getLocation() {
+      return this.activeChest != null ? new org.bukkit.Location(this.activeChest.getLevel().getWorld(), this.activeChest.getBlockPos().getX(), this.activeChest.getBlockPos().getY(), this.activeChest.getBlockPos().getZ()) : null;
+   }
+
+   public PlayerEnderChestContainer(Player owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void setActiveChest(EnderChestBlockEntity p_40106_) {
