--- a/net/minecraft/world/inventory/SmithingMenu.java
+++ b/net/minecraft/world/inventory/SmithingMenu.java
@@ -17,6 +_,20 @@
    private UpgradeRecipe selectedRecipe;
    private final List<UpgradeRecipe> recipes;
 
+   // CraftBukkit start
+   private org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView bukkitEntity;
+   @Override
+   public org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+
+      org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventorySmithing(access.getLocation(), this.inputSlots, this.resultSlots);
+      bukkitEntity = new org.bukkit.craftbukkit.v1_18_R2.inventory.CraftInventoryView(this.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    public SmithingMenu(int p_40245_, Inventory p_40246_) {
       this(p_40245_, p_40246_, ContainerLevelAccess.NULL);
    }
@@ -54,12 +_,14 @@
    public void createResult() {
       List<UpgradeRecipe> list = this.level.getRecipeManager().getRecipesFor(RecipeType.SMITHING, this.inputSlots, this.level);
       if (list.isEmpty()) {
-         this.resultSlots.setItem(0, ItemStack.EMPTY);
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), ItemStack.EMPTY); // CraftBukkit
       } else {
          this.selectedRecipe = list.get(0);
          ItemStack itemstack = this.selectedRecipe.assemble(this.inputSlots);
          this.resultSlots.setRecipeUsed(this.selectedRecipe);
-         this.resultSlots.setItem(0, itemstack);
+         // CraftBukkit start
+         org.bukkit.craftbukkit.v1_18_R2.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), itemstack);
+         // CraftBukkit end
       }
 
    }
