--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -6,6 +_,6 @@
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Lifecycle;
 import java.net.Proxy;
 import java.util.Collection;
@@ -15,10 +_,10 @@
 import net.minecraft.CrashReport;
 import net.minecraft.SystemReport;
 import net.minecraft.Util;
 import net.minecraft.commands.Commands;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.WorldStem;
 import net.minecraft.server.level.ServerLevel;
@@ -42,6 +_,6 @@
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.WorldData;
 import org.slf4j.Logger;
 
 public class GameTestServer extends MinecraftServer {
@@ -53,7 +_,7 @@
       p_177615_.m_46170_(GameRules.f_46134_).m_46246_(false, (MinecraftServer)null);
       p_177615_.m_46170_(GameRules.f_46150_).m_46246_(false, (MinecraftServer)null);
    });
-   private static final LevelSettings f_177590_ = new LevelSettings("Test Level", GameType.CREATIVE, false, Difficulty.NORMAL, true, f_177589_, DataPackConfig.f_45842_);
+   private static final LevelSettings f_177590_ = new LevelSettings("Test Level", GameType.CREATIVE, false, Difficulty.NORMAL, true, f_177589_, DataPackConfig.f_45842_, Lifecycle.experimental());
    @Nullable
    private MultipleTestTracker f_177591_;
 
@@ -93,12 +_,13 @@
    public boolean m_7038_() {
       this.m_129823_(new PlayerList(this, this.m_206579_(), this.f_129745_, 1) {
       });
-      this.m_130006_();
+      if (!net.minecraftforge.server.ServerLifecycleHooks.handleServerAboutToStart(this)) return false;
+      this.m_130006_(); // Magma
       ServerLevel serverlevel = this.m_129783_();
       serverlevel.m_8733_(this.f_177588_, 0.0F);
       int i = 20000000;
       serverlevel.m_8606_(20000000, 20000000, false, false);
-      return true;
+      return net.minecraftforge.server.ServerLifecycleHooks.handleServerStarting(this);
    }
 
    public void m_5705_(BooleanSupplier p_177619_) {
@@ -200,6 +_,11 @@
 
    public boolean m_6102_() {
       return false;
+   }
+
+   @Override
+   public org.bukkit.command.CommandSender getBukkitSender(net.minecraft.commands.CommandSourceStack wrapper) {
+      return console;
    }
 
    public boolean m_7779_(GameProfile p_177617_) {
