--- a/net/minecraft/world/level/LevelWriter.java
+++ b/net/minecraft/world/level/LevelWriter.java
@@ -24,7 +_,17 @@
 
    boolean m_7740_(BlockPos p_46957_, boolean p_46958_, @Nullable Entity p_46959_, int p_46960_);
 
+   default void prepareEntity(Entity entity, @Nullable org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason) {
+      entity.spawnReason = (spawnReason != null ? spawnReason : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
    default boolean m_7967_(Entity p_46964_) {
       return false;
    }
+
+   // CraftBukkit start
+   default boolean addFreshEntity(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      return false;
+   }
+   // CraftBukkit end
 }
