--- a/net/minecraft/world/level/ServerLevelAccessor.java
+++ b/net/minecraft/world/level/ServerLevelAccessor.java
@@ -6,7 +_,23 @@
 public interface ServerLevelAccessor extends LevelAccessor {
    ServerLevel getLevel();
 
+   default void prepareEntity(Entity entity, @org.jetbrains.annotations.Nullable org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason) {
+      entity.spawnReason = (spawnReason != null ? spawnReason : org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+   }
+
    default void addFreshEntityWithPassengers(Entity p_47206_) {
-      p_47206_.getSelfAndPassengers().forEach(this::addFreshEntity);
-   }
+      // CraftBukkit start
+      //Magma - move spawn reason to the entity class
+      this.addFreshEntityWithPassengers(p_47206_, p_47206_.spawnReason);
+   }
+
+   default void addFreshEntityWithPassengers(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+      entity.getSelfAndPassengers().forEach((e) -> this.addFreshEntity(e, reason));
+   }
+
+   @Override
+   default ServerLevel getMinecraftWorld() {
+      return getLevel();
+   }
+   // CraftBukkit end
 }
