--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -15,9 +_,19 @@
    private final WorldData worldData;
    private final ServerLevelData wrapped;
 
+   //Magma start - add level name
+   private String levelName;
+
    public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_) {
       this.worldData = p_78079_;
       this.wrapped = p_78080_;
+
+      this.levelName = worldData.getLevelName();
+      //Magma end
+   }
+
+   public ServerLevelData getWrapped() {
+      return wrapped;
    }
 
    public int getXSpawn() {
@@ -44,9 +_,15 @@
       return this.wrapped.getDayTime();
    }
 
+   //Magma start - add level name
+   public void setLevelName(String levelName) {
+      this.levelName = levelName;
+   }
+
    public String getLevelName() {
-      return this.worldData.getLevelName();
+      return levelName;
    }
+   //Magma end
 
    public int getClearWeatherTime() {
       return this.wrapped.getClearWeatherTime();
