--- a/net/minecraft/world/level/storage/LevelSummary.java
+++ b/net/minecraft/world/level/storage/LevelSummary.java
@@ -154,6 +_,10 @@
       }
    }
 
+   public boolean isExperimental() {
+      return this.settings.getLifecycle().equals(com.mojang.serialization.Lifecycle.experimental());
+   }
+
    public static enum BackupStatus {
       NONE(false, false, ""),
       DOWNGRADE(true, true, "downgrade"),
