--- a/net/minecraft/commands/arguments/coordinates/BlockPosArgument.java
+++ b/net/minecraft/commands/arguments/coordinates/BlockPosArgument.java
@@ -30,9 +_,9 @@
 
    public static BlockPos getLoadedBlockPos(CommandContext<CommandSourceStack> p_118243_, String p_118244_) throws CommandSyntaxException {
       BlockPos blockpos = p_118243_.getArgument(p_118244_, Coordinates.class).getBlockPos(p_118243_.getSource());
-      if (!p_118243_.getSource().getLevel().hasChunkAt(blockpos)) {
+      if (!p_118243_.getSource().getUnsidedLevel().hasChunkAt(blockpos)) {
          throw ERROR_NOT_LOADED.create();
-      } else if (!p_118243_.getSource().getLevel().isInWorldBounds(blockpos)) {
+      } else if (!p_118243_.getSource().getUnsidedLevel().isInWorldBounds(blockpos)) {
          throw ERROR_OUT_OF_WORLD.create();
       } else {
          return blockpos;
