--- a/net/minecraft/world/level/levelgen/feature/stateproviders/BlockStateProviderType.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/BlockStateProviderType.java
@@ -3,7 +_,7 @@
 import com.mojang.serialization.Codec;
 import net.minecraft.core.Registry;
 
-public class BlockStateProviderType<P extends BlockStateProvider> {
+public class BlockStateProviderType<P extends BlockStateProvider> extends net.minecraftforge.registries.ForgeRegistryEntry<BlockStateProviderType<?>> {
    public static final BlockStateProviderType<SimpleStateProvider> SIMPLE_STATE_PROVIDER = register("simple_state_provider", SimpleStateProvider.CODEC);
    public static final BlockStateProviderType<WeightedStateProvider> WEIGHTED_STATE_PROVIDER = register("weighted_state_provider", WeightedStateProvider.CODEC);
    public static final BlockStateProviderType<NoiseThresholdProvider> NOISE_THRESHOLD_PROVIDER = register("noise_threshold_provider", NoiseThresholdProvider.CODEC);
