--- a/net/minecraft/world/entity/ai/attributes/AttributeModifier.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeModifier.java
@@ -2,13 +_,14 @@
 
 import com.mojang.logging.LogUtils;
 import io.netty.util.internal.ThreadLocalRandom;
-import java.util.Objects;
-import java.util.UUID;
-import java.util.function.Supplier;
-import javax.annotation.Nullable;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.util.Mth;
 import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Supplier;
 
 public class AttributeModifier {
    private static final Logger f_22189_ = LogUtils.getLogger();
