--- a/net/minecraft/world/entity/ai/attributes/RangedAttribute.java
+++ b/net/minecraft/world/entity/ai/attributes/RangedAttribute.java
@@ -4,7 +_,7 @@
 
 public class RangedAttribute extends Attribute {
    private final double minValue;
-   private final double maxValue;
+   public double maxValue; // Spigot
 
    public RangedAttribute(String p_22310_, double p_22311_, double p_22312_, double p_22313_) {
       super(p_22310_, p_22311_);
@@ -28,6 +_,8 @@
    }
 
    public double sanitizeValue(double p_22315_) {
+      if (p_22315_ != p_22315_) return getDefaultValue(); // CraftBukkit
+
       return Mth.clamp(p_22315_, this.minValue, this.maxValue);
    }
 }
