--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -55,6 +_,8 @@
                CauldronInteraction.bootStrap();
                ArgumentTypes.bootStrap();
                Registry.freezeBuiltins();
+               net.minecraftforge.registries.GameData.vanillaSnapshot();
+               if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
                wrapStreams();
             }
          }
@@ -128,7 +_,6 @@
          validateThatAllBiomeFeaturesHaveBiomeFilter();
       }
 
-      DefaultAttributes.validate();
    }
 
    private static void validateThatAllBiomeFeaturesHaveBiomeFilter() {
