--- a/net/minecraft/server/ReloadableServerResources.java
+++ b/net/minecraft/server/ReloadableServerResources.java
@@ -28,18 +_,22 @@
    private static final Logger LOGGER = LogUtils.getLogger();
    private static final CompletableFuture<Unit> DATA_RELOAD_INITIAL_TASK = CompletableFuture.completedFuture(Unit.INSTANCE);
    public Commands commands;
-   private final RecipeManager recipes = new RecipeManager();
+   private final RecipeManager recipes;
    private final TagManager tagManager;
    private final PredicateManager predicateManager = new PredicateManager();
    private final LootTables lootTables = new LootTables(this.predicateManager);
    private final ItemModifierManager itemModifierManager = new ItemModifierManager(this.predicateManager, this.lootTables);
-   private final ServerAdvancementManager advancements = new ServerAdvancementManager(this.predicateManager);
+   private final ServerAdvancementManager advancements;
    private final ServerFunctionLibrary functionLibrary;
 
    public ReloadableServerResources(RegistryAccess.Frozen p_206857_, Commands.CommandSelection p_206858_, int p_206859_) {
       this.tagManager = new TagManager(p_206857_);
       this.commands = new Commands(p_206858_);
       this.functionLibrary = new ServerFunctionLibrary(p_206859_, this.commands.getDispatcher());
+      // Forge: Create context object and pass it to the recipe manager.
+      this.context = new net.minecraftforge.common.crafting.conditions.ConditionContext(this.tagManager);
+      this.recipes = new RecipeManager(context);
+      this.advancements = new ServerAdvancementManager(this.predicateManager, context);
    }
 
    public ServerFunctionLibrary getFunctionLibrary() {
@@ -76,7 +_,9 @@
 
    public static CompletableFuture<ReloadableServerResources> loadResources(ResourceManager p_206862_, RegistryAccess.Frozen p_206863_, Commands.CommandSelection p_206864_, int p_206865_, Executor p_206866_, Executor p_206867_) {
       ReloadableServerResources reloadableserverresources = new ReloadableServerResources(p_206863_, p_206864_, p_206865_);
-      return SimpleReloadInstance.create(p_206862_, reloadableserverresources.listeners(), p_206866_, p_206867_, DATA_RELOAD_INITIAL_TASK, LOGGER.isDebugEnabled()).done().thenApply((p_206880_) -> {
+      List<PreparableReloadListener> listeners = new java.util.ArrayList<>(reloadableserverresources.listeners());
+      listeners.addAll(net.minecraftforge.event.ForgeEventFactory.onResourceReload(reloadableserverresources));
+      return SimpleReloadInstance.create(p_206862_, listeners, p_206866_, p_206867_, DATA_RELOAD_INITIAL_TASK, LOGGER.isDebugEnabled()).done().thenApply((p_206880_) -> {
          return reloadableserverresources;
       });
    }
@@ -86,6 +_,7 @@
          updateRegistryTags(p_206869_, p_206884_);
       });
       Blocks.rebuildCache();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent(p_206869_, false, false));
    }
 
    private static <T> void updateRegistryTags(RegistryAccess p_206871_, TagManager.LoadResult<T> p_206872_) {
@@ -96,5 +_,16 @@
          return p_206874_.getValue().getValues();
       }));
       p_206871_.registryOrThrow(resourcekey).bindTags(map);
+   }
+
+   private final net.minecraftforge.common.crafting.conditions.ICondition.IContext context;
+
+   /**
+    * Exposes the current condition context for usage in other reload listeners.<br>
+    * This is not useful outside the reloading stage.
+    * @return The condition context for the currently active reload.
+    */
+   public net.minecraftforge.common.crafting.conditions.ICondition.IContext getConditionContext() {
+      return this.context;
    }
 }
