--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -26,18 +_,36 @@
    public static final Gson GSON = (new GsonBuilder()).create();
    public AdvancementList advancements = new AdvancementList();
    private final PredicateManager predicateManager;
+   private final net.minecraftforge.common.crafting.conditions.ICondition.IContext context; //Forge: add context
 
+   /** @deprecated Forge: use {@linkplain ServerAdvancementManager#ServerAdvancementManager(PredicateManager, net.minecraftforge.common.crafting.conditions.ICondition.IContext) constructor with context}. */
+   @Deprecated
    public ServerAdvancementManager(PredicateManager p_136027_) {
+      this(p_136027_, net.minecraftforge.common.crafting.conditions.ICondition.IContext.EMPTY);
+   }
+
+   public ServerAdvancementManager(PredicateManager p_136027_, net.minecraftforge.common.crafting.conditions.ICondition.IContext context) {
       super(GSON, "advancements");
       this.predicateManager = p_136027_;
+      this.context = context;
    }
 
    protected void apply(Map<ResourceLocation, JsonElement> p_136034_, ResourceManager p_136035_, ProfilerFiller p_136036_) {
       Map<ResourceLocation, Advancement.Builder> map = Maps.newHashMap();
       p_136034_.forEach((p_136039_, p_136040_) -> {
+         // Spigot start
+         if (org.spigotmc.SpigotConfig.disabledAdvancements != null && (org.spigotmc.SpigotConfig.disabledAdvancements.contains("*") || org.spigotmc.SpigotConfig.disabledAdvancements.contains(p_136039_.toString()) || org.spigotmc.SpigotConfig.disabledAdvancements.contains(p_136039_.getNamespace()))) {
+            return;
+         }
+         // Spigot end
+
          try {
             JsonObject jsonobject = GsonHelper.convertToJsonObject(p_136040_, "advancement");
-            Advancement.Builder advancement$builder = Advancement.Builder.fromJson(jsonobject, new DeserializationContext(p_136039_, this.predicateManager));
+            Advancement.Builder advancement$builder = Advancement.Builder.fromJson(jsonobject, new DeserializationContext(p_136039_, this.predicateManager), this.context);
+            if (advancement$builder == null) {
+                LOGGER.debug("Skipping loading advancement {} as it's conditions were not met", p_136039_);
+                return;
+            }
             map.put(p_136039_, advancement$builder);
          } catch (Exception exception) {
             LOGGER.error("Parsing error loading custom advancement {}: {}", p_136039_, exception.getMessage());
