--- a/net/minecraft/client/renderer/FogRenderer.java
+++ b/net/minecraft/client/renderer/FogRenderer.java
@@ -173,24 +_,43 @@
 
       if (fogRed != 0.0F && fogGreen != 0.0F && fogBlue != 0.0F) {
          float f8 = Math.min(1.0F / fogRed, Math.min(1.0F / fogGreen, 1.0F / fogBlue));
+         // Forge: fix MC-4647 and MC-10480
+         if (Float.isInfinite(f8)) f8 = Math.nextAfter(f8, 0.0);
          fogRed = fogRed * (1.0F - f6) + fogRed * f8 * f6;
          fogGreen = fogGreen * (1.0F - f6) + fogGreen * f8 * f6;
          fogBlue = fogBlue * (1.0F - f6) + fogBlue * f8 * f6;
       }
 
+      net.minecraftforge.client.event.EntityViewRenderEvent.FogColors event = new net.minecraftforge.client.event.EntityViewRenderEvent.FogColors(p_109019_, p_109020_, fogRed, fogGreen, fogBlue);
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+
+      fogRed = event.getRed();
+      fogGreen = event.getGreen();
+      fogBlue = event.getBlue();
+
       RenderSystem.clearColor(fogRed, fogGreen, fogBlue, 0.0F);
    }
 
    public static void setupNoFog() {
       RenderSystem.setShaderFogStart(Float.MAX_VALUE);
    }
-
+   @Deprecated // FORGE: Pass in partialTicks
    public static void setupFog(Camera p_109025_, FogRenderer.FogMode p_109026_, float p_109027_, boolean p_109028_) {
+      setupFog(p_109025_, p_109026_, p_109027_, p_109028_, 0);
+   }
+
+   public static void setupFog(Camera p_109025_, FogRenderer.FogMode p_109026_, float p_109027_, boolean p_109028_, float partialTicks) {
       FogType fogtype = p_109025_.getFluidInCamera();
       Entity entity = p_109025_.getEntity();
       FogShape fogshape = FogShape.SPHERE;
       float f;
       float f1;
+      // TODO: remove this hook in 1.19
+      float hook = net.minecraftforge.client.ForgeHooksClient.getFogDensity(p_109026_, p_109025_, partialTicks, 0.1F);
+      if (hook >= 0) {
+         f = -8.0f;
+         f1 = hook * 0.5F;
+      } else
       if (fogtype == FogType.LAVA) {
          if (entity.isSpectator()) {
             f = -8.0F;
@@ -253,6 +_,7 @@
       RenderSystem.setShaderFogStart(f);
       RenderSystem.setShaderFogEnd(f1);
       RenderSystem.setShaderFogShape(fogshape);
+      net.minecraftforge.client.ForgeHooksClient.onFogRender(p_109026_, p_109025_, partialTicks, f1);
    }
 
    public static void levelFogColor() {
