--- a/net/minecraft/client/renderer/OutlineBufferSource.java
+++ b/net/minecraft/client/renderer/OutlineBufferSource.java
@@ -105,5 +_,9 @@
       public void endVertex() {
          this.delegate.vertex(this.x, this.y, this.z).color(this.defaultR, this.defaultG, this.defaultB, this.defaultA).uv(this.u, this.v).endVertex();
       }
+
+      public com.mojang.blaze3d.vertex.VertexFormat getVertexFormat() {
+         return this.delegate.getVertexFormat();
+      }
    }
 }
