--- a/net/minecraft/client/renderer/Sheets.java
+++ b/net/minecraft/client/renderer/Sheets.java
@@ -127,7 +_,8 @@
    }
 
    private static Material createSignMaterial(WoodType p_173386_) {
-      return new Material(SIGN_SHEET, new ResourceLocation("entity/signs/" + p_173386_.name()));
+      ResourceLocation location = new ResourceLocation(p_173386_.name());
+      return new Material(SIGN_SHEET, new ResourceLocation(location.getNamespace(), "entity/signs/" + location.getPath()));
    }
 
    public static Material getSignMaterial(WoodType p_173382_) {
@@ -174,5 +_,12 @@
       default:
          return p_110773_;
       }
+   }
+
+   /**
+    * Not threadsafe. Enqueue it in client setup.
+    */
+   public static void addWoodType(WoodType woodType) {
+      SIGN_MATERIALS.put(woodType, createSignMaterial(woodType));
    }
 }
