--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -116,6 +_,12 @@
       return i;
    }
 
+   // Normal compare sorts by path first, this compares namespace first.
+   public int compareNamespaced(ResourceLocation o) {
+      int ret = this.namespace.compareTo(o.namespace);
+      return ret != 0 ? ret : this.path.compareTo(o.path);
+   }
+
    public String toDebugFileName() {
       return this.toString().replace('/', '_').replace(':', '_');
    }
@@ -141,7 +_,7 @@
       return p_135817_ >= '0' && p_135817_ <= '9' || p_135817_ >= 'a' && p_135817_ <= 'z' || p_135817_ == '_' || p_135817_ == ':' || p_135817_ == '/' || p_135817_ == '.' || p_135817_ == '-';
    }
 
-   private static boolean isValidPath(String p_135842_) {
+   public static boolean isValidPath(String p_135842_) { // Magma - private -> public
       for(int i = 0; i < p_135842_.length(); ++i) {
          if (!validPathChar(p_135842_.charAt(i))) {
             return false;
