--- a/net/minecraft/client/gui/components/BossHealthOverlay.java
+++ b/net/minecraft/client/gui/components/BossHealthOverlay.java
@@ -34,6 +_,9 @@
 
          for(LerpingBossEvent lerpingbossevent : this.events.values()) {
             int k = i / 2 - 91;
+            net.minecraftforge.client.event.RenderGameOverlayEvent.BossInfo event =
+               net.minecraftforge.client.ForgeHooksClient.renderBossEventPre(p_93705_, this.minecraft.getWindow(), lerpingbossevent, k, j, 10 + this.minecraft.font.lineHeight);
+            if (!event.isCanceled()) {
             RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
             RenderSystem.setShaderTexture(0, GUI_BARS_LOCATION);
             this.drawBar(p_93705_, k, j, lerpingbossevent);
@@ -42,7 +_,9 @@
             int i1 = i / 2 - l / 2;
             int j1 = j - 9;
             this.minecraft.font.drawShadow(p_93705_, component, (float)i1, (float)j1, 16777215);
-            j += 10 + 9;
+            }
+            j += event.getIncrement();
+            net.minecraftforge.client.ForgeHooksClient.renderBossEventPost(p_93705_, this.minecraft.getWindow());
             if (j >= this.minecraft.getWindow().getGuiScaledHeight() / 3) {
                break;
             }
