--- a/net/minecraft/client/gui/components/DebugScreenOverlay.java
+++ b/net/minecraft/client/gui/components/DebugScreenOverlay.java
@@ -399,6 +_,7 @@
             list.add("");
             list.add(ChatFormatting.UNDERLINE + "Targeted Entity");
             list.add(String.valueOf((Object)Registry.ENTITY_TYPE.getKey(entity.getType())));
+            entity.getType().builtInRegistryHolder().tags().forEach(t -> list.add("#" + t.location()));
          }
 
          return list;
