--- a/net/minecraft/world/level/levelgen/flat/FlatLevelGeneratorSettings.java
+++ b/net/minecraft/world/level/levelgen/flat/FlatLevelGeneratorSettings.java
@@ -147,7 +_,7 @@
          }
       }
 
-      return Holder.direct(Biome.BiomeBuilder.from(biome).generationSettings(biomegenerationsettings$builder.build()).build());
+      return Holder.direct(Biome.BiomeBuilder.from(biome).generationSettings(biomegenerationsettings$builder.build()).build().setRegistryName(biome.getRegistryName()));
    }
 
    public Optional<HolderSet<StructureSet>> structureOverrides() {
