--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -39,6 +_,14 @@
       }
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.f_20834_ = true;
+      this.f_20835_ = 0;
+      this.f_20836_ = ticks;
+      this.f_20837_.m_135381_(this.f_20838_, this.f_20836_);
+   }
+   // CraftBukkit end
    public void m_20847_(CompoundTag p_20848_) {
       p_20848_.m_128379_("Saddle", this.m_20851_());
    }
@@ -54,4 +_,4 @@
    public boolean m_20851_() {
       return this.f_20837_.m_135370_(this.f_20839_);
    }
 }
