--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -39,6 +_,14 @@
       }
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.boosting = true;
+      this.boostTime = 0;
+      this.boostTimeTotal = ticks;
+      this.entityData.set(this.boostTimeAccessor, this.boostTimeTotal);
+   }
+   // CraftBukkit end
    public void addAdditionalSaveData(CompoundTag p_20848_) {
       p_20848_.putBoolean("Saddle", this.hasSaddle());
    }
