--- a/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -50,7 +_,8 @@
    private volatile int totalChunks;
    private volatile int converted;
    private volatile int skipped;
-   private final Object2FloatMap<ResourceKey<Level>> progressMap = Object2FloatMaps.synchronize(new Object2FloatOpenCustomHashMap<>(Util.identityStrategy()));
+   private final Object2FloatMap<ResourceKey<Level>> progressMap = Object2FloatMaps.synchronize(new Object2FloatOpenCustomHashMap(Util.identityStrategy())); // CraftBukkit
+
    private volatile Component status = new TranslatableComponent("optimizeWorld.stage.counting");
    private static final Pattern REGEX = Pattern.compile("^r\\.(-?[0-9]+)\\.(-?[0-9]+)\\.mca$");
    private final DimensionDataStorage overworldDataStorage;
@@ -125,7 +_,7 @@
                         ChunkGenerator chunkgenerator = this.worldGenSettings.dimensions().get(WorldGenSettings.levelToLevelStem(resourcekey2)).generator();
                         CompoundTag compoundtag1 = chunkstorage.upgradeChunkTag(resourcekey2, () -> {
                            return this.overworldDataStorage;
-                        }, compoundtag, chunkgenerator.getTypeNameForDataFixer());
+                        }, compoundtag, chunkgenerator.getTypeNameForDataFixer(),chunkpos,  null); // Magma - maybe fix this?!
                         ChunkPos chunkpos1 = new ChunkPos(compoundtag1.getInt("xPos"), compoundtag1.getInt("zPos"));
                         if (!chunkpos1.equals(chunkpos)) {
                            LOGGER.warn("Chunk {} has invalid position {}", chunkpos, chunkpos1);
