--- a/net/minecraft/data/DataGenerator.java
+++ b/net/minecraft/data/DataGenerator.java
@@ -16,10 +_,11 @@
    private final Collection<Path> inputFolders;
    private final Path outputFolder;
    private final List<DataProvider> providers = Lists.newArrayList();
+   private final List<DataProvider> providerView = java.util.Collections.unmodifiableList(providers);
 
    public DataGenerator(Path p_123911_, Collection<Path> p_123912_) {
       this.outputFolder = p_123911_;
-      this.inputFolders = p_123912_;
+      this.inputFolders = Lists.newArrayList(p_123912_);
    }
 
    public Collection<Path> getInputFolders() {
@@ -38,6 +_,7 @@
 
       for(DataProvider dataprovider : this.providers) {
          LOGGER.info("Starting provider: {}", (Object)dataprovider.getName());
+         net.minecraftforge.fml.StartupMessageManager.addModMessage("Generating: " + dataprovider.getName());
          stopwatch1.start();
          dataprovider.run(hashcache);
          stopwatch1.stop();
@@ -51,6 +_,14 @@
 
    public void addProvider(DataProvider p_123915_) {
       this.providers.add(p_123915_);
+   }
+
+   public List<DataProvider> getProviders() {
+       return this.providerView;
+   }
+
+   public void addInput(Path value) {
+      this.inputFolders.add(value);
    }
 
    static {
