--- a/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/EnderDragonPhaseManager.java
@@ -23,6 +_,26 @@
             this.currentPhase.end();
          }
 
+         // CraftBukkit start - Call EnderDragonChangePhaseEvent
+         //Magma start - don't call event if phase is unknown
+         final org.bukkit.entity.EnderDragon.Phase bukkitPhase = org.bukkit.craftbukkit.v1_18_R2.entity.CraftEnderDragon.getBukkitPhase(p_31417_);
+
+         if (bukkitPhase != org.bukkit.entity.EnderDragon.Phase.UNKNOWN) {
+            org.bukkit.event.entity.EnderDragonChangePhaseEvent event = new org.bukkit.event.entity.EnderDragonChangePhaseEvent(
+                    (org.bukkit.craftbukkit.v1_18_R2.entity.CraftEnderDragon) this.dragon.getBukkitEntity(),
+                    (this.currentPhase == null) ? null : org.bukkit.craftbukkit.v1_18_R2.entity.CraftEnderDragon.getBukkitPhase(this.currentPhase.getPhase()),
+                    bukkitPhase
+            );
+            this.dragon.level.getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               return;
+            }
+            p_31417_ = org.bukkit.craftbukkit.v1_18_R2.entity.CraftEnderDragon.getMinecraftPhase(event.getNewPhase());
+         }
+         //Magma end
+
+         // CraftBukkit end
+
          this.currentPhase = this.getPhase(p_31417_);
          if (!this.dragon.level.isClientSide) {
             this.dragon.getEntityData().set(EnderDragon.DATA_PHASE, p_31417_.getId());
