--- a/net/minecraft/client/renderer/texture/MipmapGenerator.java
+++ b/net/minecraft/client/renderer/texture/MipmapGenerator.java
@@ -34,15 +_,19 @@
             }
          }
 
+         int maxMipmapLevel = net.minecraftforge.client.ForgeHooksClient.getMaxMipmapLevel(p_118055_.getWidth(), p_118055_.getHeight());
          for(int k1 = 1; k1 <= p_118056_; ++k1) {
             NativeImage nativeimage1 = anativeimage[k1 - 1];
-            NativeImage nativeimage = new NativeImage(nativeimage1.getWidth() >> 1, nativeimage1.getHeight() >> 1, false);
-            int k = nativeimage.getWidth();
-            int l = nativeimage.getHeight();
+            // Forge: guard against invalid texture size, because we allow generating mipmaps regardless of texture sizes
+            NativeImage nativeimage = new NativeImage(Math.max(1, nativeimage1.getWidth() >> 1), Math.max(1, nativeimage1.getHeight() >> 1), false);
+            if (k1 <= maxMipmapLevel) {
+               int k = nativeimage.getWidth();
+               int l = nativeimage.getHeight();
 
-            for(int i1 = 0; i1 < k; ++i1) {
-               for(int j1 = 0; j1 < l; ++j1) {
-                  nativeimage.setPixelRGBA(i1, j1, alphaBlend(nativeimage1.getPixelRGBA(i1 * 2 + 0, j1 * 2 + 0), nativeimage1.getPixelRGBA(i1 * 2 + 1, j1 * 2 + 0), nativeimage1.getPixelRGBA(i1 * 2 + 0, j1 * 2 + 1), nativeimage1.getPixelRGBA(i1 * 2 + 1, j1 * 2 + 1), flag));
+               for (int i1 = 0; i1 < k; ++i1) {
+                  for (int j1 = 0; j1 < l; ++j1) {
+                     nativeimage.setPixelRGBA(i1, j1, alphaBlend(nativeimage1.getPixelRGBA(i1 * 2 + 0, j1 * 2 + 0), nativeimage1.getPixelRGBA(i1 * 2 + 1, j1 * 2 + 0), nativeimage1.getPixelRGBA(i1 * 2 + 0, j1 * 2 + 1), nativeimage1.getPixelRGBA(i1 * 2 + 1, j1 * 2 + 1), flag));
+                  }
                }
             }
 
