--- a/net/minecraft/client/renderer/texture/TextureManager.java
+++ b/net/minecraft/client/renderer/texture/TextureManager.java
@@ -167,6 +_,7 @@
    public void release(ResourceLocation p_118514_) {
       AbstractTexture abstracttexture = this.getTexture(p_118514_, MissingTextureAtlasSprite.getTexture());
       if (abstracttexture != MissingTextureAtlasSprite.getTexture()) {
+         this.byPath.remove(p_118514_); // Forge: fix MC-98707
          TextureUtil.releaseTextureId(abstracttexture.getId());
       }
 
