--- a/net/minecraft/client/renderer/texture/TextureAtlasSprite.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlasSprite.java
@@ -145,6 +_,7 @@
 
    void upload(int p_118376_, int p_118377_, NativeImage[] p_118378_) {
       for(int i = 0; i < this.mainImage.length; ++i) {
+         if ((this.width >> i <= 0) || (this.height >> i <= 0)) break;
          p_118378_[i].upload(i, this.x >> i, this.y >> i, p_118376_ >> i, p_118377_ >> i, this.width >> i, this.height >> i, this.mainImage.length > 1, false);
       }
 
@@ -389,7 +_,8 @@
             int j = p_118446_.width >> i;
             int k = p_118446_.height >> i;
             if (this.activeFrame[i] == null) {
-               this.activeFrame[i] = new NativeImage(j, k, false);
+               // Forge: guard against invalid texture size, because we allow generating mipmaps regardless of texture sizes
+               this.activeFrame[i] = new NativeImage(Math.max(1, j), Math.max(1, k), false);
             }
          }
 
@@ -404,6 +_,8 @@
             for(int k = 0; k < this.activeFrame.length; ++k) {
                int l = TextureAtlasSprite.this.width >> k;
                int i1 = TextureAtlasSprite.this.height >> k;
+               // Forge: guard against invalid texture size, because we allow generating mipmaps regardless of texture sizes
+               if (l == 0 || i1 == 0) continue;
 
                for(int j1 = 0; j1 < i1; ++j1) {
                   for(int k1 = 0; k1 < l; ++k1) {
@@ -438,5 +_,15 @@
          }
 
       }
+   }
+
+   // Forge Start
+   public int getPixelRGBA(int frameIndex, int x, int y) {
+       if (this.animatedTexture != null) {
+           x += this.animatedTexture.getFrameX(frameIndex) * this.width;
+           y += this.animatedTexture.getFrameY(frameIndex) * this.height;
+       }
+
+       return this.mainImage[0].getPixelRGBA(x, y);
    }
 }
